diff --git a/gfx/wr/webrender/src/renderer/mod.rs b/gfx/wr/webrender/src/renderer/mod.rs
index a292e9b89912..06eea476ccf8 100644
--- a/gfx/wr/webrender/src/renderer/mod.rs
+++ b/gfx/wr/webrender/src/renderer/mod.rs
@@ -265,10 +265,12 @@ const GPU_TAG_COMPOSITE: GpuProfileTag = GpuProfileTag {
     color: debug_colors::TOMATO,
 };
 
+#[derive(Debug)]
 // Defines the content that we will draw to a given swapchain / layer, calculated
 // after occlusion culling.
 struct SwapChainLayer {
-    occlusion: occlusion::FrontToBackBuilder<usize>,
+    opaque_items: Vec<occlusion::Item<usize>>,
+    alpha_items: Vec<occlusion::Item<usize>>,
     clear_tiles: Vec<occlusion::Item<usize>>,
 }
 
@@ -3381,6 +3383,7 @@ impl Renderer {
         projection: &default::Transform3D<f32>,
         results: &mut RenderResults,
         partial_present_mode: Option<PartialPresentMode>,
+        occlusion: &occlusion::FrontToBackBuilder<usize>,
         layer: &SwapChainLayer,
     ) {
         self.device.bind_draw_target(draw_target);
@@ -3407,7 +3410,7 @@ impl Renderer {
                 // on Mali-G77 we have observed artefacts when calling glClear (even with
                 // the empty scissor rect set) after calling eglSetDamageRegion with an
                 // empty damage region. So avoid clearing in that case. See bug 1709548.
-                if !dirty_rect.is_empty() && layer.occlusion.test(&dirty_rect) {
+                if !dirty_rect.is_empty() && occlusion.test(&dirty_rect) {
                     // We have a single dirty rect, so clear only that
                     self.device.clear_target(clear_color,
                                              None,
@@ -3423,12 +3426,11 @@ impl Renderer {
         }
 
         // Draw opaque tiles
-        let opaque_items = layer.occlusion.opaque_items();
-        if !opaque_items.is_empty() {
+        if !layer.opaque_items.is_empty() {
             let opaque_sampler = self.gpu_profiler.start_sampler(GPU_SAMPLER_TAG_OPAQUE);
             self.set_blend(false, FramebufferKind::Main);
             self.draw_tile_list(
-                opaque_items.iter(),
+                layer.opaque_items.iter(),
                 &composite_state,
                 &composite_state.external_surfaces,
                 projection,
@@ -3453,13 +3455,12 @@ impl Renderer {
         }
 
         // Draw alpha tiles
-        let alpha_items = layer.occlusion.alpha_items();
-        if !alpha_items.is_empty() {
+        if !layer.alpha_items.is_empty() {
             let transparent_sampler = self.gpu_profiler.start_sampler(GPU_SAMPLER_TAG_TRANSPARENT);
             self.set_blend(true, FramebufferKind::Main);
             self.set_blend_mode_premultiplied_alpha(FramebufferKind::Main);
             self.draw_tile_list(
-                alpha_items.iter().rev(),
+                layer.alpha_items.iter(),
                 &composite_state,
                 &composite_state.external_surfaces,
                 projection,
@@ -3485,13 +3486,6 @@ impl Renderer {
         let _gm = self.gpu_profiler.start_marker("framebuffer");
         let _timer = self.gpu_profiler.start_timer(GPU_TAG_COMPOSITE);
 
-        // We are only interested in tiles backed with actual cached pixels so we don't
-        // count clear tiles here.
-        let num_tiles = composite_state.tiles
-            .iter()
-            .filter(|tile| tile.kind != TileKind::Clear).count();
-        self.profile.set(profiler::PICTURE_TILES, num_tiles);
-
         let window_is_opaque = match self.compositor_config.layer_compositor() {
             Some(ref compositor) => {
                 let props = compositor.get_window_properties();
@@ -3503,8 +3497,20 @@ impl Renderer {
         let mut input_layers: Vec<CompositorInputLayer> = Vec::new();
         let mut swapchain_layers = Vec::new();
         let cap = composite_state.tiles.len();
+        let mut occlusion = occlusion::FrontToBackBuilder::with_capacity(cap, cap);
+        let mut clear_tiles = Vec::new();
+
+        // We are only interested in tiles backed with actual cached pixels so we don't
+        // count clear tiles here.
+        let num_tiles = composite_state.tiles
+            .iter()
+            .filter(|tile| tile.kind != TileKind::Clear).count();
+        self.profile.set(profiler::PICTURE_TILES, num_tiles);
 
         for (idx, tile) in composite_state.tiles.iter().enumerate() {
+            // Clear tiles overwrite whatever is under them, so they are treated as opaque.
+            let is_opaque = tile.kind != TileKind::Alpha;
+
             let device_tile_box = composite_state.get_device_rect(
                 &tile.local_rect,
                 tile.transform_index
@@ -3530,6 +3536,30 @@ impl Renderer {
                 continue;
             }
 
+            // For normal tiles, add to occlusion tracker. For clear tiles, add directly
+            // to the swapchain tile list
+            match tile.kind {
+                TileKind::Opaque | TileKind::Alpha => {
+                    // Store (index of tile, index of layer) so we can segment them below
+                    occlusion.add(&rect, is_opaque, idx);
+                }
+                TileKind::Clear => {
+                    // Clear tiles are specific to how we render the window buttons on
+                    // Windows 8. They clobber what's under them so they can be treated as opaque,
+                    // but require a different blend state so they will be rendered after the opaque
+                    // tiles and before transparent ones.
+                    clear_tiles.push(occlusion::Item { rectangle: rect, key: idx });
+                }
+            }
+        }
+
+        assert_eq!(swapchain_layers.len(), input_layers.len());
+
+        for item in occlusion.opaque_items().iter().chain(occlusion.alpha_items().iter().rev()) {
+            let tile = &composite_state.tiles[item.key];
+            // Clear tiles overwrite whatever is under them, so they are treated as opaque.
+            let is_opaque = tile.kind != TileKind::Alpha;
+
             // Determine if the tile is an external surface or content
             let usage = match tile.surface {
                 CompositeTileSurface::Texture { .. } |
@@ -3618,7 +3648,6 @@ impl Renderer {
                         );
 
                         let clip_rect = tile.device_clip_rect.to_i32();
-                        let is_opaque = tile.kind != TileKind::Alpha;
 
                         (rect.min.to_i32(), clip_rect, is_opaque)
                     }
@@ -3633,39 +3662,27 @@ impl Renderer {
                 });
 
                 swapchain_layers.push(SwapChainLayer {
+                    opaque_items: Vec::new(),
+                    alpha_items: Vec::new(),
                     clear_tiles: Vec::new(),
-                    occlusion: occlusion::FrontToBackBuilder::with_capacity(cap, cap),
                 })
             }
 
-            // For normal tiles, add to occlusion tracker. For clear tiles, add directly
-            // to the swapchain tile list
-            let layer = swapchain_layers.last_mut().unwrap();
+            let item = occlusion::Item {
+                rectangle: item.rectangle,
+                key: item.key,
+            };
 
-            // Clear tiles overwrite whatever is under them, so they are treated as opaque.
-            match tile.kind {
-                TileKind::Opaque => {
-                    // Store (index of tile, index of layer) so we can segment them below
-                    layer.occlusion.add(&rect, true, idx);
-                }
-                TileKind::Alpha => {
-                    // Store (index of tile, index of layer) so we can segment them below
-                    layer.occlusion.add(&rect, false, idx);
-                }
-                TileKind::Clear => {
-                    // Clear tiles are specific to how we render the window buttons on
-                    // Windows 8. They clobber what's under them so they can be treated as opaque,
-                    // but require a different blend state so they will be rendered after the opaque
-                    // tiles and before transparent ones.
-                    layer.clear_tiles.push(occlusion::Item { rectangle: rect, key: idx });
-                }
+            let layer = swapchain_layers.last_mut().unwrap();
+            if is_opaque {
+                layer.opaque_items.push(item);
+            } else {
+                layer.alpha_items.push(item);
             }
         }
 
-        assert_eq!(swapchain_layers.len(), input_layers.len());
-
         // If no tiles were present, and we expect an opaque window,
-        // add an empty layer to force a composite that clears the screen,
+        // ddd an empty layer to force a composite that clears the screen,
         // to match existing semantics.
         if window_is_opaque && input_layers.is_empty() {
             input_layers.push(CompositorInputLayer {
@@ -3676,8 +3693,9 @@ impl Renderer {
             });
 
             swapchain_layers.push(SwapChainLayer {
+                opaque_items: Vec::new(),
+                alpha_items: Vec::new(),
                 clear_tiles: Vec::new(),
-                occlusion: occlusion::FrontToBackBuilder::with_capacity(cap, cap),
             })
         }
 
@@ -3693,8 +3711,9 @@ impl Renderer {
             });
 
             swapchain_layers.push(SwapChainLayer {
+                opaque_items: Vec::new(),
+                alpha_items: Vec::new(),
                 clear_tiles: Vec::new(),
-                occlusion: occlusion::FrontToBackBuilder::with_capacity(cap, cap),
             })
         }
 
@@ -3749,6 +3768,7 @@ impl Renderer {
                 projection,
                 results,
                 partial_present_mode,
+                &occlusion,
                 swapchain_layer,
             );
 
