diff --git a/third_party/dav1d/src/cpu.c b/third_party/dav1d/src/cpu.c
index 6b57fe4a9790..13c883c025e7 100644
--- a/third_party/dav1d/src/cpu.c
+++ b/third_party/dav1d/src/cpu.c
@@ -85,12 +85,23 @@ COLD int dav1d_num_logical_processors(Dav1dContext *const c) {
 #ifdef _WIN32
 #if WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP)
     GROUP_AFFINITY affinity;
-    if (GetThreadGroupAffinity(GetCurrentThread(), &affinity)) {
+  typedef BOOL (WINAPI *GETTHREADGROUPAFFINITY)(HANDLE hThread,PGROUP_AFFINITY GroupAffinity);
+  GETTHREADGROUPAFFINITY pGetThreadGroupAffinity=NULL;
+    HMODULE hKernel=GetModuleHandle(L"kernel32.dll");
+    if (hKernel!=NULL)
+    {
+      pGetThreadGroupAffinity=(GETTHREADGROUPAFFINITY)GetProcAddress(hKernel,"GetThreadGroupAffinity");
+
+      if (pGetThreadGroupAffinity!=NULL && pGetThreadGroupAffinity(GetCurrentThread(), &affinity)) {
         int num_processors = 1;
         while (affinity.Mask &= affinity.Mask - 1)
             num_processors++;
         return num_processors;
+      }
     }
+    SYSTEM_INFO system_info;
+    GetNativeSystemInfo(&system_info);
+    return system_info.dwNumberOfProcessors;
 #else
     SYSTEM_INFO system_info;
     GetNativeSystemInfo(&system_info);
