diff --git a/widget/windows/nsWindow.cpp b/widget/windows/nsWindow.cpp
index eeca575df548..161662b4e4fc 100644
--- a/widget/windows/nsWindow.cpp
+++ b/widget/windows/nsWindow.cpp
@@ -1055,7 +1055,17 @@ nsresult nsWindow::Create(nsIWidget* aParent, const LayoutDeviceIntRect& aRect,
         aInitData.mIsPrivate &&
         !StaticPrefs::browser_privatebrowsing_autostart();
     RefPtr<IPropertyStore> pPropStore;
-    if (!FAILED(SHGetPropertyStoreForWindow(mWnd, IID_IPropertyStore,
+const wchar_t kShellLibraryName[] =  L"shell32.dll";
+
+  typedef HRESULT (WINAPI * SHGetPropertyStoreForWindowPtr)
+                    (HWND hwnd, REFIID riid, void** ppv);
+  SHGetPropertyStoreForWindowPtr funcGetProStore = nullptr;
+
+  HMODULE hDLL = ::LoadLibraryW(kShellLibraryName);
+  funcGetProStore = (SHGetPropertyStoreForWindowPtr)
+    GetProcAddress(hDLL, "SHGetPropertyStoreForWindow");
+
+    if (funcGetProStore && !FAILED(funcGetProStore(mWnd, IID_IPropertyStore,
                                             getter_AddRefs(pPropStore)))) {
       PROPVARIANT pv;
       nsAutoString aumid;
@@ -1085,6 +1095,7 @@ nsresult nsWindow::Create(nsIWidget* aParent, const LayoutDeviceIntRect& aRect,
         PropVariantClear(&pv);
       }
     }
+    ::FreeLibrary(hDLL);
     HICON icon = ::LoadIconW(
         ::GetModuleHandleW(nullptr),
         MAKEINTRESOURCEW(usePrivateAumid ? IDI_PBMODE : IDI_APPICON));
