diff --git a/toolkit/xre/nsWindowsWMain.cpp b/toolkit/xre/nsWindowsWMain.cpp
index 7eb9e1104682..ea14a59b80bb 100644
--- a/toolkit/xre/nsWindowsWMain.cpp
+++ b/toolkit/xre/nsWindowsWMain.cpp
@@ -15,8 +15,10 @@
 
 #include "mozilla/Char16.h"
 #include "nsUTF8Utils.h"
+#include "nsWindowsHelpers.h"
 
 #include <windows.h>
+#include <versionhelpers.h>
 
 #ifdef __MINGW32__
 
@@ -112,6 +114,20 @@ static void FreeAllocStrings(int argc, char** argv) {
 }
 
 int wmain(int argc, WCHAR** argv) {
+  // In Windows 7 32-bit, user32.dll must be mapped to the same virtual
+  // address in all processes.  Otherwise, win32k's user-mode callback causes
+  // crash.  Since we delayload user32.dll, if our code or injected code
+  // reserves the user32 address before user32.dll is loaded, it is loaded
+  // to a new address and we crash.  To mitigate this problem, we explicitly
+  // load user32.dll as early as possible. See bug 1730033 for details.
+  if (!IsWindows8OrGreater()) {
+    SYSTEM_INFO sysInfo;
+    ::GetNativeSystemInfo(&sysInfo);
+    if (sysInfo.wProcessorArchitecture == PROCESSOR_ARCHITECTURE_INTEL) {
+      LoadLibrarySystem32(L"user32.dll");
+    }
+  }
+
   SanitizeEnvironmentVariables();
   SetDllDirectoryW(L"");
 
