diff --git a/widget/windows/nsWindow.cpp b/widget/windows/nsWindow.cpp
index ec5713420ef5..c7072c627aea 100644
--- a/widget/windows/nsWindow.cpp
+++ b/widget/windows/nsWindow.cpp
@@ -346,12 +346,6 @@ static SystemTimeConverter<DWORD>& TimeConverter() {
   return timeConverterSingleton;
 }
 
-static const wchar_t* GetMainWindowClass();
-static const wchar_t* ChooseWindowClass(mozilla::widget::WindowType);
-// This method registers the given window class, and returns the class name.
-static void RegisterWindowClass(const wchar_t* aClassName, UINT aExtraStyle,
-                                LPWSTR aIconID);
-
 // Global event hook for window cloaking. Never deregistered.
 //  - `Nothing` if not yet set.
 //  - `Some(nullptr)` if no attempt should be made to set it.
@@ -1191,28 +1185,40 @@ void nsWindow::Destroy() {
  *
  **************************************************************/
 
-static void RegisterWindowClass(const wchar_t* aClassName, UINT aExtraStyle,
-                                LPWSTR aIconID) {
-  WNDCLASSW wc = {};
+/* static */
+const wchar_t* nsWindow::RegisterWindowClass(const wchar_t* aClassName,
+                                             UINT aExtraStyle, LPWSTR aIconID) {
+  WNDCLASSW wc;
   if (::GetClassInfoW(nsToolkit::mDllInstance, aClassName, &wc)) {
     // already registered
-    return;
+    return aClassName;
   }
 
   wc.style = CS_DBLCLKS | aExtraStyle;
   wc.lpfnWndProc = WinUtils::NonClientDpiScalingDefWindowProcW;
+  wc.cbClsExtra = 0;
+  wc.cbWndExtra = 0;
   wc.hInstance = nsToolkit::mDllInstance;
   wc.hIcon =
       aIconID ? ::LoadIconW(::GetModuleHandleW(nullptr), aIconID) : nullptr;
+  wc.hCursor = nullptr;
+  wc.hbrBackground = nullptr;
+  wc.lpszMenuName = nullptr;
   wc.lpszClassName = aClassName;
 
-  // Failures are ignored as they are handled when ::CreateWindow fails
-  ::RegisterClassW(&wc);
+  if (!::RegisterClassW(&wc)) {
+    // For older versions of Win32 (i.e., not XP), the registration may
+    // fail with aExtraStyle, so we have to re-register without it.
+    wc.style = CS_DBLCLKS;
+    ::RegisterClassW(&wc);
+  }
+  return aClassName;
 }
 
 static LPWSTR const gStockApplicationIcon = MAKEINTRESOURCEW(32512);
 
-static const wchar_t* ChooseWindowClass(WindowType aWindowType) {
+/* static */
+const wchar_t* nsWindow::ChooseWindowClass(WindowType aWindowType) {
   const wchar_t* className = [aWindowType] {
     switch (aWindowType) {
       case WindowType::Invisible:
@@ -1225,8 +1231,7 @@ static const wchar_t* ChooseWindowClass(WindowType aWindowType) {
         return GetMainWindowClass();
     }
   }();
-  RegisterWindowClass(className, 0, gStockApplicationIcon);
-  return className;
+  return RegisterWindowClass(className, 0, gStockApplicationIcon);
 }
 
 /**************************************************************
@@ -8160,7 +8165,7 @@ bool nsWindow::CanTakeFocus() {
   return false;
 }
 
-static const wchar_t* GetMainWindowClass() {
+/* static */ const wchar_t* nsWindow::GetMainWindowClass() {
   static const wchar_t* sMainWindowClass = nullptr;
   if (!sMainWindowClass) {
     nsAutoString className;
diff --git a/widget/windows/nsWindow.h b/widget/windows/nsWindow.h
index 2cbf4804c04b..dc1d84e298b4 100644
--- a/widget/windows/nsWindow.h
+++ b/widget/windows/nsWindow.h
@@ -521,6 +521,7 @@ class nsWindow final : public nsBaseWidget {
   void UpdateDarkModeToolbar();
   void ResetLayout();
   void InvalidateNonClientRegion();
+  static const wchar_t* GetMainWindowClass();
   HWND GetOwnerWnd() const { return ::GetWindow(mWnd, GW_OWNER); }
   bool IsOwnerForegroundWindow() const {
     HWND owner = GetOwnerWnd();
@@ -591,6 +592,11 @@ class nsWindow final : public nsBaseWidget {
   DWORD WindowStyle();
   DWORD WindowExStyle();
 
+  static const wchar_t* ChooseWindowClass(WindowType);
+  // This method registers the given window class, and returns the class name.
+  static const wchar_t* RegisterWindowClass(const wchar_t* aClassName,
+                                            UINT aExtraStyle, LPWSTR aIconID);
+
   /**
    * Popup hooks
    */
