diff --git a/widget/windows/nsWindowGfx.cpp b/widget/windows/nsWindowGfx.cpp
index 030c2dec92c9..4c10acf9fa11 100644
--- a/widget/windows/nsWindowGfx.cpp
+++ b/widget/windows/nsWindowGfx.cpp
@@ -204,12 +204,8 @@ bool nsWindow::OnPaint(uint32_t aNestingLevel) {
   }
   if (!regionToClear.IsEmpty()) {
     auto black = reinterpret_cast<HBRUSH>(::GetStockObject(BLACK_BRUSH));
-    // We could use RegionToHRGN, but at least for simple regions (and possibly
-    // for complex ones too?) FillRect is faster; see bug 1946365 comment 12.
-    for (auto it = regionToClear.RectIter(); !it.Done(); it.Next()) {
-      auto rect = WinUtils::ToWinRect(it.Get());
-      ::FillRect(hDC, &rect, black);
-    }
+    nsAutoRegion hRgnToClear(WinUtils::RegionToHRGN(regionToClear));
+    ::FillRgn(hDC, hRgnToClear, black);
   }
 
   bool didPaint = false;
