diff --git a/mozglue/misc/AwakeTimeStamp.cpp b/mozglue/misc/AwakeTimeStamp.cpp
index f447b4bca131..c92739cf584e 100644
--- a/mozglue/misc/AwakeTimeStamp.cpp
+++ b/mozglue/misc/AwakeTimeStamp.cpp
@@ -84,11 +84,15 @@ AwakeTimeStamp AwakeTimeStamp::NowLoRes() { return Now(); }
 static constexpr uint64_t kHNSperUS = 10;
 
 AwakeTimeStamp AwakeTimeStamp::NowLoRes() {
-  ULONGLONG interrupt_time;
-  DebugOnly<bool> rv = QueryUnbiasedInterruptTime(&interrupt_time);
-  MOZ_ASSERT(rv);
+LARGE_INTEGER ticks;
+LARGE_INTEGER Frequency;
 
-  return AwakeTimeStamp(interrupt_time / kHNSperUS);
+QueryPerformanceFrequency(&Frequency); 
+QueryPerformanceCounter(&ticks);
+ticks.QuadPart *= 1000000;
+ticks.QuadPart /= Frequency.QuadPart;
+
+  return AwakeTimeStamp(ticks.QuadPart);
 }
 
 AwakeTimeStamp AwakeTimeStamp::Now() { return NowLoRes(); }
diff --git a/mozglue/misc/Now.cpp b/mozglue/misc/Now.cpp
index 924b1543591e..6ba2ab5b739c 100644
--- a/mozglue/misc/Now.cpp
+++ b/mozglue/misc/Now.cpp
@@ -48,11 +48,14 @@ Maybe<uint64_t> NowIncludingSuspendMs() {
 static constexpr uint64_t kHNSperMS = 10000;
 
 Maybe<uint64_t> NowExcludingSuspendMs() {
-  ULONGLONG interrupt_time;
-  if (!QueryUnbiasedInterruptTime(&interrupt_time)) {
-    return Nothing();
-  }
-  return Some(interrupt_time / kHNSperMS);
+LARGE_INTEGER ticks;
+LARGE_INTEGER Frequency;
+
+QueryPerformanceFrequency(&Frequency); 
+QueryPerformanceCounter(&ticks);
+ticks.QuadPart *= 1000;
+ticks.QuadPart /= Frequency.QuadPart;
+  return Some(ticks.QuadPart);
 }
 
 Maybe<uint64_t> NowIncludingSuspendMs() {
