diff --git a/servo/components/style/values/specified/box.rs b/servo/components/style/values/specified/box.rs
index 26e248ad9a9d..92fa2b662c28 100644
--- a/servo/components/style/values/specified/box.rs
+++ b/servo/components/style/values/specified/box.rs
@@ -1575,12 +1575,23 @@ pub enum Appearance {
     /// The scroll corner
     #[parse(condition = "ParserContext::chrome_rules_enabled")]
     Scrollcorner,
+    /// A spin control (up/down control for time/date pickers).
+    #[parse(condition = "ParserContext::chrome_rules_enabled")]
+    Spinner,
     /// The up button of a spin control.
     #[parse(condition = "ParserContext::chrome_rules_enabled")]
     SpinnerUpbutton,
     /// The down button of a spin control.
     #[parse(condition = "ParserContext::chrome_rules_enabled")]
     SpinnerDownbutton,
+    /// The textfield of a spin control
+    #[parse(condition = "ParserContext::chrome_rules_enabled")]
+    SpinnerTextfield,
+    /// The tabs scroll arrows (left/right).
+    #[parse(condition = "ParserContext::chrome_rules_enabled")]
+    TabScrollArrowBack,
+    #[parse(condition = "ParserContext::chrome_rules_enabled")]
+    TabScrollArrowForward,
     /// A single toolbar button (with no associated dropdown).
     #[parse(condition = "ParserContext::chrome_rules_enabled")]
     Toolbarbutton,
diff --git a/toolkit/themes/shared/checkbox.css b/toolkit/themes/shared/checkbox.css
index 63dc95ecc855..9f9a135b023b 100644
--- a/toolkit/themes/shared/checkbox.css
+++ b/toolkit/themes/shared/checkbox.css
@@ -62,8 +62,6 @@ checkbox {
   }
 
   @media (-moz-platform: linux) {
-    width: 1.3em;
-    height: 1.3em;
     margin: 2px;
     margin-inline-end: 4px;
   }
diff --git a/widget/nsNativeTheme.cpp b/widget/nsNativeTheme.cpp
index a2fa0d883f01..60e94b35d343 100644
--- a/widget/nsNativeTheme.cpp
+++ b/widget/nsNativeTheme.cpp
@@ -581,8 +581,6 @@ bool nsNativeTheme::IsWidgetAlwaysNonNative(nsIFrame* aFrame,
                                             StyleAppearance aAppearance) {
   return IsWidgetScrollbarPart(aAppearance) ||
          aAppearance == StyleAppearance::FocusOutline ||
-         aAppearance == StyleAppearance::SpinnerUpbutton ||
-         aAppearance == StyleAppearance::SpinnerDownbutton ||
          aAppearance == StyleAppearance::Toolbarbutton ||
          aAppearance == StyleAppearance::ProgressBar ||
          aAppearance == StyleAppearance::Meter ||
diff --git a/widget/windows/nsNativeThemeWin.cpp b/widget/windows/nsNativeThemeWin.cpp
index bdfc5809a433..fb4ef0142a39 100644
--- a/widget/windows/nsNativeThemeWin.cpp
+++ b/widget/windows/nsNativeThemeWin.cpp
@@ -65,7 +65,9 @@ bool nsNativeThemeWin::IsWidgetAlwaysNonNative(nsIFrame* aFrame,
   return Theme::IsWidgetAlwaysNonNative(aFrame, aAppearance) ||
          aAppearance == StyleAppearance::Checkbox ||
          aAppearance == StyleAppearance::Radio ||
-         aAppearance == StyleAppearance::MozMenulistArrowButton;
+         aAppearance == StyleAppearance::MozMenulistArrowButton ||
+         aAppearance == StyleAppearance::SpinnerUpbutton ||
+         aAppearance == StyleAppearance::SpinnerDownbutton;
 }
 
 auto nsNativeThemeWin::IsWidgetNonNative(nsIFrame* aFrame,
