diff --git a/widget/nsXPLookAndFeel.cpp b/widget/nsXPLookAndFeel.cpp
index 53f72ffebfdf..a2bc17ac56ac 100644
--- a/widget/nsXPLookAndFeel.cpp
+++ b/widget/nsXPLookAndFeel.cpp
@@ -751,7 +751,5 @@ Maybe<nscolor> nsXPLookAndFeel::GenericDarkColor(ColorID aID) {
     case ColorID::MozButtonactivetext:
     case ColorID::MozHeaderbartext:
     case ColorID::MozHeaderbarinactivetext:
-    case ColorID::Captiontext:
-    case ColorID::Inactivecaptiontext:  // TODO(emilio): Maybe make
                                         // Inactivecaptiontext Graytext?
       color = kWindowText;
@@ -815,14 +813,8 @@ Maybe<nscolor> nsXPLookAndFeel::GenericDarkColor(ColorID aID) {
       // value of browser.visited_color.dark.
       color = NS_RGB(0xff, 0xad, 0xff);
       break;
-    case ColorID::Activeborder:
-    case ColorID::Inactiveborder:
-      color = NS_RGB(57, 57, 57);
-      break;
     case ColorID::MozHeaderbar:
     case ColorID::MozHeaderbarinactive:
-    case ColorID::Activecaption:
-    case ColorID::Inactivecaption:
       color = NS_RGB(28, 27, 34);
       break;
     case ColorID::MozAutofillBackground:
diff --git a/widget/windows/nsLookAndFeel.cpp b/widget/windows/nsLookAndFeel.cpp
index d955f2a637c8..8671dd387a1a 100644
--- a/widget/windows/nsLookAndFeel.cpp
+++ b/widget/windows/nsLookAndFeel.cpp
@@ -184,67 +184,6 @@ nsresult nsLookAndFeel::NativeGetColor(ColorID aID, ColorScheme aScheme,
     return NS_OK;
   }
 
-  // Titlebar and menu colors are color-scheme aware.
-  switch (aID) {
-    case ColorID::Activecaption:
-      aColor = mTitlebarColors.Get(aScheme, true).mBg;
-      return NS_OK;
-    case ColorID::Captiontext:
-      aColor = mTitlebarColors.Get(aScheme, true).mFg;
-      return NS_OK;
-    case ColorID::Activeborder:
-      aColor = mTitlebarColors.Get(aScheme, true).mBorder;
-      return NS_OK;
-    case ColorID::Inactivecaption:
-      aColor = mTitlebarColors.Get(aScheme, false).mBg;
-      return NS_OK;
-    case ColorID::Inactivecaptiontext:
-      aColor = mTitlebarColors.Get(aScheme, false).mFg;
-      return NS_OK;
-    case ColorID::Inactiveborder:
-      aColor = mTitlebarColors.Get(aScheme, false).mBorder;
-      return NS_OK;
-    case ColorID::MozMenuhover:
-      MOZ_ASSERT(UseNonNativeMenuColors());
-      if (WinUtils::MicaPopupsEnabled()) {
-        aColor = aScheme == ColorScheme::Dark ? NS_RGBA(255, 255, 255, 30)
-                                              : NS_RGBA(0, 0, 0, 30);
-      } else {
-        aColor = aScheme == ColorScheme::Dark ? *GenericDarkColor(aID)
-                                              : NS_RGB(0xe0, 0xe0, 0xe6);
-      }
-      return NS_OK;
-    case ColorID::MozMenuhoverdisabled:
-      if (UseNonNativeMenuColors()) {
-        if (WinUtils::MicaPopupsEnabled()) {
-          aColor = aScheme == ColorScheme::Dark ? NS_RGBA(255, 255, 255, 10)
-                                                : NS_RGBA(0, 0, 0, 10);
-        } else {
-          aColor = aScheme == ColorScheme::Dark ? *GenericDarkColor(aID)
-                                                : NS_RGB(0xf0, 0xf0, 0xf3);
-        }
-      } else {
-        aColor = NS_TRANSPARENT;
-      }
-      return NS_OK;
-    case ColorID::Menu: {
-      if (UseNonNativeMenuColors()) {
-        if (WinUtils::MicaPopupsEnabled()) {
-          aColor = aScheme == ColorScheme::Dark ? NS_RGBA(0, 0, 0, 153)
-                                                : NS_RGBA(255, 255, 255, 153);
-        } else {
-          aColor = aScheme == ColorScheme::Dark ? *GenericDarkColor(aID)
-                                                : NS_RGB(0xf9, 0xf9, 0xfb);
-        }
-      } else {
-        aColor = GetColorForSysColorIndex(COLOR_MENU);
-      }
-      return NS_OK;
-    }
-    default:
-      break;
-  }
-
   if (aScheme == ColorScheme::Dark) {
     if (auto color = GenericDarkColor(aID)) {
       aColor = *color;
@@ -297,6 +273,61 @@ nsresult nsLookAndFeel::NativeGetColor(ColorID aID, ColorScheme aScheme,
       }
       idx = COLOR_HIGHLIGHTTEXT;
       break;
+    case ColorID::Activecaption:
+      aColor = mTitlebarColors.Get(aScheme, true).mBg;
+      return NS_OK;
+    case ColorID::Captiontext:
+      aColor = mTitlebarColors.Get(aScheme, true).mFg;
+      return NS_OK;
+    case ColorID::Activeborder:
+      aColor = mTitlebarColors.Get(aScheme, true).mBorder;
+      return NS_OK;
+    case ColorID::Inactivecaption:
+      aColor = mTitlebarColors.Get(aScheme, false).mBg;
+      return NS_OK;
+    case ColorID::Inactivecaptiontext:
+      aColor = mTitlebarColors.Get(aScheme, false).mFg;
+      return NS_OK;
+    case ColorID::Inactiveborder:
+      aColor = mTitlebarColors.Get(aScheme, false).mBorder;
+      return NS_OK;
+    case ColorID::MozMenuhover:
+      MOZ_ASSERT(UseNonNativeMenuColors());
+      if (WinUtils::MicaPopupsEnabled()) {
+        aColor = aScheme == ColorScheme::Dark ? NS_RGBA(255, 255, 255, 15)
+                                              : NS_RGBA(0, 0, 0, 15);
+      } else {
+        aColor = aScheme == ColorScheme::Dark ? *GenericDarkColor(aID)
+                                              : NS_RGB(0xe0, 0xe0, 0xe6);
+      }
+      return NS_OK;
+    case ColorID::MozMenuhoverdisabled:
+      if (UseNonNativeMenuColors()) {
+        if (WinUtils::MicaPopupsEnabled()) {
+          aColor = aScheme == ColorScheme::Dark ? NS_RGBA(255, 255, 255, 10)
+                                                : NS_RGBA(0, 0, 0, 10);
+        } else {
+          aColor = aScheme == ColorScheme::Dark ? *GenericDarkColor(aID)
+                                                : NS_RGB(0xf0, 0xf0, 0xf3);
+        }
+      } else {
+        aColor = NS_TRANSPARENT;
+      }
+      return NS_OK;
+    case ColorID::Menu: {
+      if (UseNonNativeMenuColors()) {
+        if (WinUtils::MicaPopupsEnabled()) {
+          aColor = aScheme == ColorScheme::Dark ? NS_RGBA(0, 0, 0, 153)
+                                                : NS_RGBA(255, 255, 255, 153);
+        } else {
+          aColor = aScheme == ColorScheme::Dark ? *GenericDarkColor(aID)
+                                                : NS_RGB(0xf9, 0xf9, 0xfb);
+        }
+      } else {
+        aColor = GetColorForSysColorIndex(COLOR_MENU);
+      }
+      return NS_OK;
+    }
     case ColorID::Infobackground:
       idx = COLOR_INFOBK;
       break;
@@ -788,26 +782,11 @@ auto nsLookAndFeel::ComputeTitlebarColors() -> TitlebarColors {
                            GetColorForSysColorIndex(COLOR_INACTIVECAPTIONTEXT),
                            GetColorForSysColorIndex(COLOR_INACTIVEBORDER)};
 
-  if (!mHighContrastOn) {
-    // Use our non-native colors.
-    result.mActiveLight = {
-        GetStandinForNativeColor(ColorID::Activecaption, ColorScheme::Light),
-        GetStandinForNativeColor(ColorID::Captiontext, ColorScheme::Light),
-        GetStandinForNativeColor(ColorID::Activeborder, ColorScheme::Light)};
-    result.mInactiveLight = {
-        GetStandinForNativeColor(ColorID::Inactivecaption, ColorScheme::Light),
-        GetStandinForNativeColor(ColorID::Inactivecaptiontext,
-                                 ColorScheme::Light),
-        GetStandinForNativeColor(ColorID::Inactiveborder, ColorScheme::Light)};
-  }
-
-  // Our dark colors are always non-native.
-  result.mActiveDark = {*GenericDarkColor(ColorID::Activecaption),
-                        *GenericDarkColor(ColorID::Captiontext),
-                        *GenericDarkColor(ColorID::Activeborder)};
-  result.mInactiveDark = {*GenericDarkColor(ColorID::Inactivecaption),
-                          *GenericDarkColor(ColorID::Inactivecaptiontext),
-                          *GenericDarkColor(ColorID::Inactiveborder)};
+  // Foreground and background taken from Windows Mica material theme colors.
+  result.mActiveDark = {NS_RGB(0x2e, 0x2e, 0x2e), NS_RGB(0xff, 0xff, 0xff),
+                        NS_RGB(57, 57, 57)};
+  result.mInactiveDark = {NS_RGB(0x33, 0x33, 0x33), NS_RGB(0xff, 0xff, 0xff),
+                          NS_RGB(57, 57, 57)};
 
   // TODO(bug 1825241): Somehow get notified when this changes? Hopefully the
   // sys color notification is enough.
