diff --git a/browser/app/profile/firefox.js b/browser/app/profile/firefox.js
index 37180f621f27..f42ed7e48429 100644
--- a/browser/app/profile/firefox.js
+++ b/browser/app/profile/firefox.js
@@ -1471,7 +1471,7 @@ pref("browser.bookmarks.editDialog.maxRecentFolders", 7);
   // See - security/sandbox/win/src/sandboxbroker/sandboxBroker.cpp
   // SetSecurityLevelForContentProcess() for what the different settings mean.
   #if defined(EARLY_BETA_OR_EARLIER)
-    pref("security.sandbox.content.level", 8);
+    pref("security.sandbox.content.level", 7);
   #else
     pref("security.sandbox.content.level", 7);
   #endif
