diff --git a/gfx/layers/ipc/CompositorBridgeChild.cpp b/gfx/layers/ipc/CompositorBridgeChild.cpp
index 9b8553a77fdb..a5a1cf482cd7 100644
--- a/gfx/layers/ipc/CompositorBridgeChild.cpp
+++ b/gfx/layers/ipc/CompositorBridgeChild.cpp
@@ -364,7 +364,7 @@ bool CompositorBridgeChild::SendFlushRenderingAsync(
   if (!mCanSend) {
     return false;
   }
-  return PCompositorBridgeChild::SendFlushRenderingAsync(aReasons);
+  return PCompositorBridgeChild::SendFlushRendering(aReasons);
 }
 
 void CompositorBridgeChild::SetForceSyncFlushRendering(
