diff --git a/toolkit/crashreporter/client/app/src/ui/windows/mod.rs b/toolkit/crashreporter/client/app/src/ui/windows/mod.rs
index 924b1986f619..6e77f8eec3e1 100644
--- a/toolkit/crashreporter/client/app/src/ui/windows/mod.rs
+++ b/toolkit/crashreporter/client/app/src/ui/windows/mod.rs
@@ -338,21 +338,6 @@ impl CustomWindowClass for AppWindow {
                 for child in children {
                     renderer.render_window(child);
                 }
-
-                // Set the window size to the initial (DPI-aware) min width/height. We don't have
-                // the DPI information until we create the window, so we can't set the initial
-                // window size correctly at that time.
-                unsafe {
-                    win::SetWindowPos(
-                        hwnd,
-                        win::HWND_TOP,
-                        0,
-                        0,
-                        me.renderer.min_width().try_into().unwrap(),
-                        me.renderer.min_height().try_into().unwrap(),
-                        win::SWP_NOZORDER | win::SWP_NOACTIVATE | win::SWP_NOMOVE,
-                    )
-                };
             }
             win::WM_CLOSE => {
                 if model.modal {
