diff --git a/toolkit/crashreporter/mozwer-rust/lib.rs b/toolkit/crashreporter/mozwer-rust/lib.rs
index 61fc89446229..90b4f269e255 100644
--- a/toolkit/crashreporter/mozwer-rust/lib.rs
+++ b/toolkit/crashreporter/mozwer-rust/lib.rs
@@ -40,7 +40,7 @@ use windows_sys::Win32::{
     System::Memory::{
         VirtualAllocEx, VirtualFreeEx, MEM_COMMIT, MEM_RELEASE, MEM_RESERVE, PAGE_READWRITE,
     },
-    System::ProcessStatus::K32GetModuleFileNameExW,
+    System::ProcessStatus::GetModuleFileNameExW,
     System::SystemInformation::{
         VerSetConditionMask, VerifyVersionInfoW, OSVERSIONINFOEXW, VER_MAJORVERSION,
         VER_MINORVERSION, VER_SERVICEPACKMAJOR, VER_SERVICEPACKMINOR,
@@ -539,7 +539,7 @@ impl ApplicationInformation {
     fn get_application_path(process: HANDLE) -> Result<PathBuf> {
         let mut path: [u16; MAX_PATH as usize + 1] = [0; MAX_PATH as usize + 1];
         unsafe {
-            let res = K32GetModuleFileNameExW(
+            let res = GetModuleFileNameExW(
                 process,
                 0,
                 (&mut path).as_mut_ptr(),
