diff --git a/modules/libpref/init/all.js b/modules/libpref/init/all.js
index 57c5cf787bb7..82d473e8461b 100644
--- a/modules/libpref/init/all.js
+++ b/modules/libpref/init/all.js
@@ -2262,9 +2262,22 @@ pref("font.size.monospace.x-math", 13);
   pref("gfx.font_rendering.cleartype_params.pixel_structure", -1);
   pref("gfx.font_rendering.cleartype_params.rendering_mode", -1);
 
+#if 0
   // We no longer force "GDI Classic" mode on any fonts by default.
   pref("gfx.font_rendering.cleartype_params.force_gdi_classic_for_families", "");
   pref("gfx.font_rendering.cleartype_params.force_gdi_classic_max_size", 0);
+#else
+  // A comma-separated list of font family names. Fonts in these families will
+  // be forced to use "GDI Classic" ClearType mode, provided the value
+  // of gfx.font_rendering.cleartype_params.rendering_mode is -1
+  // (i.e. a specific rendering_mode has not been explicitly set).
+  // Currently we apply this setting to the sans-serif Microsoft "core Web fonts".
+  pref("gfx.font_rendering.cleartype_params.force_gdi_classic_for_families",
+       "Arial,Consolas,Courier New,Microsoft Sans Serif,Segoe UI,Tahoma,Trebuchet MS,Verdana");
+  // The maximum size at which we will force GDI classic mode using
+  // force_gdi_classic_for_families.
+  pref("gfx.font_rendering.cleartype_params.force_gdi_classic_max_size", 15);
+#endif
 
   // Switch the keyboard layout per window
   pref("intl.keyboard.per_window_layout", false);
