diff --git a/build/moz.configure/windows.configure b/build/moz.configure/windows.configure
index 098cf06a667ee..0c456a5802284 100644
--- a/build/moz.configure/windows.configure
+++ b/build/moz.configure/windows.configure
@@ -595,7 +595,7 @@ with only_when(depends(c_compiler)(lambda c: c.type == "clang-cl")):
     set_define("HAVE_IO_H", True)
 
     def win32_subsystem_version():
-        return "10.0"
+        return "5.01"
 
     set_config(
         "WIN32_CONSOLE_EXE_LDFLAGS",
