diff --git a/hal/windows/WindowsHeterogeneousCpuInfo.cpp b/hal/windows/WindowsHeterogeneousCpuInfo.cpp
index 52b305008d2e2..a0b277ddace12 100644
--- a/hal/windows/WindowsHeterogeneousCpuInfo.cpp
+++ b/hal/windows/WindowsHeterogeneousCpuInfo.cpp
@@ -15,7 +15,14 @@ namespace mozilla::hal_impl {
 
 static mozilla::Maybe<hal::HeterogeneousCpuInfo> CreateHeterogeneousCpuInfo() {
   ULONG returnedLength;
-  GetSystemCpuSetInformation(NULL, 0, &returnedLength, NULL, 0);
+  auto pGetSystemCpuSetInformation =
+      reinterpret_cast<decltype(&GetSystemCpuSetInformation)>(
+          GetProcAddress(GetModuleHandleW(L"kernel32.dll"),
+                           "GetSystemCpuSetInformation"));
+  if (!pGetSystemCpuSetInformation) {
+    return Nothing();
+  }
+  pGetSystemCpuSetInformation(NULL, 0, &returnedLength, NULL, 0);
 
   if (!returnedLength) {
     return Nothing();
@@ -25,7 +32,7 @@ static mozilla::Maybe<hal::HeterogeneousCpuInfo> CreateHeterogeneousCpuInfo() {
 
   cpuSets.SetLength(returnedLength);
 
-  if (!GetSystemCpuSetInformation(
+  if (!pGetSystemCpuSetInformation(
           reinterpret_cast<SYSTEM_CPU_SET_INFORMATION*>(cpuSets.Elements()),
           returnedLength, &returnedLength, NULL, 0)) {
     return Nothing();
