diff --git a/widget/windows/nsWindow.cpp b/widget/windows/nsWindow.cpp
index 96db1530eda0..14206cf057cd 100644
--- a/widget/windows/nsWindow.cpp
+++ b/widget/windows/nsWindow.cpp
@@ -1221,19 +1221,18 @@ static LPWSTR const gStockApplicationIcon = MAKEINTRESOURCEW(32512);
 
 /* static */
 const wchar_t* nsWindow::ChooseWindowClass(WindowType aWindowType) {
-  const wchar_t* className = [aWindowType] {
-    switch (aWindowType) {
-      case WindowType::Invisible:
-        return kClassNameHidden;
-      case WindowType::Dialog:
-        return kClassNameDialog;
-      case WindowType::Popup:
-        return kClassNameDropShadow;
-      default:
-        return GetMainWindowClass();
-    }
-  }();
-  return RegisterWindowClass(className, 0, gStockApplicationIcon);
+  switch (aWindowType) {
+    case WindowType::Invisible:
+      return RegisterWindowClass(kClassNameHidden, 0, gStockApplicationIcon);
+    case WindowType::Dialog:
+      return RegisterWindowClass(kClassNameDialog, 0, nullptr);
+    case WindowType::Popup:
+      return RegisterWindowClass(kClassNameDropShadow, 0,
+                                 gStockApplicationIcon);
+    default:
+      return RegisterWindowClass(GetMainWindowClass(), 0,
+                                 gStockApplicationIcon);
+  }
 }
 
 /**************************************************************
@@ -1333,6 +1332,7 @@ DWORD nsWindow::WindowStyle() {
 
 // Return nsWindow extended styles
 DWORD nsWindow::WindowExStyle() {
+  MOZ_ASSERT_IF(mIsAlert, mWindowType == WindowType::Dialog);
   switch (mWindowType) {
     case WindowType::Child:
       return 0;
@@ -1343,16 +1343,16 @@ DWORD nsWindow::WindowExStyle() {
       }
       return extendedStyle;
     }
+    case WindowType::Dialog: {
+      if (mIsAlert) {
+        return WS_EX_TOOLWINDOW;
+      }
+      return WS_EX_WINDOWEDGE | WS_EX_DLGMODALFRAME;
+    }
-    case WindowType::Dialog:
     case WindowType::TopLevel:
     case WindowType::Invisible:
       break;
   }
-  if (mIsAlert) {
-    MOZ_ASSERT(mWindowType == WindowType::Dialog,
-               "Expect alert windows to have type=dialog");
-    return WS_EX_TOOLWINDOW | WS_EX_TOPMOST;
-  }
   return WS_EX_WINDOWEDGE;
 }
 
