diff --git a/gfx/webrender_bindings/src/bindings.rs b/gfx/webrender_bindings/src/bindings.rs
index d2bdcc7413a3..47c077cb4c6c 100644
--- a/gfx/webrender_bindings/src/bindings.rs
+++ b/gfx/webrender_bindings/src/bindings.rs
@@ -42,7 +42,7 @@ use webrender::{
     MappableCompositor, MappedTileInfo, NativeSurfaceId, NativeSurfaceInfo, NativeTileId, PartialPresentCompositor,
     PipelineInfo, ProfilerHooks, RecordedFrameHandle, RenderBackendHooks, Renderer, RendererStats,
     SWGLCompositeSurfaceInfo, SceneBuilderHooks, ShaderPrecacheFlags, Shaders, SharedShaders, TextureCacheConfig,
-    UploadMethod, WebRenderOptions, WindowVisibility, WindowProperties, ONE_TIME_USAGE_HINT, CompositorInputConfig, CompositorSurfaceUsage,
+    UploadMethod, WebRenderOptions, WindowVisibility, WindowProperties, ONE_TIME_USAGE_HINT, CompositorInputConfig,
 };
 use wr_malloc_size_of::MallocSizeOfOps;
 
@@ -1506,7 +1506,6 @@ struct NativeLayer {
     size: DeviceIntSize,
     is_opaque: bool,
     frames_since_used: usize,
-    usage: CompositorSurfaceUsage,
 }
 
 pub struct WrLayerCompositor {
@@ -1543,8 +1542,7 @@ impl LayerCompositor for WrLayerCompositor {
             let size = request.clip_rect.size();
 
             let existing_index = self.surface_pool.iter().position(|layer| {
-                layer.is_opaque == request.is_opaque &&
-                layer.usage.matches(&request.usage)
+                layer.is_opaque == request.is_opaque
             });
 
             let mut layer = match existing_index {
@@ -1553,9 +1551,6 @@ impl LayerCompositor for WrLayerCompositor {
 
                     layer.frames_since_used = 0;
 
-                    // Copy across (potentially) updated external image id
-                    layer.usage = request.usage;
-
                     layer
                 }
                 None => {
@@ -1563,23 +1558,12 @@ impl LayerCompositor for WrLayerCompositor {
                     self.next_layer_id += 1;
 
                     unsafe {
-                        match request.usage {
-                            CompositorSurfaceUsage::Content => {
-                                wr_compositor_create_swapchain_surface(
-                                    self.compositor,
-                                    id,
-                                    size,
-                                    request.is_opaque,
-                                );
-                            }
-                            CompositorSurfaceUsage::External { .. } => {
-                                wr_compositor_create_external_surface(
-                                    self.compositor,
-                                    id,
-                                    request.is_opaque,
-                                );
-                            }
-                        }
+                        wr_compositor_create_swapchain_surface(
+                            self.compositor,
+                            id,
+                            size,
+                            request.is_opaque,
+                        );
                     }
 
                     NativeLayer {
@@ -1587,33 +1571,19 @@ impl LayerCompositor for WrLayerCompositor {
                         size,
                         is_opaque: request.is_opaque,
                         frames_since_used: 0,
-                        usage: request.usage,
                     }
                 }
             };
 
-            match layer.usage {
-                CompositorSurfaceUsage::Content => {
-                    if layer.size.width != size.width || layer.size.height != size.height {
-                        unsafe {
-                            wr_compositor_resize_swapchain(
-                                self.compositor,
-                                layer.id,
-                                size
-                            );
-                        }
-                        layer.size = size;
-                    }
-                }
-                CompositorSurfaceUsage::External { external_image_id, .. } => {
-                    unsafe {
-                        wr_compositor_attach_external_image(
-                            self.compositor,
-                            layer.id,
-                            external_image_id,
-                        );
-                    }
+            if layer.size.width != size.width || layer.size.height != size.height {
+                unsafe {
+                    wr_compositor_resize_swapchain(
+                        self.compositor,
+                        layer.id,
+                        size
+                    );
                 }
+                layer.size = size;
             }
 
             self.visual_tree.push(layer);
@@ -1651,12 +1621,14 @@ impl LayerCompositor for WrLayerCompositor {
     fn add_surface(
         &mut self,
         index: usize,
-        transform: CompositorSurfaceTransform,
         clip_rect: DeviceIntRect,
         image_rendering: ImageRendering,
     ) {
         let layer = &self.visual_tree[index];
 
+        // TODO(gwc): Need to set correct transform once adding video layers
+        let transform = CompositorSurfaceTransform::identity();
+
         unsafe {
             wr_compositor_add_surface(
                 self.compositor,
diff --git a/gfx/wr/webrender/src/composite.rs b/gfx/wr/webrender/src/composite.rs
index d18297583a5b..560e64911d17 100644
--- a/gfx/wr/webrender/src/composite.rs
+++ b/gfx/wr/webrender/src/composite.rs
@@ -2,7 +2,7 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-use api::{ColorF, ExternalImageId, ImageBufferKind, ImageKey, ImageRendering, YuvFormat, YuvRangedColorSpace};
+use api::{ColorF, YuvRangedColorSpace, YuvFormat, ImageRendering, ExternalImageId, ImageBufferKind};
 use api::units::*;
 use api::ColorDepth;
 use crate::image_source::resolve_image;
@@ -204,8 +204,6 @@ pub struct ExternalSurfaceDescriptor {
     /// If the native surface needs to be updated, this will contain the size
     /// of the native surface as Some(size). If not dirty, this is None.
     pub update_params: Option<DeviceIntSize>,
-    /// If using external compositing, a user key for the client
-    pub external_image_id: Option<ExternalImageId>,
 }
 
 impl ExternalSurfaceDescriptor {
@@ -285,8 +283,6 @@ pub struct ResolvedExternalSurface {
     pub image_buffer_kind: ImageBufferKind,
     // Update information for a native surface if it's dirty
     pub update_params: Option<(NativeSurfaceId, DeviceIntSize)>,
-    /// If using external compositing, a user key for the client
-    pub external_image_id: Option<ExternalImageId>,
 }
 
 /// Public interface specified in `WebRenderOptions` that configures
@@ -465,32 +461,10 @@ pub struct CompositeTileDescriptor {
 // by WR to render content, or is an external swapchain for video
 #[cfg_attr(feature = "capture", derive(Serialize))]
 #[cfg_attr(feature = "replay", derive(Deserialize))]
-#[derive(Debug, Copy, Clone)]
+#[derive(Debug, PartialEq, Copy, Clone)]
 pub enum CompositorSurfaceUsage {
     Content,
-    External {
-        image_key: ImageKey,
-        external_image_id: ExternalImageId,
-        transform_index: CompositorTransformIndex,
-    },
-}
-
-impl CompositorSurfaceUsage {
-    // Returns true if usage is compatible
-    pub fn matches(&self, other: &CompositorSurfaceUsage) -> bool {
-        match (self, other) {
-            // Surfaces used for content are always compatible
-            (CompositorSurfaceUsage::Content, CompositorSurfaceUsage::Content) => true,
-
-            (CompositorSurfaceUsage::Content, CompositorSurfaceUsage::External { .. }) |
-            (CompositorSurfaceUsage::External { .. }, CompositorSurfaceUsage::Content) => false,
-
-            // External surfaces are matched by image-key (which doesn't change per-frame)
-            (CompositorSurfaceUsage::External { image_key: key1, .. }, CompositorSurfaceUsage::External { image_key: key2, .. }) => {
-                key1 == key2
-            }
-        }
-    }
+    External,
 }
 
 /// Describes the properties that identify a surface composition uniquely.
@@ -778,7 +752,7 @@ impl CompositeState {
         // when drawing with the simple (Draw) compositor, and to schedule compositing
         // of any required updates into the surfaces.
         let needs_external_surface_update = match self.compositor_kind {
-            CompositorKind::Draw { .. } | CompositorKind::Layer { .. } => true,
+            CompositorKind::Draw { .. } => true,
             _ => external_surface.update_params.is_some(),
         };
         let external_surface_index = if needs_external_surface_update {
@@ -1061,7 +1035,6 @@ impl CompositeState {
                         },
                     image_buffer_kind,
                     update_params,
-                    external_image_id: external_surface.external_image_id,
                 });
             },
             ExternalSurfaceDependency::Rgb { .. } => {
@@ -1074,7 +1047,6 @@ impl CompositeState {
                     },
                     image_buffer_kind,
                     update_params,
-                    external_image_id: external_surface.external_image_id,
                 });
             },
         }
@@ -1361,7 +1333,6 @@ pub trait Compositor {
 
 // Describes the configuration for an input layer that the compositor
 // implemention should prepare
-#[derive(Debug)]
 pub struct CompositorInputLayer {
     // Device space location of the layer (pre-clip)
     pub offset: DeviceIntPoint,
@@ -1373,7 +1344,6 @@ pub struct CompositorInputLayer {
 
 // Provides the parameters about the frame to the compositor implementation.
 // TODO(gw): Include information about picture cache slices and external surfaces.
-#[derive(Debug)]
 pub struct CompositorInputConfig<'a> {
     pub layers: &'a [CompositorInputLayer],
 }
@@ -1398,7 +1368,6 @@ pub trait LayerCompositor {
     fn add_surface(
         &mut self,
         index: usize,
-        transform: CompositorSurfaceTransform,
         clip_rect: DeviceIntRect,
         image_rendering: ImageRendering,
     );
diff --git a/gfx/wr/webrender/src/picture.rs b/gfx/wr/webrender/src/picture.rs
index 1a89f1d6a806..ab49c678339a 100644
--- a/gfx/wr/webrender/src/picture.rs
+++ b/gfx/wr/webrender/src/picture.rs
@@ -2928,7 +2928,6 @@ impl TileCacheInstance {
             z_id: ZBufferId::invalid(),
             native_surface_id,
             update_params,
-            external_image_id,
         };
 
         // If the surface is opaque, we can draw it an an underlay (which avoids
diff --git a/gfx/wr/webrender/src/renderer/mod.rs b/gfx/wr/webrender/src/renderer/mod.rs
index 55e585c541c1..6a28893ac495 100644
--- a/gfx/wr/webrender/src/renderer/mod.rs
+++ b/gfx/wr/webrender/src/renderer/mod.rs
@@ -3521,36 +3521,8 @@ impl Renderer {
                 CompositeTileSurface::Clear => {
                     CompositorSurfaceUsage::Content
                 }
-                CompositeTileSurface::ExternalSurface { external_surface_index } => {
-                    match self.current_compositor_kind {
-                        CompositorKind::Native { .. } | CompositorKind::Draw { .. } => {
-                            CompositorSurfaceUsage::Content
-                        }
-                        CompositorKind::Layer { .. } => {
-                            let surface = &composite_state.external_surfaces[external_surface_index.0];
-
-                            // TODO(gwc): For now, we only select a hardware overlay swapchain if we
-                            // have an external image, but it may make sense to do for compositor
-                            // surfaces without in future.
-                            match surface.external_image_id {
-                                Some(external_image_id) => {
-                                    let image_key = match surface.color_data {
-                                        ResolvedExternalSurfaceColorData::Rgb { image_dependency, .. } => image_dependency.key,
-                                        ResolvedExternalSurfaceColorData::Yuv { image_dependencies, .. } => image_dependencies[0].key,
-                                    };
-
-                                    CompositorSurfaceUsage::External {
-                                        image_key,
-                                        external_image_id,
-                                        transform_index: tile.transform_index,
-                                    }
-                                }
-                                None => {
-                                    CompositorSurfaceUsage::Content
-                                }
-                            }
-                        }
-                    }
+                CompositeTileSurface::ExternalSurface { .. } => {
+                    CompositorSurfaceUsage::Content
                 }
             };
 
@@ -3560,17 +3532,15 @@ impl Renderer {
                     match (curr_layer.usage, usage) {
                         // Content -> content, composite in to same layer
                         (CompositorSurfaceUsage::Content, CompositorSurfaceUsage::Content) => None,
-                        (CompositorSurfaceUsage::External { .. }, CompositorSurfaceUsage::Content) => Some(usage),
 
                         // Switch of layer type, or video -> video, need new swapchain
-                        (CompositorSurfaceUsage::Content, CompositorSurfaceUsage::External { .. }) |
-                        (CompositorSurfaceUsage::External { .. }, CompositorSurfaceUsage::External { .. }) => {
+                        (CompositorSurfaceUsage::Content, CompositorSurfaceUsage::External) |
+                        (CompositorSurfaceUsage::External, CompositorSurfaceUsage::Content) |
+                        (CompositorSurfaceUsage::External, CompositorSurfaceUsage::External) => {
                             // Only create a new layer if we're using LayerCompositor
                             match self.compositor_config {
                                 CompositorConfig::Draw { .. } | CompositorConfig::Native { .. } => None,
-                                CompositorConfig::Layer { .. } => {
-                                    Some(usage)
-                                }
+                                CompositorConfig::Layer { .. } => Some(usage),
                             }
                         }
                     }
@@ -3586,7 +3556,7 @@ impl Renderer {
                             input_layers.is_empty() && window_is_opaque,
                         )
                     }
-                    CompositorSurfaceUsage::External { .. } => {
+                    CompositorSurfaceUsage::External => {
                         let rect = composite_state.get_device_rect(
                             &tile.local_rect,
                             tile.transform_index
@@ -3650,14 +3620,6 @@ impl Renderer {
         for (layer_index, (layer, swapchain_layer)) in input_layers.iter().zip(swapchain_layers.iter()).enumerate() {
             self.device.reset_state();
 
-            // Skip compositing external images
-            match layer.usage {
-                CompositorSurfaceUsage::Content => {}
-                CompositorSurfaceUsage::External { .. } => {
-                    continue;
-                }
-            }
-
             let clear_color = if layer_index == 0 {
                 self.clear_color
             } else {
@@ -3702,16 +3664,8 @@ impl Renderer {
         // End frame notify for experimental compositor
         if let Some(ref mut compositor) = self.compositor_config.layer_compositor() {
             for (layer_index, layer) in input_layers.iter().enumerate() {
-                // External surfaces need transform applied, but content
-                // surfaces are always at identity
-                let transform = match layer.usage {
-                    CompositorSurfaceUsage::Content => CompositorSurfaceTransform::identity(),
-                    CompositorSurfaceUsage::External { transform_index, .. } => composite_state.get_compositor_transform(transform_index),
-                };
-
                 compositor.add_surface(
                     layer_index,
-                    transform,
                     layer.clip_rect,
                     ImageRendering::Auto,
                 );
diff --git a/gfx/wr/wrench/src/composite.rs b/gfx/wr/wrench/src/composite.rs
index 867eb5a78303..2c71ff798feb 100644
--- a/gfx/wr/wrench/src/composite.rs
+++ b/gfx/wr/wrench/src/composite.rs
@@ -6,7 +6,7 @@ use std::os::raw::c_void;
 use std::ptr;
 
 use glutin::platform::windows::EGLContext;
-use webrender::{CompositorInputConfig, CompositorSurfaceTransform, LayerCompositor};
+use webrender::{LayerCompositor, CompositorInputConfig};
 use winit::platform::windows::WindowExtWindows;
 
 use crate::WindowWrapper;
@@ -248,12 +248,11 @@ impl LayerCompositor for WrCompositor {
     }
 
     fn add_surface(
-        &mut self,
-        index: usize,
-        _transform: CompositorSurfaceTransform,
-        _clip_rect: webrender::api::units::DeviceIntRect,
-        _image_rendering: webrender::api::ImageRendering,
-    ) {
+            &mut self,
+            index: usize,
+            _clip_rect: webrender::api::units::DeviceIntRect,
+            _image_rendering: webrender::api::ImageRendering,
+        ) {
         let layer = &self.layers[index];
 
         unsafe {
