diff --git a/security/sandbox/win/SandboxInitialization.cpp b/security/sandbox/win/SandboxInitialization.cpp
index 641577c4071a..97b5cdc5153a 100644
--- a/security/sandbox/win/SandboxInitialization.cpp
+++ b/security/sandbox/win/SandboxInitialization.cpp
@@ -180,6 +180,8 @@ static sandbox::BrokerServices* InitializeBrokerServices() {
   // will be broken. This has to run before threads and windows are created.
   (void)brokerServices->CreateAlternateDesktop(
       sandbox::Desktop::kAlternateWinstation);
+  (void)brokerServices->CreateAlternateDesktop(
+      sandbox::Desktop::kAlternateDesktop);
 
   // Ensure the relevant mitigations are enforced.
   mozilla::sandboxing::ApplyParentProcessMitigations();
diff --git a/security/sandbox/win/src/sandboxbroker/sandboxBroker.cpp b/security/sandbox/win/src/sandboxbroker/sandboxBroker.cpp
index 429a6dff12a0..fa5c80fdba0d 100644
--- a/security/sandbox/win/src/sandboxbroker/sandboxBroker.cpp
+++ b/security/sandbox/win/src/sandboxbroker/sandboxBroker.cpp
@@ -1588,7 +1588,7 @@ struct WindowsUtilitySandboxProps : public UtilitySandboxProps {
   WindowsUtilitySandboxProps() {
     mJobLevel = sandbox::JobLevel::kInteractive;
     mDelayedTokenLevel = sandbox::USER_RESTRICTED_SAME_ACCESS;
-    mDesktop = sandbox::Desktop::kDefault;
+    mDesktop = sandbox::Desktop::kAlternateDesktop;
     mInitialIntegrityLevel = sandbox::INTEGRITY_LEVEL_MEDIUM;
     mDelayedIntegrityLevel = sandbox::INTEGRITY_LEVEL_MEDIUM;
     mUseWin32kLockdown = false;
