diff --git a/browser/installer/Makefile.in b/browser/installer/Makefile.in
index a291755adbd1..92ed86d87dcd 100644
--- a/browser/installer/Makefile.in
+++ b/browser/installer/Makefile.in
@@ -102,6 +102,13 @@ ifdef MOZ_NOTIFICATION_SERVER
 DEFINES += -DMOZ_NOTIFICATION_SERVER=1
 endif
 
+ifeq (cocoa,$(MOZ_WIDGET_TOOLKIT))
+MOZ_PKG_MAC_DSSTORE=$(topsrcdir)/$(MOZ_BRANDING_DIRECTORY)/dsstore
+MOZ_PKG_MAC_BACKGROUND=$(topsrcdir)/$(MOZ_BRANDING_DIRECTORY)/background.png
+MOZ_PKG_MAC_ICON=$(topsrcdir)/$(MOZ_BRANDING_DIRECTORY)/disk.icns
+MOZ_PKG_MAC_EXTRA=--symlink '/Applications:/ '
+endif
+
 include $(topsrcdir)/toolkit/mozapps/installer/packager.mk
 
 ifeq (Darwin,$(OS_TARGET))
diff --git a/browser/locales/Makefile.in b/browser/locales/Makefile.in
index 2f6a2c9be2e9..d3b76c92e48a 100644
--- a/browser/locales/Makefile.in
+++ b/browser/locales/Makefile.in
@@ -16,6 +16,8 @@ SUBMAKEFILES += \
 
 PWD := $(CURDIR)
 
+# These are defaulted to be compatible with the files the wget-en-US target
+# pulls. You may override them if you provide your own files.
 ZIP_IN ?= $(ABS_DIST)/$(PACKAGE)
 
 ifdef MOZ_DEV_EDITION
@@ -23,6 +25,17 @@ MOZ_LANGPACK_EID=langpack-$(AB_CD)@devedition.mozilla.org
 else
 MOZ_LANGPACK_EID=langpack-$(AB_CD)@firefox.mozilla.org
 endif
+# For Nightly, we know where to get the builds from to do local repacks
+ifdef NIGHTLY_BUILD
+export EN_US_BINARY_URL ?= https://archive.mozilla.org/pub/firefox/nightly/latest-mozilla-central
+endif
+
+ifneq (,$(filter cocoa,$(MOZ_WIDGET_TOOLKIT)))
+MOZ_PKG_MAC_DSSTORE=$(ABS_DIST)/branding/dsstore
+MOZ_PKG_MAC_BACKGROUND=$(ABS_DIST)/branding/background.png
+MOZ_PKG_MAC_ICON=$(ABS_DIST)/branding/disk.icns
+MOZ_PKG_MAC_EXTRA=--symlink '/Applications:/ '
+endif
 
 # Required for l10n.mk - defines a list of app sub dirs that should
 # be included in langpack xpis.
diff --git a/config/makefiles/makeutils.mk b/config/makefiles/makeutils.mk
index 5be39dec7fad..95c7791a4b3f 100644
--- a/config/makefiles/makeutils.mk
+++ b/config/makefiles/makeutils.mk
@@ -116,3 +116,6 @@ endif
 ifdef USE_AUTOTARGETS_MK # mkdir_deps
   include $(topORerr)/config/makefiles/autotargets.mk
 endif
+
+## copy(src, dst): recursive copy
+copy_dir = (cd $(1)/. && $(TAR) $(TAR_CREATE_FLAGS) - .) | (cd $(2)/. && tar -xf -)
diff --git a/python/mozbuild/mozbuild/artifact_commands.py b/python/mozbuild/mozbuild/artifact_commands.py
index 1951a41546c7..83da36b3940b 100644
--- a/python/mozbuild/mozbuild/artifact_commands.py
+++ b/python/mozbuild/mozbuild/artifact_commands.py
@@ -89,7 +89,6 @@ def _make_artifacts(
     download_symbols=False,
     download_maven_zip=False,
     no_process=False,
-    unfiltered_project_package=False,
 ):
     state_dir = command_context._mach_context.state_dir
     cache_dir = os.path.join(state_dir, "package-frontend")
@@ -130,7 +129,6 @@ def _make_artifacts(
         download_symbols=download_symbols,
         download_maven_zip=download_maven_zip,
         no_process=no_process,
-        unfiltered_project_package=unfiltered_project_package,
         mozbuild=command_context,
     )
     return artifacts
@@ -165,11 +163,6 @@ def _make_artifacts(
     action="store_true",
     help="Don't process (unpack) artifact packages, just download them.",
 )
-@CommandArgument(
-    "--unfiltered-project-package",
-    action="store_true",
-    help="Minimally process (only) main project package artifact, unpacking it to the given `--distdir`.",
-)
 @CommandArgument(
     "--maven-zip", action="store_true", help="Download Maven zip (Android-only)."
 )
@@ -184,7 +177,6 @@ def artifact_install(
     symbols=False,
     distdir=None,
     no_process=False,
-    unfiltered_project_package=False,
     maven_zip=False,
 ):
     command_context._set_log_level(verbose)
@@ -197,7 +189,6 @@ def artifact_install(
         download_symbols=symbols,
         download_maven_zip=maven_zip,
         no_process=no_process,
-        unfiltered_project_package=unfiltered_project_package,
     )
 
     return artifacts.install_from(source, distdir or command_context.distdir)
diff --git a/python/mozbuild/mozbuild/artifacts.py b/python/mozbuild/mozbuild/artifacts.py
index 9fb1d6afb335..60b483bf1d8c 100644
--- a/python/mozbuild/mozbuild/artifacts.py
+++ b/python/mozbuild/mozbuild/artifacts.py
@@ -57,7 +57,7 @@ import pylru
 import requests
 from mach.util import UserError
 from mozpack import executables
-from mozpack.files import FileFinder, JarFinder, TarFinder
+from mozpack.files import JarFinder, TarFinder
 from mozpack.mozjar import JarReader, JarWriter
 from mozpack.packager.unpack import UnpackFinder
 from taskcluster.exceptions import TaskclusterRestFailure
@@ -84,14 +84,10 @@ MAX_CACHED_TASKS = 400  # Number of pushheads to cache Task Cluster task data fo
 # copying from DMG files is very slow, we extract the desired binaries to a
 # separate archive for fast re-installation.
 PROCESSED_SUFFIX = ".processed.jar"
-UNFILTERED_PROJECT_PACKAGE_PROCESSED_SUFFIX = (
-    ".unfiltered_project_package.processed.jar"
-)
 
 
-class GeckoJobConfiguration:
+class ArtifactJob:
     trust_domain = "gecko"
-    product = "firefox"
     default_candidate_trees = [
         "releases/mozilla-release",
     ]
@@ -110,33 +106,6 @@ class GeckoJobConfiguration:
     ]
     try_tree = "try"
 
-
-class AndroidJobConfiguration(GeckoJobConfiguration):
-    product = "mobile"
-
-
-class ThunderbirdJobConfiguration:
-    trust_domain = "comm"
-    product = "thunderbird"
-    default_candidate_trees = [
-        "releases/comm-release",
-    ]
-    nightly_candidate_trees = [
-        "comm-central",
-    ]
-    beta_candidate_trees = [
-        "releases/comm-beta",
-    ]
-    # The list below list should be updated when we have new ESRs.
-    esr_candidate_trees = [
-        "releases/comm-esr115",
-        "releases/comm-esr128",
-        "releases/comm-esr140",
-    ]
-    try_tree = "try-comm-central"
-
-
-class ArtifactJob:
     def _get_orig_basename(self, filename):
         """Extract the original basename from a filename, removing hash prefixes if present."""
         orig_basename = os.path.basename(filename)
@@ -203,13 +173,9 @@ class ArtifactJob:
         download_tests=True,
         download_symbols=False,
         download_maven_zip=False,
-        override_job_configuration=None,
         substs=None,
         mozbuild=None,
     ):
-        if override_job_configuration is not None:
-            self.job_configuration = override_job_configuration
-
         self._package_re = re.compile(self.package_re)
         self._tests_re = None
         if download_tests:
@@ -496,14 +462,6 @@ class ArtifactJob:
         else:
             raise RuntimeError("Unsupported archive type for %s" % filename)
 
-    @property
-    def product(self):
-        return self.job_configuration.product
-
-    @property
-    def trust_domain(self):
-        return self.job_configuration.trust_domain
-
     @property
     def candidate_trees(self):
         if not self._candidate_trees:
@@ -515,18 +473,18 @@ class ArtifactJob:
         version_display = buildconfig.substs.get("MOZ_APP_VERSION_DISPLAY")
 
         if "esr" in version_display or "esr" in source_repo:
-            return self.job_configuration.esr_candidate_trees
+            return self.esr_candidate_trees
         elif re.search(r"a\d+$", version_display):
-            return self.job_configuration.nightly_candidate_trees
+            return self.nightly_candidate_trees
         elif re.search(r"b\d+$", version_display):
-            return self.job_configuration.beta_candidate_trees
+            return self.beta_candidate_trees
 
-        return self.job_configuration.default_candidate_trees
+        return self.default_candidate_trees
 
 
 class AndroidArtifactJob(ArtifactJob):
     package_re = r"public/build/geckoview_example\.apk$"
-    job_configuration = AndroidJobConfiguration
+    product = "mobile"
 
     package_artifact_patterns = {"**/*.so"}
 
@@ -591,7 +549,7 @@ class AndroidArtifactJob(ArtifactJob):
 
 class LinuxArtifactJob(ArtifactJob):
     package_re = r"public/build/target\.tar\.(bz2|xz)$"
-    job_configuration = GeckoJobConfiguration
+    product = "firefox"
 
     _package_artifact_patterns = {
         "{product}/crashhelper",
@@ -684,7 +642,7 @@ class ResignJarWriter(JarWriter):
 
 class MacArtifactJob(ArtifactJob):
     package_re = r"public/build/target\.dmg$"
-    job_configuration = GeckoJobConfiguration
+    product = "firefox"
 
     # These get copied into dist/bin without the path, so "root/a/b/c" -> "dist/bin/c".
     _paths_no_keep_path = (
@@ -806,7 +764,7 @@ class MacArtifactJob(ArtifactJob):
 
 class WinArtifactJob(ArtifactJob):
     package_re = r"public/build/target\.(zip|tar\.gz)$"
-    job_configuration = GeckoJobConfiguration
+    product = "firefox"
 
     _package_artifact_patterns = {
         "{product}/dependentlibs.list",
@@ -870,78 +828,37 @@ class WinArtifactJob(ArtifactJob):
             )
 
 
-class UnfilteredProjectPackageArtifactJob(ArtifactJob):
-    """An `ArtifactJob` that processes only the main project package and is
-    unfiltered, i.e., does not change the internal structure of the main
-    package.  For use in repackaging, where the artifact build mode VCS and
-    Taskcluster integration is convenient but the whole package is needed (and
-    DMGs are slow to work with locally).
-
-    Desktop-only at this time.
-
-    """
-
-    # Can't yet handle `AndroidArtifactJob` uniformly, since the `product` is "mobile".
-    package_re = "|".join(
-        [
-            f"({cls.package_re})"
-            for cls in (LinuxArtifactJob, MacArtifactJob, WinArtifactJob)
-        ]
-    )
-    job_configuration = GeckoJobConfiguration
+class ThunderbirdMixin:
+    trust_domain = "comm"
+    product = "thunderbird"
+    try_tree = "try-comm-central"
+    default_candidate_trees = [
+        "releases/comm-release",
+    ]
+    nightly_candidate_trees = [
+        "comm-central",
+    ]
+    beta_candidate_trees = [
+        "releases/comm-beta",
+    ]
+    # The list below list should be updated when we have new ESRs.
+    esr_candidate_trees = [
+        "releases/comm-esr115",
+        "releases/comm-esr128",
+        "releases/comm-esr140",
+    ]
 
-    @property
-    def _extra_archives(self):
-        return {}
 
-    def process_package_artifact(self, filename, processed_filename):
-        tempdir = tempfile.mkdtemp()
-        try:
-            self.log(
-                logging.DEBUG,
-                "artifact",
-                {"tempdir": tempdir},
-                "Unpacking into {tempdir}",
-            )
-            mozinstall.install(filename, tempdir)
+class LinuxThunderbirdArtifactJob(ThunderbirdMixin, LinuxArtifactJob):
+    pass
 
-            # Avoid mismatches between local packages (Nightly.app) and CI artifacts
-            # (Firefox Nightly.app).
-            if filename.endswith(".dmg"):
-                bundle_dirs = glob.glob(mozpath.join(tempdir, "*.app"))
-            else:
-                bundle_dirs = glob.glob(
-                    mozpath.join(tempdir, self._substs["MOZ_APP_NAME"])
-                )
 
-            if len(bundle_dirs) != 1:
-                raise ValueError(f"Expected one source bundle, found: {bundle_dirs}")
-            (source,) = bundle_dirs
+class MacThunderbirdArtifactJob(ThunderbirdMixin, MacArtifactJob):
+    pass
 
-            with self.get_writer(file=processed_filename, compress_level=5) as writer:
-                finder = FileFinder(source)
-                for p, f in finder.find("*"):
-                    q = p
-                    if filename.endswith(".dmg"):
-                        q = mozpath.join(self._substs["MOZ_MACBUNDLE_NAME"], q)
-                    self.log(
-                        logging.DEBUG,
-                        "artifact",
-                        {"path": q},
-                        "Adding {path} to unfiltered project package archive",
-                    )
-                    writer.add(q.encode("utf-8"), f.open(), mode=f.mode)
 
-        finally:
-            try:
-                shutil.rmtree(tempdir)
-            except OSError:
-                self.log(
-                    logging.WARN,
-                    "artifact",
-                    {"tempdir": tempdir},
-                    "Unable to delete {tempdir}",
-                )
+class WinThunderbirdArtifactJob(ThunderbirdMixin, WinArtifactJob):
+    pass
 
 
 def startswithwhich(s, prefixes):
@@ -950,7 +867,7 @@ def startswithwhich(s, prefixes):
             return prefix
 
 
-JOB_DETAILS = {
+MOZ_JOB_DETAILS = {
     j: {
         "android": AndroidArtifactJob,
         "linux": LinuxArtifactJob,
@@ -959,6 +876,15 @@ JOB_DETAILS = {
     }[startswithwhich(j, ("android", "linux", "macosx", "win"))]
     for j in JOB_CHOICES
 }
+COMM_JOB_DETAILS = {
+    j: {
+        "android": None,
+        "linux": LinuxThunderbirdArtifactJob,
+        "macosx": MacThunderbirdArtifactJob,
+        "win": WinThunderbirdArtifactJob,
+    }[startswithwhich(j, ("android", "linux", "macosx", "win"))]
+    for j in JOB_CHOICES
+}
 
 
 def cachedmethod(cachefunc):
@@ -1126,7 +1052,7 @@ class TaskCache(CacheManager):
         )
 
     @cachedmethod(operator.attrgetter("_cache"))
-    def artifacts(self, tree, job, job_configuration, rev):
+    def artifacts(self, tree, job, artifact_job_class, rev):
         # Grab the second part of the repo name, which is generally how things
         # are indexed. Eg: 'integration/autoland' is indexed as
         # 'autoland'
@@ -1135,7 +1061,7 @@ class TaskCache(CacheManager):
         if job.endswith("-opt"):
             tree += ".shippable"
 
-        namespace = f"{job_configuration.trust_domain}.v2.{tree}.revision.{rev}.{job_configuration.product}.{job}"
+        namespace = f"{artifact_job_class.trust_domain}.v2.{tree}.revision.{rev}.{artifact_job_class.product}.{job}"
         self.log(
             logging.DEBUG,
             "artifact",
@@ -1171,17 +1097,11 @@ class Artifacts:
         download_symbols=False,
         download_maven_zip=False,
         no_process=False,
-        unfiltered_project_package=False,
         mozbuild=None,
     ):
         if (hg and git) or (not hg and not git):
             raise ValueError("Must provide path to exactly one of hg and git")
 
-        if no_process and unfiltered_project_package:
-            raise ValueError(
-                "Must provide only one of no_process and unfiltered_project_package"
-            )
-
         self._substs = substs
         self._defines = defines
         self._tree = tree
@@ -1193,40 +1113,23 @@ class Artifacts:
         self._cache_dir = cache_dir
         self._skip_cache = skip_cache
         self._no_process = no_process
-        self._unfiltered_project_package = unfiltered_project_package
 
-        job_configuration = (
-            ThunderbirdJobConfiguration
-            if substs.get("MOZ_BUILD_APP") == "comm/mail"
-            else None
-        )
-        if not self._unfiltered_project_package:
-            try:
-                cls = JOB_DETAILS[self._job]
-                self._artifact_job = cls(
-                    log=self._log,
-                    download_tests=download_tests,
-                    download_symbols=download_symbols,
-                    download_maven_zip=download_maven_zip,
-                    override_job_configuration=job_configuration,
-                    substs=self._substs,
-                    mozbuild=mozbuild,
-                )
-            except KeyError:
-                self.log(
-                    logging.INFO, "artifact", {"job": self._job}, "Unknown job {job}"
-                )
-                raise KeyError("Unknown job")
-        else:
-            self._artifact_job = UnfilteredProjectPackageArtifactJob(
+        app = self._substs.get("MOZ_BUILD_APP")
+        job_details = COMM_JOB_DETAILS if app == "comm/mail" else MOZ_JOB_DETAILS
+
+        try:
+            cls = job_details[self._job]
+            self._artifact_job = cls(
                 log=self._log,
-                download_tests=False,
-                download_symbols=False,
-                download_maven_zip=False,
-                override_job_configuration=job_configuration,
+                download_tests=download_tests,
+                download_symbols=download_symbols,
+                download_maven_zip=download_maven_zip,
                 substs=self._substs,
                 mozbuild=mozbuild,
             )
+        except KeyError:
+            self.log(logging.INFO, "artifact", {"job": self._job}, "Unknown job {job}")
+            raise KeyError("Unknown job")
 
         self._task_cache = TaskCache(
             self._cache_dir, log=self._log, skip_cache=self._skip_cache
@@ -1529,7 +1432,7 @@ https://firefox-source-docs.mozilla.org/contributing/vcs/mercurial_bundles.html
     def find_pushhead_artifacts(self, task_cache, job, tree, pushhead):
         try:
             taskId, artifacts = task_cache.artifacts(
-                tree, job, self._artifact_job.job_configuration, pushhead
+                tree, job, self._artifact_job.__class__, pushhead
             )
         except ValueError:
             return None
@@ -1579,8 +1482,6 @@ https://firefox-source-docs.mozilla.org/contributing/vcs/mercurial_bundles.html
 
         # Do we need to post-process?
         processed_filename = filename + PROCESSED_SUFFIX
-        if self._unfiltered_project_package:
-            processed_filename = filename + UNFILTERED_PROJECT_PACKAGE_PROCESSED_SUFFIX
 
         if self._skip_cache and os.path.exists(processed_filename):
             self.log(
diff --git a/python/mozbuild/mozbuild/test/test_artifacts.py b/python/mozbuild/mozbuild/test/test_artifacts.py
index d76f626c9e77..6897136d4a8b 100644
--- a/python/mozbuild/mozbuild/test/test_artifacts.py
+++ b/python/mozbuild/mozbuild/test/test_artifacts.py
@@ -7,16 +7,11 @@ from unittest import TestCase
 import buildconfig
 import mozunit
 
-from mozbuild.artifacts import (
-    ArtifactJob,
-    GeckoJobConfiguration,
-    ThunderbirdJobConfiguration,
-)
+from mozbuild.artifacts import ArtifactJob, ThunderbirdMixin
 
 
 class FakeArtifactJob(ArtifactJob):
     package_re = r""
-    job_configuration = GeckoJobConfiguration
 
 
 class TestArtifactJob(TestCase):
@@ -29,43 +24,38 @@ class TestArtifactJob(TestCase):
 
     def test_candidate_trees_with_empty_file(self):
         self._assert_candidate_trees(
-            version_display="",
-            expected_trees=GeckoJobConfiguration.default_candidate_trees,
+            version_display="", expected_trees=ArtifactJob.default_candidate_trees
         )
 
     def test_candidate_trees_with_beta_version(self):
         self._assert_candidate_trees(
-            version_display="92.1b2",
-            expected_trees=GeckoJobConfiguration.beta_candidate_trees,
+            version_display="92.1b2", expected_trees=ArtifactJob.beta_candidate_trees
         )
 
     def test_candidate_trees_with_esr_version(self):
         self._assert_candidate_trees(
-            version_display="91.3.0esr",
-            expected_trees=GeckoJobConfiguration.esr_candidate_trees,
+            version_display="91.3.0esr", expected_trees=ArtifactJob.esr_candidate_trees
         )
 
     def test_candidate_trees_with_nightly_version(self):
         self._assert_candidate_trees(
-            version_display="95.0a1",
-            expected_trees=GeckoJobConfiguration.nightly_candidate_trees,
+            version_display="95.0a1", expected_trees=ArtifactJob.nightly_candidate_trees
         )
 
     def test_candidate_trees_with_release_version(self):
         self._assert_candidate_trees(
-            version_display="93.0.1",
-            expected_trees=GeckoJobConfiguration.default_candidate_trees,
+            version_display="93.0.1", expected_trees=ArtifactJob.default_candidate_trees
         )
 
     def test_candidate_trees_with_newline_before_version(self):
         self._assert_candidate_trees(
             version_display="\n\n91.3.0esr",
-            expected_trees=GeckoJobConfiguration.esr_candidate_trees,
+            expected_trees=ArtifactJob.esr_candidate_trees,
         )
 
     def test_property_is_cached(self):
         job = FakeArtifactJob()
-        expected_trees = GeckoJobConfiguration.esr_candidate_trees
+        expected_trees = ArtifactJob.esr_candidate_trees
 
         buildconfig.substs["MOZ_APP_VERSION_DISPLAY"] = "91.3.0.esr"
         self.assertEqual(job.candidate_trees, expected_trees)
@@ -75,19 +65,21 @@ class TestArtifactJob(TestCase):
         self.assertEqual(job.candidate_trees, expected_trees)
 
 
+class FakeThunderbirdJob(ThunderbirdMixin, FakeArtifactJob):
+    pass
+
+
 class TestThunderbirdMixin(TestCase):
     def _assert_candidate_trees(self, version_display, source_repo, expected_trees):
         with mock.patch.object(
             buildconfig,
             "substs",
             {
                 "MOZ_APP_VERSION_DISPLAY": version_display,
                 "MOZ_SOURCE_REPO": source_repo,
             },
         ):
-            job = FakeArtifactJob(
-                override_job_configuration=ThunderbirdJobConfiguration
-            )
+            job = FakeThunderbirdJob()
             self.assertGreater(len(job.candidate_trees), 0)
             self.assertEqual(job.candidate_trees, expected_trees)
 
@@ -88,33 +82,33 @@ class TestThunderbirdMixin(TestCase):
         self._assert_candidate_trees(
             version_display="92.1b2",
             source_repo="https://hg.mozilla.org/releases/comm-beta",
-            expected_trees=ThunderbirdJobConfiguration.beta_candidate_trees,
+            expected_trees=ThunderbirdMixin.beta_candidate_trees,
         )
 
     def test_candidate_trees_with_esr_version(self):
         self._assert_candidate_trees(
             version_display="91.3.0esr",
             source_repo="https://hg.mozilla.org/releases/comm-esr91",
-            expected_trees=ThunderbirdJobConfiguration.esr_candidate_trees,
+            expected_trees=ThunderbirdMixin.esr_candidate_trees,
         )
 
     def test_candidate_trees_with_nightly_version(self):
         self._assert_candidate_trees(
             version_display="95.0a1",
             source_repo="https://hg.mozilla.org/comm-central",
-            expected_trees=ThunderbirdJobConfiguration.nightly_candidate_trees,
+            expected_trees=ThunderbirdMixin.nightly_candidate_trees,
         )
 
     def test_candidate_trees_with_release_version(self):
         self._assert_candidate_trees(
             version_display="93.0.1",
             source_repo="https://hg.mozilla.org/releases/comm-release",
-            expected_trees=ThunderbirdJobConfiguration.default_candidate_trees,
+            expected_trees=ThunderbirdMixin.default_candidate_trees,
         )
 
     def test_property_is_cached(self):
-        job = FakeArtifactJob(override_job_configuration=ThunderbirdJobConfiguration)
-        expected_trees = ThunderbirdJobConfiguration.esr_candidate_trees
+        job = FakeThunderbirdJob()
+        expected_trees = ThunderbirdMixin.esr_candidate_trees
 
         buildconfig.substs["MOZ_APP_VERSION_DISPLAY"] = "91.3.0.esr"
         self.assertEqual(job.candidate_trees, expected_trees)
diff --git a/taskcluster/kinds/l10n/kind.yml b/taskcluster/kinds/l10n/kind.yml
index 537fe85fda49..d5a8a4ead7e6 100644
--- a/taskcluster/kinds/l10n/kind.yml
+++ b/taskcluster/kinds/l10n/kind.yml
@@ -89,12 +89,18 @@ tasks:
                     win64: windows2012-64/opt
         env:
             by-build-platform:
-                linux.*:
-                    MOZ_ARTIFACT_TASK: {task-reference: '<build-signing>'}
+                linux.*:    # linux64 and 32 get same treatment here
+                    EN_US_PACKAGE_NAME: target.tar.xz
+                    EN_US_BINARY_URL:
+                        artifact-reference: <build-signing/{artifact_prefix}>
                 macosx64.*:
-                    MOZ_ARTIFACT_TASK: {task-reference: '<repackage>'}
+                    EN_US_PACKAGE_NAME: target.dmg
+                    EN_US_BINARY_URL:
+                        artifact-reference: <repackage/{artifact_prefix}>
                 win.*:
-                    MOZ_ARTIFACT_TASK: {task-reference: '<build-signing>'}
+                    EN_US_PACKAGE_NAME: target.zip
+                    EN_US_BINARY_URL:
+                        artifact-reference: <build-signing/{artifact_prefix}>
         mozharness:
             config:
                 by-build-platform:
diff --git a/taskcluster/kinds/shippable-l10n/kind.yml b/taskcluster/kinds/shippable-l10n/kind.yml
index 2525369fa22a..ac7615d33037 100644
--- a/taskcluster/kinds/shippable-l10n/kind.yml
+++ b/taskcluster/kinds/shippable-l10n/kind.yml
@@ -112,12 +112,18 @@ tasks:
                     win64-aarch64-devedition: windows2012-aarch64-devedition/opt
         env:
             by-build-platform:
-                linux.*:
-                    MOZ_ARTIFACT_TASK: {task-reference: '<build-signing>'}
+                linux.*:    # linux64 and 32 get same treatment here
+                    EN_US_PACKAGE_NAME: target.tar.xz
+                    EN_US_BINARY_URL:
+                        artifact-reference: <build-signing/{artifact_prefix}>
                 macosx64.*:
-                    MOZ_ARTIFACT_TASK: {task-reference: '<repackage>'}
+                    EN_US_PACKAGE_NAME: target.dmg
+                    EN_US_BINARY_URL:
+                        artifact-reference: <repackage/{artifact_prefix}>
                 win.*:
-                    MOZ_ARTIFACT_TASK: {task-reference: '<build-signing>'}
+                    EN_US_PACKAGE_NAME: target.zip
+                    EN_US_BINARY_URL:
+                        artifact-reference: <build-signing/{artifact_prefix}>
         mozharness:
             config:
                 by-build-platform:
diff --git a/testing/mozharness/configs/single_locale/tc_linux_common.py b/testing/mozharness/configs/single_locale/tc_linux_common.py
index bb6ede48db52..4ae752847378 100644
--- a/testing/mozharness/configs/single_locale/tc_linux_common.py
+++ b/testing/mozharness/configs/single_locale/tc_linux_common.py
@@ -8,6 +8,7 @@ config = {
     "bootstrap_env": {
         "NO_MERCURIAL_SETUP_CHECK": "1",
         "MOZ_OBJDIR": "%(abs_obj_dir)s",
+        "EN_US_BINARY_URL": os.environ["EN_US_BINARY_URL"],
         "DIST": "%(abs_obj_dir)s",
         "L10NBASEDIR": "../../l10n",
         "TOOLTOOL_CACHE": os.environ.get("TOOLTOOL_CACHE"),
diff --git a/testing/mozharness/configs/single_locale/tc_macosx64.py b/testing/mozharness/configs/single_locale/tc_macosx64.py
index bb6ede48db52..4ae752847378 100644
--- a/testing/mozharness/configs/single_locale/tc_macosx64.py
+++ b/testing/mozharness/configs/single_locale/tc_macosx64.py
@@ -8,6 +8,7 @@ config = {
     "bootstrap_env": {
         "NO_MERCURIAL_SETUP_CHECK": "1",
         "MOZ_OBJDIR": "%(abs_obj_dir)s",
+        "EN_US_BINARY_URL": os.environ["EN_US_BINARY_URL"],
         "DIST": "%(abs_obj_dir)s",
         "L10NBASEDIR": "../../l10n",
         "TOOLTOOL_CACHE": os.environ.get("TOOLTOOL_CACHE"),
diff --git a/testing/mozharness/configs/single_locale/tc_win32.py b/testing/mozharness/configs/single_locale/tc_win32.py
index bb6ede48db52..34bcde89bc0b 100644
--- a/testing/mozharness/configs/single_locale/tc_win32.py
+++ b/testing/mozharness/configs/single_locale/tc_win32.py
@@ -8,8 +8,10 @@ config = {
     "bootstrap_env": {
         "NO_MERCURIAL_SETUP_CHECK": "1",
         "MOZ_OBJDIR": "%(abs_obj_dir)s",
+        "EN_US_BINARY_URL": os.environ["EN_US_BINARY_URL"],
         "DIST": "%(abs_obj_dir)s",
         "L10NBASEDIR": "../../l10n",
         "TOOLTOOL_CACHE": os.environ.get("TOOLTOOL_CACHE"),
+        "EN_US_PACKAGE_NAME": "target.zip",
     },
 }
diff --git a/testing/mozharness/configs/single_locale/tc_win64.py b/testing/mozharness/configs/single_locale/tc_win64.py
index bb6ede48db52..34bcde89bc0b 100644
--- a/testing/mozharness/configs/single_locale/tc_win64.py
+++ b/testing/mozharness/configs/single_locale/tc_win64.py
@@ -8,8 +8,10 @@ config = {
     "bootstrap_env": {
         "NO_MERCURIAL_SETUP_CHECK": "1",
         "MOZ_OBJDIR": "%(abs_obj_dir)s",
+        "EN_US_BINARY_URL": os.environ["EN_US_BINARY_URL"],
         "DIST": "%(abs_obj_dir)s",
         "L10NBASEDIR": "../../l10n",
         "TOOLTOOL_CACHE": os.environ.get("TOOLTOOL_CACHE"),
+        "EN_US_PACKAGE_NAME": "target.zip",
     },
 }
diff --git a/testing/mozharness/scripts/desktop_l10n.py b/testing/mozharness/scripts/desktop_l10n.py
index 70d56e978960..13f4508828f6 100755
--- a/testing/mozharness/scripts/desktop_l10n.py
+++ b/testing/mozharness/scripts/desktop_l10n.py
@@ -58,6 +58,17 @@ class DesktopSingleLocale(LocalesMixin, AutomationMixin, VCSMixin, BaseScript):
                 " revision separated by colon, en-GB:default.",
             },
         ],
+        [
+            [
+                "--en-us-installer-url",
+            ],
+            {
+                "action": "store",
+                "dest": "en_us_installer_url",
+                "type": "string",
+                "help": "Specify the url of the en-us binary",
+            },
+        ],
     ]
 
     def __init__(self, require_config_file=True):
@@ -190,9 +201,11 @@ class DesktopSingleLocale(LocalesMixin, AutomationMixin, VCSMixin, BaseScript):
         self._copy_mozconfig()
         self._mach_configure()
         self._make_export()
+        self.make_wget_en_US()
+        self.make_unpack_en_US()
 
     def _make_export(self):
-        """this step creates nsinstall, needed on Windows hosts
+        """this step creates nsinstall, needed my make_wget_en_US() on Windows hosts
         and creates buildid.h, used by NSIS installer for Windows UBR telemetry
         """
         dirs = self.query_abs_dirs()
@@ -267,6 +280,21 @@ class DesktopSingleLocale(LocalesMixin, AutomationMixin, VCSMixin, BaseScript):
             ignore_errors=ignore_errors,
         )
 
+    def make_unpack_en_US(self):
+        """wrapper for make unpack"""
+        config = self.config
+        dirs = self.query_abs_dirs()
+        env = self.query_bootstrap_env()
+        cwd = os.path.join(dirs["abs_obj_dir"], config["locales_dir"])
+        return self._make(target=["unpack"], cwd=cwd, env=env)
+
+    def make_wget_en_US(self):
+        """wrapper for make wget-en-US"""
+        env = self.query_bootstrap_env()
+        dirs = self.query_abs_dirs()
+        cwd = dirs["abs_locales_dir"]
+        return self._make(target=["wget-en-US"], cwd=cwd, env=env)
+
     def make_upload(self, locale):
         """wrapper for make upload command"""
         env = self.query_l10n_env()
diff --git a/toolkit/locales/l10n.mk b/toolkit/locales/l10n.mk
index 4e27d1815a0e..e9c7235b3b95 100644
--- a/toolkit/locales/l10n.mk
+++ b/toolkit/locales/l10n.mk
@@ -37,6 +37,11 @@ LPROJ_ROOT := $(subst -,_,$(AB_CD))
 endif
 endif
 
+# These are defaulted to be compatible with the files the wget-en-US target
+# pulls. You may override them if you provide your own files.
+ZIP_IN ?= $(ABS_DIST)/$(PACKAGE)
+WIN32_INSTALLER_IN ?= $(ABS_DIST)/$(PKG_INST_PATH)$(PKG_INST_BASENAME).exe
+
 # Allows overriding the final destination of the repackaged file
 ZIP_OUT ?= $(ABS_DIST)/$(PACKAGE)
 
@@ -71,19 +76,41 @@ STAGEDIST = $(ABS_DIST)/l10n-stage/$(MOZ_PKG_DIR)/$(_APPNAME)/Contents/Resources
 else
 STAGEDIST = $(ABS_DIST)/l10n-stage/$(MOZ_PKG_DIR)
 endif
+UNPACKED_INSTALLER = $(ABS_DIST)/unpacked-installer
 
 include $(MOZILLA_DIR)/toolkit/mozapps/installer/packager.mk
 
-unpack:
-	$(RM) -r -f '$(ABS_DIST)/l10n-stage'
-	$(PYTHON3) $(topsrcdir)/mach --log-no-times artifact install --unfiltered-project-package --distdir '$(ABS_DIST)/l10n-stage/$(MOZ_PKG_DIR)' --verbose
+PACKAGE_BASE_DIR = $(ABS_DIST)/l10n-stage
+
+$(UNPACKED_INSTALLER): AB_CD:=en-US
+$(UNPACKED_INSTALLER): UNPACKAGE=$(call ESCAPE_WILDCARD,$(ZIP_IN))
+$(UNPACKED_INSTALLER): $(call ESCAPE_WILDCARD,$(ZIP_IN))
+# only mac needs to remove the parent of STAGEDIST...
+ifeq (cocoa,$(MOZ_WIDGET_TOOLKIT))
+	$(RM) -r -v $(UNPACKED_INSTALLER)
+else
+# ... and windows doesn't like removing STAGEDIST itself, remove all children
+# 	find $(UNPACKED_INSTALLER) -maxdepth 1 -print0 | xargs -0 $(RM) -r
+endif
+	$(NSINSTALL) -D $(UNPACKED_INSTALLER)
+	$(call INNER_UNMAKE_PACKAGE,$(UNPACKED_INSTALLER))
+
+unpack: $(UNPACKED_INSTALLER)
+ifeq ($(OS_ARCH), WINNT)
+	$(RM) -r -f $(ABS_DIST)/l10n-stage
+	$(NSINSTALL) -D $(ABS_DIST)/l10n-stage
+	$(call copy_dir, $(UNPACKED_INSTALLER), $(ABS_DIST)/l10n-stage)
+else
+	rsync -rav --delete $(UNPACKED_INSTALLER)/ $(ABS_DIST)/l10n-stage
+endif
 
 # The path to the object dir for the mozilla-central build system,
 # may be overridden if necessary.
 MOZDEPTH ?= $(DEPTH)
 
+repackage-zip: UNPACKAGE='$(ZIP_IN)'
 repackage-zip:
-	$(PYTHON3) $(MOZILLA_DIR)/toolkit/mozapps/installer/l10n-repack.py '$(STAGEDIST)' $(ABS_DIST)/xpi-stage/locale-$(AB_CD) \
+	$(PYTHON3) $(MOZILLA_DIR)/toolkit/mozapps/installer/l10n-repack.py '$(STAGEDIST)' $(DIST)/xpi-stage/locale-$(AB_CD) \
 		$(MOZ_PKG_EXTRAL10N) \
 		$(if $(MOZ_PACKAGER_MINIFY),--minify) \
 		$(if $(filter omni,$(MOZ_PACKAGER_FORMAT)),$(if $(NON_OMNIJAR_FILES),--non-resource $(NON_OMNIJAR_FILES)))
@@ -111,7 +138,7 @@ endif
 	if test -f '$(DIST)/l10n-stage/$(PACKAGE).asc'; then mv -f '$(DIST)/l10n-stage/$(PACKAGE).asc' '$(ZIP_OUT).asc'; fi
 
 repackage-zip-%: unpack
-	@$(MAKE) repackage-zip AB_CD=$*
+	@$(MAKE) repackage-zip AB_CD=$* ZIP_IN='$(ZIP_IN)'
 
 # Dealing with app sub dirs: If DIST_SUBDIRS is defined it contains a
 # listing of app sub-dirs we should include in langpack xpis. If not,
@@ -153,3 +180,24 @@ package-langpack-%:
 	$(NSINSTALL) -D $(DIST)/$(PKG_LANGPACK_PATH)
 	$(call py_action,langpack_manifest $(AB_CD),--locales $(AB_CD) --app-version $(MOZ_APP_VERSION) --max-app-ver $(MOZ_APP_MAXVERSION) --app-name '$(MOZ_APP_DISPLAYNAME)' --l10n-basedir '$(L10NBASEDIR)' --metadata $(LANGPACK_METADATA) --langpack-eid '$(MOZ_LANGPACK_EID)' --input $(DIST)/xpi-stage/locale-$(AB_CD))
 	$(call py_action,zip $(PKG_LANGPACK_BASENAME).xpi,-C $(DIST)/xpi-stage/locale-$(AB_CD) -x **/*.manifest -x **/*.js -x **/*.ini $(LANGPACK_FILE) $(PKG_ZIP_DIRS) manifest.json)
+
+# This variable is to allow the wget-en-US target to know which ftp server to download from
+ifndef EN_US_BINARY_URL
+EN_US_BINARY_URL = $(error You must set EN_US_BINARY_URL)
+endif
+
+# Allow the overriding of PACKAGE format so we can get an EN_US build with a different
+# PACKAGE format than we are creating l10n packages with.
+EN_US_PACKAGE_NAME ?= $(PACKAGE)
+
+# This make target allows us to wget the latest en-US binary from a specified website
+# The make installers-% target needs the en-US binary in dist/
+# and for the windows repackages we need the .installer.exe in dist/sea
+wget-en-US:
+ifndef WGET
+	$(error Wget not installed)
+endif
+	$(NSINSTALL) -D $(ABS_DIST)/$(PKG_PATH)
+	(cd $(ABS_DIST)/$(PKG_PATH) && \
+        $(WGET) --no-cache -nv --no-iri -N -O $(PACKAGE) '$(EN_US_BINARY_URL)/$(EN_US_PACKAGE_NAME)')
+	@echo 'Downloaded $(EN_US_BINARY_URL)/$(EN_US_PACKAGE_NAME) to $(ABS_DIST)/$(PKG_PATH)/$(PACKAGE)'
diff --git a/toolkit/mozapps/installer/upload-files.mk b/toolkit/mozapps/installer/upload-files.mk
index c13145da1383..eb7932527d8b 100644
--- a/toolkit/mozapps/installer/upload-files.mk
+++ b/toolkit/mozapps/installer/upload-files.mk
@@ -35,6 +35,7 @@ _RESPATH = $(_APPNAME)/Contents/Resources
 endif
 endif
 
+PACKAGE_BASE_DIR = $(ABS_DIST)
 PACKAGE       = $(PKG_PATH)$(PKG_BASENAME)$(PKG_SUFFIX)
 
 # JavaScript Shell packaging
@@ -98,21 +99,25 @@ endif
 TAR_CREATE_FLAGS := --exclude=.mkdir.done $(TAR_CREATE_FLAGS)
 CREATE_FINAL_TAR = $(TAR) -c --owner=0 --group=0 --numeric-owner \
   --mode=go-w --exclude=.mkdir.done -f
+UNPACK_TAR       = tar -xf-
 
 ifeq ($(MOZ_PKG_FORMAT),TAR)
   PKG_SUFFIX	= .tar
   INNER_MAKE_PACKAGE 	= cd $(1) && $(CREATE_FINAL_TAR) - $(MOZ_PKG_DIR) > $(PACKAGE)
+  INNER_UNMAKE_PACKAGE	= cd $(1) && $(UNPACK_TAR) < $(UNPACKAGE)
 endif
 
 ifeq ($(MOZ_PKG_FORMAT),TGZ)
   PKG_SUFFIX	= .tar.gz
   INNER_MAKE_PACKAGE 	= cd $(1) && $(CREATE_FINAL_TAR) - $(MOZ_PKG_DIR) | gzip -vf9 > $(PACKAGE)
+  INNER_UNMAKE_PACKAGE	= cd $(1) && gunzip -c $(UNPACKAGE) | $(UNPACK_TAR)
 endif
 
 ifeq ($(MOZ_PKG_FORMAT),XZ)
   PKG_SUFFIX = .tar.xz
   # For non-shippable builds, we would rather finish the build sooner than have optimal compression.
   INNER_MAKE_PACKAGE 	= cd $(1) && $(CREATE_FINAL_TAR) - $(MOZ_PKG_DIR) | xz --compress --stdout $(if $(MOZ_PROFILE_USE),-9 --extreme) > $(PACKAGE)
+  INNER_UNMAKE_PACKAGE	= cd $(1) && xz --decompress --stdout $(UNPACKAGE) | $(UNPACK_TAR)
 endif
 
 ifeq ($(MOZ_PKG_FORMAT),BZ2)
@@ -121,15 +126,18 @@ ifeq ($(MOZ_PKG_FORMAT),BZ2)
   else
     INNER_MAKE_PACKAGE 	= cd $(1) && $(CREATE_FINAL_TAR) - $(MOZ_PKG_DIR) | bzip2 -vf > $(PACKAGE)
   endif
+  INNER_UNMAKE_PACKAGE	= cd $(1) && bunzip2 -c $(UNPACKAGE) | $(UNPACK_TAR)
 endif
 
 ifeq ($(MOZ_PKG_FORMAT),ZIP)
   PKG_SUFFIX	= .zip
   INNER_MAKE_PACKAGE = $(call py_action,zip,'$(PACKAGE)' '$(MOZ_PKG_DIR)' -x '**/.mkdir.done',$(1))
+  INNER_UNMAKE_PACKAGE = $(call py_action,make_unzip,$(UNPACKAGE),$(1))
 endif
 
 ifeq ($(MOZ_PKG_FORMAT),APK)
 INNER_MAKE_PACKAGE = true
+INNER_UNMAKE_PACKAGE = true
 endif
 
 ifeq ($(MOZ_PKG_FORMAT),DMG)
@@ -210,9 +221,6 @@ ifeq ($(MOZ_PKG_FORMAT),DMG)
 
   _ABS_MOZSRCDIR = $(shell cd $(MOZILLA_DIR) && pwd)
   PKG_DMG_SOURCE = $(MOZ_PKG_DIR)
-  MOZ_PKG_MAC_DSSTORE=$(topsrcdir)/$(MOZ_BRANDING_DIRECTORY)/dsstore
-  MOZ_PKG_MAC_BACKGROUND=$(topsrcdir)/$(MOZ_BRANDING_DIRECTORY)/background.png
-  MOZ_PKG_MAC_ICON=$(topsrcdir)/$(MOZ_BRANDING_DIRECTORY)/disk.icns
   INNER_MAKE_PACKAGE = \
     $(call py_action,make_dmg, \
         $(if $(MOZ_PKG_MAC_DSSTORE),--dsstore '$(MOZ_PKG_MAC_DSSTORE)') \
@@ -221,6 +229,13 @@ ifeq ($(MOZ_PKG_FORMAT),DMG)
         --volume-name '$(MOZ_APP_DISPLAYNAME)' \
         '$(PKG_DMG_SOURCE)' '$(PACKAGE)', \
         $(1))
+  INNER_UNMAKE_PACKAGE = \
+    $(call py_action,unpack_dmg, \
+        $(if $(MOZ_PKG_MAC_DSSTORE),--dsstore '$(MOZ_PKG_MAC_DSSTORE)') \
+        $(if $(MOZ_PKG_MAC_BACKGROUND),--background '$(MOZ_PKG_MAC_BACKGROUND)') \
+        $(if $(MOZ_PKG_MAC_ICON),--icon '$(MOZ_PKG_MAC_ICON)') \
+        $(UNPACKAGE) $(MOZ_PKG_DIR), \
+        $(1))
 endif
 
 MAKE_PACKAGE = $(INNER_MAKE_PACKAGE)
