diff --git a/browser/components/shell/nsWindowsShellService.cpp b/browser/components/shell/nsWindowsShellService.cpp
index cdb99b7f31bd..c0297a2f7048 100644
--- a/browser/components/shell/nsWindowsShellService.cpp
+++ b/browser/components/shell/nsWindowsShellService.cpp
@@ -332,6 +332,11 @@ nsresult nsWindowsShellService::LaunchControlPanelDefaultsSelectionUI() {
 }
 
 nsresult nsWindowsShellService::LaunchControlPanelDefaultPrograms() {
+  // This Default Programs feature is Win7+ only.
+  if (!IsWin7OrLater()) {
+    return NS_ERROR_FAILURE;
+  }
+
   return ::LaunchControlPanelDefaultPrograms() ? NS_OK : NS_ERROR_FAILURE;
 }
 
diff --git a/xpcom/io/nsLocalFileWin.cpp b/xpcom/io/nsLocalFileWin.cpp
index 2aa6cf4fbdc4..227ecbf804a8 100644
--- a/xpcom/io/nsLocalFileWin.cpp
+++ b/xpcom/io/nsLocalFileWin.cpp
@@ -11,6 +11,7 @@
 #include "mozilla/UniquePtrExtensions.h"
 #include "mozilla/Utf8.h"
 #include "mozilla/WinHeaderOnlyUtils.h"
+#include "mozilla/WindowsVersion.h"
 
 #include "nsCOMPtr.h"
 
@@ -1825,12 +1826,14 @@ nsresult nsLocalFile::CopySingleFile(nsIFile* aSourceFile, nsIFile* aDestParent,
     // without COPY_FILE_NO_BUFFERING takes < 1ms. So we only use
     // COPY_FILE_NO_BUFFERING when we have a remote drive.
     DWORD dwCopyFlags = COPY_FILE_ALLOW_DECRYPTED_DESTINATION;
+  if (IsVistaOrLater()) {
     bool path1Remote, path2Remote;
     if (!IsRemoteFilePath(filePath.get(), path1Remote) ||
         !IsRemoteFilePath(destPath.get(), path2Remote) || path1Remote ||
         path2Remote) {
       dwCopyFlags |= COPY_FILE_NO_BUFFERING;
     }
+  }
 
     copyOK = ::CopyFileExW(filePath.get(), destPath.get(), nullptr, nullptr,
                            nullptr, dwCopyFlags);
