diff --git a/browser/app/profile/firefox.js b/browser/app/profile/firefox.js
index 9167f292cb04..aaea8bad75fe 100644
--- a/browser/app/profile/firefox.js
+++ b/browser/app/profile/firefox.js
@@ -1510,7 +1510,11 @@ pref("browser.bookmarks.editDialog.maxRecentFolders", 7);
   // On windows these levels are:
   // See - security/sandbox/win/src/sandboxbroker/sandboxBroker.cpp
   // SetSecurityLevelForContentProcess() for what the different settings mean.
-  pref("security.sandbox.content.level", 8);
+  #if defined(EARLY_BETA_OR_EARLIER)
+    pref("security.sandbox.content.level", 8);
+  #else
+    pref("security.sandbox.content.level", 7);
+  #endif
 
   // Pref controlling if messages relevant to sandbox violations are logged.
   pref("security.sandbox.logging.enabled", false);
