diff --git a/third_party/llama.cpp/ggml/src/ggml-cpu/ggml-cpu-c.c b/third_party/llama.cpp/ggml/src/ggml-cpu/ggml-cpu-c.c
index f6bea3df34a0..c7835cbf532f 100644
--- a/third_party/llama.cpp/ggml/src/ggml-cpu/ggml-cpu-c.c
+++ b/third_party/llama.cpp/ggml/src/ggml-cpu/ggml-cpu-c.c
@@ -2416,7 +2416,7 @@ static bool ggml_thread_apply_priority(int32_t prio) {
         // Newer Windows 11 versions aggresively park (offline) CPU cores and often place
         // all our threads onto the first 4 cores which results in terrible performance with
         // n_threads > 4
-        #if _WIN32_WINNT >= 0x0602
+        #if 0
         THREAD_POWER_THROTTLING_STATE t;
         ZeroMemory(&t, sizeof(t));
         t.Version     = THREAD_POWER_THROTTLING_CURRENT_VERSION;
