diff --git a/modules/libpref/init/StaticPrefList.yaml b/modules/libpref/init/StaticPrefList.yaml
index 454e5d3ec4f4..d4ee70b9c10a 100644
--- a/modules/libpref/init/StaticPrefList.yaml
+++ b/modules/libpref/init/StaticPrefList.yaml
@@ -18676,6 +18676,20 @@
   type: RelaxedAtomicInt32
   value: 0
   mirror: always
+
+
+# Whether we expand the client area into the taskbar hidden edge on maximized
+# windows.
+- name: widget.windows.hidden-taskbar-hack.size
+  type: bool
+  value: false
+  mirror: always
+
+# Whether we paint the bottom taskbar edge on maximized windows.
+- name: widget.windows.hidden-taskbar-hack.paint
+  type: bool
+  value: false
+  mirror: always
 #endif
 
 # Whether to disable SwipeTracker (e.g. swipe-to-nav).
diff --git a/widget/windows/nsWindow.cpp b/widget/windows/nsWindow.cpp
index 699d6a31a957..81d2037ec645 100644
--- a/widget/windows/nsWindow.cpp
+++ b/widget/windows/nsWindow.cpp
@@ -945,6 +945,9 @@ nsresult nsWindow::Create(nsIWidget* aParent, const LayoutDeviceIntRect& aRect,
 
       // Skeleton ui is disabled when custom titlebar is off, see bug 1673092.
       SetCustomTitlebar(true);
+      // The skeleton UI already painted over the NC area, so there's no need
+      // to do that again; the effective non-client margins haven't changed.
+      mNeedsNCAreaClear = false;
 
       // Reset the WNDPROC for this window and its whole class, as we had
       // to use our own WNDPROC when creating the the skeleton UI window.
@@ -2682,6 +2685,19 @@ bool nsWindow::UpdateNonClientMargins(bool aReflowWindow) {
     // frame sizes for left, right and bottom since Windows will automagically
     // position the edges "offscreen" for maximized windows.
     metrics.mOffset.top = metrics.mCaptionHeight;
+
+    if (StaticPrefs::widget_windows_hidden_taskbar_hack_size()) {
+      if (mozilla::Maybe<UINT> maybeEdge = GetHiddenTaskbarEdge()) {
+        auto edge = maybeEdge.value();
+        if (ABE_LEFT == edge) {
+          metrics.mOffset.left -= kHiddenTaskbarSize;
+        } else if (ABE_RIGHT == edge) {
+          metrics.mOffset.right -= kHiddenTaskbarSize;
+        } else if (ABE_BOTTOM == edge || ABE_TOP == edge) {
+          metrics.mOffset.bottom -= kHiddenTaskbarSize;
+        }
+      }
+    }
   } else if (mPIPWindow &&
              !StaticPrefs::widget_windows_pip_decorations_enabled()) {
     metrics.mOffset = metrics.DefaultMargins();
@@ -2690,6 +2706,12 @@ bool nsWindow::UpdateNonClientMargins(bool aReflowWindow) {
   }
 
   UpdateOpaqueRegionInternal();
+  if (StaticPrefs::widget_windows_hidden_taskbar_hack_paint()) {
+    // We probably shouldn't need to clear the NC-area, but we need to in
+    // order to work around bug 642851.
+    mNeedsNCAreaClear = true;
+  }
+
   if (aReflowWindow) {
     // Force a reflow of content based on the new client
     // dimensions.
@@ -2759,6 +2781,26 @@ void nsWindow::SetResizeMargin(mozilla::LayoutDeviceIntCoord aResizeMargin) {
   mCustomResizeMargin = aResizeMargin;
 }
 
+LayoutDeviceIntRegion nsWindow::ComputeNonClientRegion() {
+  // +-+-----------------------+-+
+  // | | app non-client chrome | |
+  // | +-----------------------+ |
+  // | |   app client chrome   | | }
+  // | +-----------------------+ | }
+  // | |      app content      | | } area we don't want to invalidate
+  // | +-----------------------+ | }
+  // | |   app client chrome   | | }
+  // | +-----------------------+ |
+  // +---------------------------+ <
+  //  ^                         ^    windows non-client chrome
+  // client area = app *
+  auto winRect = LayoutDeviceIntRect(LayoutDeviceIntPoint(), GetSize());
+  LayoutDeviceIntRegion region{winRect};
+  winRect.Deflate(mCustomNonClientMetrics.DefaultMargins());
+  region.SubOut(winRect);
+  return region;
+}
+
 /**************************************************************
  *
  * SECTION: nsIWidget::SetCursor
@@ -7171,10 +7213,9 @@ LayoutDeviceIntRegion nsWindow::GetTranslucentRegion() {
   if (mTransparencyMode != TransparencyMode::Transparent) {
     return {};
   }
-  const auto clientRect =
-      LayoutDeviceIntRect(LayoutDeviceIntPoint(), GetClientSize());
-  LayoutDeviceIntRegion translucentRegion{clientRect};
-  translucentRegion.SubOut(mOpaqueRegion);
+  const auto winRect = LayoutDeviceIntRect(LayoutDeviceIntPoint(), GetSize());
+  LayoutDeviceIntRegion translucentRegion{winRect};
+  translucentRegion.SubOut(mOpaqueRegion.MovedBy(GetClientOffset()));
   return translucentRegion;
 }
 
diff --git a/widget/windows/nsWindow.h b/widget/windows/nsWindow.h
index 219e421ee8fc..b24a5fe6b594 100644
--- a/widget/windows/nsWindow.h
+++ b/widget/windows/nsWindow.h
@@ -554,6 +554,7 @@ class nsWindow final : public nsBaseWidget,
   bool UpdateNonClientMargins(bool aReflowWindow = true);
   void UpdateDarkModeToolbar();
   void ResetLayout();
+  LayoutDeviceIntRegion ComputeNonClientRegion();
   HWND GetOwnerWnd() const { return ::GetWindow(mWnd, GW_OWNER); }
   bool IsOwnerForegroundWindow() const {
     HWND owner = GetOwnerWnd();
@@ -645,7 +646,8 @@ class nsWindow final : public nsBaseWidget,
   LayoutDeviceIntRegion GetOpaqueRegionForTesting() const override {
     return mOpaqueRegion;
   }
-  // Gets the translucent region, in client coordinates.
+  // Gets the translucent region, relative to the whole window, including the
+  // NC area.
   LayoutDeviceIntRegion GetTranslucentRegion();
   void MaybeInvalidateTranslucentRegion();
 
@@ -847,8 +849,8 @@ class nsWindow final : public nsBaseWidget,
 
   // Graphics
   LayoutDeviceIntRect mLastPaintBounds;
-  // The region of the window we know is cleared to transparent already,
-  // in client coords.
+  // The region of the window we know is cleared to transparent already
+  // (relative to the whole window).
   LayoutDeviceIntRegion mClearedRegion;
 
   ResizeState mResizeState = NOT_RESIZING;
@@ -940,6 +942,10 @@ class nsWindow final : public nsBaseWidget,
 
   nsString mDesktopId MOZ_GUARDED_BY(mozilla::sMainThreadCapability);
 
+  // If set, indicates the non-client-area region must be cleared to black on
+  // next paint.
+  bool mNeedsNCAreaClear = false;
+
   friend class nsWindowGfx;
 
   static constexpr int kHiddenTaskbarSize = 2;
diff --git a/widget/windows/nsWindowGfx.cpp b/widget/windows/nsWindowGfx.cpp
index e492e9a62be2..b1e3b2eead80 100644
--- a/widget/windows/nsWindowGfx.cpp
+++ b/widget/windows/nsWindowGfx.cpp
@@ -181,6 +181,7 @@ bool nsWindow::OnPaint(uint32_t aNestingLevel) {
   // https://learn.microsoft.com/en-us/windows/win32/gdi/the-wm-paint-message
   HDC hDC = ::BeginPaint(mWnd, &ps);
   LayoutDeviceIntRegion region = GetRegionToPaint(ps, hDC);
+  LayoutDeviceIntRegion regionToClear;
   // Clear the translucent region if needed.
   if (mTransparencyMode == TransparencyMode::Transparent) {
     auto translucentRegion = GetTranslucentRegion();
@@ -191,27 +192,25 @@ bool nsWindow::OnPaint(uint32_t aNestingLevel) {
     //   regionToClear = translucentRegion - (mClearedRegion - region)
     //   mClearedRegion = translucentRegion;
     //   And add translucentRegion to region afterwards.
-    LayoutDeviceIntRegion regionToClear = translucentRegion;
+    regionToClear = translucentRegion;
     if (!mClearedRegion.IsEmpty()) {
       mClearedRegion.SubOut(region);
       regionToClear.SubOut(mClearedRegion);
     }
     region.OrWith(translucentRegion);
     mClearedRegion = std::move(translucentRegion);
-
-    // Don't clear the region for unaccelerated transparent windows;
-    // We clear the whole window below anyways, and doing so could cause
-    // flicker, as Windows doesn't guarantee atomicity even between
-    // ::BeginPaint and ::EndPaint, see bug 1958631.
-    if (!regionToClear.IsEmpty() && !isFallback) {
-      auto black = reinterpret_cast<HBRUSH>(::GetStockObject(BLACK_BRUSH));
-      // We could use RegionToHRGN, but at least for simple regions (and
-      // possibly for complex ones too?) FillRect is faster; see bug 1946365
-      // comment 12.
-      for (auto it = regionToClear.RectIter(); !it.Done(); it.Next()) {
-        auto rect = WinUtils::ToWinRect(it.Get());
-        ::FillRect(hDC, &rect, black);
-      }
+  }
+  if (mNeedsNCAreaClear) {
+    regionToClear.OrWith(ComputeNonClientRegion());
+    mNeedsNCAreaClear = false;
+  }
+  if (!regionToClear.IsEmpty()) {
+    auto black = reinterpret_cast<HBRUSH>(::GetStockObject(BLACK_BRUSH));
+    // We could use RegionToHRGN, but at least for simple regions (and possibly
+    // for complex ones too?) FillRect is faster; see bug 1946365 comment 12.
+    for (auto it = regionToClear.RectIter(); !it.Done(); it.Next()) {
+      auto rect = WinUtils::ToWinRect(it.Get());
+      ::FillRect(hDC, &rect, black);
     }
   }
 
