diff --git a/dom/media/mp4/MP4Decoder.cpp b/dom/media/mp4/MP4Decoder.cpp
index 9fce45c08fe6..b2230f490fd8 100644
--- a/dom/media/mp4/MP4Decoder.cpp
+++ b/dom/media/mp4/MP4Decoder.cpp
@@ -7,6 +7,9 @@
 #include "MP4Decoder.h"
 #include "H264.h"
 #include "VPXDecoder.h"
+#ifdef XP_WIN
+#include "mozilla/WindowsVersion.h"
+#endif
 #ifdef MOZ_AV1
 #  include "AOMDecoder.h"
 #endif
@@ -32,6 +35,14 @@ static bool IsWhitelistedH264Codec(const nsAString& aCodec) {
     return false;
   }
 
+#ifdef XP_WIN
+  // Disable 4k video on windows vista since it performs poorly.
+  if (!IsWin7OrLater() &&
+      level >= H264_LEVEL::H264_LEVEL_5) {
+    return false;
+  }
+#endif
+
   // Just assume what we can play on all platforms the codecs/formats that
   // WMF can play, since we don't have documentation about what other
   // platforms can play... According to the WMF documentation:
diff --git a/dom/media/platforms/wmf/WMFUtils.cpp b/dom/media/platforms/wmf/WMFUtils.cpp
index 3ed1b2467df7..8d55a9581212 100644
--- a/dom/media/platforms/wmf/WMFUtils.cpp
+++ b/dom/media/platforms/wmf/WMFUtils.cpp
@@ -494,13 +494,12 @@ LoadDLLs() {
 
 HRESULT
 MediaFoundationInitializer::MFStartup() {
-  if (IsWin7AndPre2000Compatible()) {
-    /*
-     * Specific exclude the usage of WMF on Win 7 with compatibility mode
-     * prior to Win 2000 as we may crash while trying to startup WMF.
-     * Using GetVersionEx API which takes compatibility mode into account.
-     * See Bug 1279171.
-     */
+  if (!IsVistaOrLater()) {
+    // *Only* use WMF on Vista and later, as if Firefox is run in Windows 95
+    // compatibility mode on Windows 7 (it does happen!) we may crash trying
+    // to startup WMF. So we need to detect the OS version here, as in
+    // compatibility mode IsVistaOrLater() and friends behave as if we're on
+    // the emulated version of Windows. See bug 1279171.
     return E_FAIL;
   }
 
@@ -509,6 +508,7 @@ MediaFoundationInitializer::MFStartup() {
     return hr;
   }
 
+  const int MF_VISTA_VERSION = (0x0001 << 16 | MF_API_VERSION);
   const int MF_WIN7_VERSION = (0x0002 << 16 | MF_API_VERSION);
 
   // decltype is unusable for functions having default parameters
@@ -516,8 +516,12 @@ MediaFoundationInitializer::MFStartup() {
   ENSURE_FUNCTION_PTR_(MFStartup, Mfplat.dll)
 
   hr = E_FAIL;
-  mozilla::mscom::EnsureMTA(
-      [&]() -> void { hr = MFStartupPtr(MF_WIN7_VERSION, MFSTARTUP_FULL); });
+  if (!IsWin7OrLater())
+    mozilla::mscom::EnsureMTA(
+        [&]() -> void { hr = MFStartupPtr(MF_VISTA_VERSION, MFSTARTUP_FULL); });
+  else
+    mozilla::mscom::EnsureMTA(
+        [&]() -> void { hr = MFStartupPtr(MF_WIN7_VERSION, MFSTARTUP_FULL); });
   return hr;
 }
 
