diff --git a/netwerk/sctp/src/netinet/sctp_os_userspace.h b/netwerk/sctp/src/netinet/sctp_os_userspace.h
index 493ae0267f77..20a1238e5011 100644
--- a/netwerk/sctp/src/netinet/sctp_os_userspace.h
+++ b/netwerk/sctp/src/netinet/sctp_os_userspace.h
@@ -51,7 +51,7 @@
 #include <windows.h>
 #include "user_environment.h"
 typedef CRITICAL_SECTION userland_mutex_t;
-#if WINVER < 0x0600
+#if defined(_WIN32)
 typedef CRITICAL_SECTION userland_rwlock_t;
 enum {
 	C_SIGNAL = 0,
diff --git a/netwerk/sctp/src/netinet/sctp_process_lock.h b/netwerk/sctp/src/netinet/sctp_process_lock.h
index feaa2c6acd33..91a6c3d9616f 100644
--- a/netwerk/sctp/src/netinet/sctp_process_lock.h
+++ b/netwerk/sctp/src/netinet/sctp_process_lock.h
@@ -130,7 +130,7 @@
 	LeaveCriticalSection(&SCTP_BASE_INFO(wq_addr_mtx))
 #define SCTP_WQ_ADDR_LOCK_ASSERT()
 
-#if WINVER < 0x0600
+#if defined(_WIN32)
 #define SCTP_INP_INFO_LOCK_INIT() \
 	InitializeCriticalSection(&SCTP_BASE_INFO(ipi_ep_mtx))
 #define SCTP_INP_INFO_LOCK_DESTROY() \
@@ -536,7 +536,7 @@
 
 #if defined(_WIN32)
 /* address list locks */
-#if WINVER < 0x0600
+#if defined(_WIN32)
 #define SCTP_IPI_ADDR_INIT() \
 	InitializeCriticalSection(&SCTP_BASE_INFO(ipi_addr_mtx))
 #define SCTP_IPI_ADDR_DESTROY() \
diff --git a/netwerk/sctp/src/netinet/sctp_userspace.c b/netwerk/sctp/src/netinet/sctp_userspace.c
index 71888901fec3..cd81d56f82bc 100644
--- a/netwerk/sctp/src/netinet/sctp_userspace.c
+++ b/netwerk/sctp/src/netinet/sctp_userspace.c
@@ -258,7 +258,7 @@ win_if_nametoindex(const char *ifname)
 	return index;
 }
 
-#if WINVER < 0x0600
+#if defined(_WIN32)
 /* These functions are written based on the code at
  * http://www.cs.wustl.edu/~schmidt/win32-cv-1.html
  * Therefore, for the rest of the file the following applies:
