diff --git a/dom/media/gmp/ChromiumCDMAdapter.cpp b/dom/media/gmp/ChromiumCDMAdapter.cpp
index ae29347fb08a..c27f600bbc19 100644
--- a/dom/media/gmp/ChromiumCDMAdapter.cpp
+++ b/dom/media/gmp/ChromiumCDMAdapter.cpp
@@ -23,6 +23,7 @@
 #  include "mozilla/NativeNt.h"
 #  include "mozilla/WindowsVersion.h"
 #  include <windows.h>
+#  include <bcrypt.h>
 #  include <strsafe.h>
 #  include <unordered_map>
 #  include <vector>
@@ -201,6 +202,9 @@ typedef BOOL(WINAPI* SetThreadInformationFnPtr)(
 static WindowsDllInterceptor::FuncHookType<SetThreadInformationFnPtr>
     sOriginalSetThreadInformationFnPtr;
 
+static WindowsDllInterceptor::FuncHookType<decltype(&GetProcAddress)>
+    sOriginalGetProcAddressFnPtr;
+
 static std::unordered_map<std::wstring, std::wstring>* sDeviceNames = nullptr;
 
 DWORD WINAPI QueryDosDeviceWHook(LPCWSTR lpDeviceName, LPWSTR lpTargetPath,
@@ -288,6 +292,19 @@ BOOL WINAPI MozSetThreadInformation(
   return TRUE;
 }
 
+BOOL WINAPI MozProcessPrng(PBYTE pbData, SIZE_T cbData) {
+  return ::BCryptGenRandom(nullptr, pbData, cbData,
+                           BCRYPT_USE_SYSTEM_PREFERRED_RNG) == STATUS_SUCCESS;
+}
+
+FARPROC WINAPI MozGetProcAddress(HMODULE hModule, LPCSTR lpProcName) {
+  if (strcmp(lpProcName, "ProcessPrng") == 0) {
+    return reinterpret_cast<FARPROC>(&MozProcessPrng);
+  }
+
+  return sOriginalGetProcAddressFnPtr(hModule, lpProcName);
+}
+
 static void InitializeHooks() {
   static bool initialized = false;
   if (initialized) {
@@ -323,6 +340,11 @@ static void InitializeHooks() {
                                                   &MozSetThreadInformation)) {
         GMP_LOG_WARNING("Failed to hook SetThreadInformation");
       }
+
+      if (!sOriginalGetProcAddressFnPtr.Set(
+              sKernel32Intercept, "GetProcAddress", &MozGetProcAddress)) {
+        GMP_LOG_WARNING("Failed to hook GetProcAddress");
+      }
     } else {
       GMP_LOG_WARNING(
           "Failed to rename to GetProcessMitigationPolicy and/or "
