diff --git a/layout/forms/crashtests/crashtests.list b/layout/forms/crashtests/crashtests.list
index da43270ba6ba..5bc0fa774607 100644
--- a/layout/forms/crashtests/crashtests.list
+++ b/layout/forms/crashtests/crashtests.list
@@ -68,7 +68,7 @@ load 1405830.html
 load 1418477.html
 load 1432853.html
 asserts(1-4) load 1460787-1.html
-load 1464165-1.html
+asserts(1-4) load 1464165-1.html # Big sizes.
 load 1471157.html
 load 1488219.html
 load 1600207.html
diff --git a/layout/generic/nsBlockFrame.cpp b/layout/generic/nsBlockFrame.cpp
index f6e9ba279426..5eaff0ebc5da 100644
--- a/layout/generic/nsBlockFrame.cpp
+++ b/layout/generic/nsBlockFrame.cpp
@@ -671,19 +671,13 @@ Maybe<nscoord> nsBlockFrame::GetNaturalBaselineBOffset(
                                aExportContext)
           : GetBaselineBOffset(LinesRBegin(), LinesREnd(), aWM, aBaselineGroup,
                                aExportContext);
-  if (!offset && IsButtonLike()) {
-    for (const auto& line : Reversed(Lines())) {
-      if (line.IsEmpty()) {
-        continue;
-      }
-      // Buttons use the end of the content as a baseline if we haven't found
-      // one yet.
-      nscoord bEnd = line.BEnd();
-      offset.emplace(aBaselineGroup == BaselineSharingGroup::Last
-                         ? BSize(aWM) - bEnd
-                         : bEnd);
-      break;
-    }
+  if (!offset && IsButtonLike() && !mLines.empty()) {
+    // Buttons use the end of the content as a baseline if we haven't found one
+    // yet.
+    nscoord bEnd = mLines.back()->BEnd();
+    offset.emplace(aBaselineGroup == BaselineSharingGroup::Last
+                       ? BSize(aWM) - bEnd
+                       : bEnd);
   }
   return offset;
 }
@@ -4154,6 +4148,7 @@ bool nsBlockFrame::IsEmpty() {
   if (!IsSelfEmpty()) {
     return false;
   }
+
   return LinesAreEmpty();
 }
 
@@ -8347,7 +8342,6 @@ void nsBlockFrame::SetInitialChildList(ChildListID aListID,
           !GetParent()->IsListControlFrame()) ||
          pseudo == PseudoStyleType::MozSvgText) &&
         !IsMathMLFrame() && !IsColumnSetWrapperFrame() &&
-        !IsComboboxControlFrame() &&
         RefPtr<ComputedStyle>(GetFirstLetterStyle(PresContext())) != nullptr;
     NS_ASSERTION(haveFirstLetterStyle ==
                      HasAnyStateBits(NS_BLOCK_HAS_FIRST_LETTER_STYLE),
