diff --git a/mfbt/WindowsVersion.h b/mfbt/WindowsVersion.h
index 2e9621cb5d6e..fcf24acc2e92 100644
--- a/mfbt/WindowsVersion.h
+++ b/mfbt/WindowsVersion.h
@@ -84,44 +84,6 @@ IsWindowsBuildOrLater(uint32_t aBuild)
   return false;
 }
 
-#if defined(_M_X64) || defined(_M_AMD64)
-// We support only Win7 or later on Win64.
-MOZ_ALWAYS_INLINE bool
-IsXPSP3OrLater()
-{
-  return true;
-}
-
-MOZ_ALWAYS_INLINE bool
-IsWin2003OrLater()
-{
-  return true;
-}
-
-MOZ_ALWAYS_INLINE bool
-IsWin2003SP2OrLater()
-{
-  return true;
-}
-
-MOZ_ALWAYS_INLINE bool
-IsVistaOrLater()
-{
-  return true;
-}
-
-MOZ_ALWAYS_INLINE bool
-IsVistaSP1OrLater()
-{
-  return true;
-}
-
-MOZ_ALWAYS_INLINE bool
-IsWin7OrLater()
-{
-  return true;
-}
-#else
 MOZ_ALWAYS_INLINE bool
 IsXPSP3OrLater()
 {
@@ -157,7 +119,6 @@ IsWin7OrLater()
 {
   return IsWindowsVersionOrLater(0x06010000ul);
 }
-#endif
 
 MOZ_ALWAYS_INLINE bool IsWin7SP1OrLater() {
   return IsWindowsVersionOrLater(0x06010100ul);
