diff --git a/widget/windows/nsWindow.cpp b/widget/windows/nsWindow.cpp
index ce127e1a393c..aa03e13b9721 100644
--- a/widget/windows/nsWindow.cpp
+++ b/widget/windows/nsWindow.cpp
@@ -2409,15 +2409,6 @@ LayoutDeviceIntRect nsWindow::GetBounds() {
   return rect;
 }
 
-LayoutDeviceIntSize nsWindow::GetSize() const {
-  if (!mWnd) {
-    return mBounds.Size();
-  }
-  RECT r;
-  VERIFY(::GetWindowRect(mWnd, &r));
-  return {r.right - r.left, r.bottom - r.top};
-}
-
 // Get this component dimension
 LayoutDeviceIntRect nsWindow::GetClientBounds() {
   if (!mWnd) {
@@ -2651,8 +2642,8 @@ bool nsWindow::UpdateNonClientMargins(bool aReflowWindow) {
 
   UpdateOpaqueRegionInternal();
   if (StaticPrefs::widget_windows_hidden_taskbar_hack_paint()) {
-    // We probably shouldn't need to clear the NC-area, but we need to in
-    // order to work around bug 642851.
+    // We probably shouldn't need to clear the NC-area if we're an opaque window,
+    // but we need to in order to work around bug 642851.
     mNeedsNCAreaClear = true;
   }
 
@@ -2724,7 +2715,7 @@ void nsWindow::SetResizeMargin(mozilla::LayoutDeviceIntCoord aResizeMargin) {
   mCustomResizeMargin = aResizeMargin;
 }
 
-LayoutDeviceIntRegion nsWindow::ComputeNonClientRegion() {
+nsAutoRegion nsWindow::ComputeNonClientHRGN() {
   // +-+-----------------------+-+
   // | | app non-client chrome | |
   // | +-----------------------+ |
@@ -2737,11 +2728,24 @@ LayoutDeviceIntRegion nsWindow::ComputeNonClientRegion() {
   // +---------------------------+ <
   //  ^                         ^    windows non-client chrome
   // client area = app *
-  auto winRect = LayoutDeviceIntRect(LayoutDeviceIntPoint(), GetSize());
-  LayoutDeviceIntRegion region{winRect};
-  winRect.Deflate(mCustomNonClientMetrics.DefaultMargins());
-  region.SubOut(winRect);
-  return region;
+  RECT rect;
+  GetWindowRect(mWnd, &rect);
+  MapWindowPoints(nullptr, mWnd, (LPPOINT)&rect, 2);
+  nsAutoRegion winRgn(::CreateRectRgnIndirect(&rect));
+
+  // Subtract app client chrome and app content leaving
+  // windows non-client chrome and app non-client chrome
+  // in winRgn.
+  ::GetWindowRect(mWnd, &rect);
+  rect.top += mCustomNonClientMetrics.mCaptionHeight +
+              mCustomNonClientMetrics.mVertResizeMargin;
+  rect.right -= mCustomNonClientMetrics.mHorResizeMargin;
+  rect.bottom -= mCustomNonClientMetrics.mVertResizeMargin;
+  rect.left += mCustomNonClientMetrics.mHorResizeMargin;
+  ::MapWindowPoints(nullptr, mWnd, (LPPOINT)&rect, 2);
+  nsAutoRegion clientRgn(::CreateRectRgnIndirect(&rect));
+  ::CombineRgn(winRgn, winRgn, clientRgn, RGN_DIFF);
+  return nsAutoRegion(winRgn.out());
 }
 
 /**************************************************************
@@ -3856,10 +3860,10 @@ void nsWindow::SetIsEarlyBlankWindow(bool aIsEarlyBlankWindow) {
     return;
   }
   mIsEarlyBlankWindow = aIsEarlyBlankWindow;
-  if (!aIsEarlyBlankWindow) {
+  if (!aIsEarlyBlankWindow && mNeedsNCAreaClear) {
     // We skip processing WM_PAINT messages while we're the blank window;
     // ensure we get one to do any work we might have missed.
-    MaybeInvalidateTranslucentRegion();
+    ::RedrawWindow(mWnd, nullptr, nullptr, RDW_INVALIDATE | RDW_INTERNALPAINT);
   }
 }
 
@@ -4994,10 +4998,18 @@ bool nsWindow::ProcessMessageInternal(UINT msg, WPARAM& wParam, LPARAM& lParam,
     } break;
 
     case WM_NCPAINT: {
-      // ClearType changes often don't send a WM_SETTINGCHANGE message. But
-      // they do seem to always send a WM_NCPAINT message, so let's update on
-      // that.
+      /*
+       * ClearType changes often don't send a WM_SETTINGCHANGE message. But they
+       * do seem to always send a WM_NCPAINT message, so let's update on that.
+       */
       gfxDWriteFont::UpdateSystemTextVars();
+      if (mCustomNonClient &&
+          mTransparencyMode == TransparencyMode::Transparent) {
+        // We rely on dwm for glass / semi-transparent window painting, so we
+        // just need to make sure to clear the non-client area next time we get
+        // around to doing a main-thread paint.
+        mNeedsNCAreaClear = true;
+      }
     } break;
 
     case WM_POWERBROADCAST:
@@ -7156,34 +7168,6 @@ void nsWindow::UpdateOpaqueRegion(const LayoutDeviceIntRegion& aRegion) {
   UpdateOpaqueRegionInternal();
 }
 
-LayoutDeviceIntRegion nsWindow::GetTranslucentRegion() {
-  if (mTransparencyMode != TransparencyMode::Transparent) {
-    return {};
-  }
-  const auto winRect = LayoutDeviceIntRect(LayoutDeviceIntPoint(), GetSize());
-  LayoutDeviceIntRegion translucentRegion{winRect};
-  translucentRegion.SubOut(mOpaqueRegion.MovedBy(GetClientOffset()));
-  return translucentRegion;
-}
-
-void nsWindow::MaybeInvalidateTranslucentRegion() {
-  if (mTransparencyMode != TransparencyMode::Transparent) {
-    return;
-  }
-  const auto translucent = GetTranslucentRegion();
-  if (translucent.IsEmpty() || mClearedRegion.Contains(translucent)) {
-    return;
-  }
-  // We need to clear some part of the window that isn't cleared already, make
-  // sure we trigger a WM_PAINT message.
-  //
-  // NOTE(emilio): we could provide a finer grained region here (i.e., only
-  // invalidate the translucent region, or even only the bits that are not yet
-  // cleared), but we don't do much with that region in OnPaint message so this
-  // seems fine for now.
-  ::RedrawWindow(mWnd, nullptr, nullptr, RDW_INVALIDATE | RDW_INTERNALPAINT);
-}
-
 void nsWindow::UpdateOpaqueRegionInternal() {
   MARGINS margins{0};
   if (mTransparencyMode == TransparencyMode::Transparent) {
@@ -7208,7 +7192,10 @@ void nsWindow::UpdateOpaqueRegionInternal() {
     }
   }
   DwmExtendFrameIntoClientArea(mWnd, &margins);
-  MaybeInvalidateTranslucentRegion();
+  if (mTransparencyMode == TransparencyMode::Transparent) {
+    mNeedsNCAreaClear = true;
+    ::RedrawWindow(mWnd, nullptr, nullptr, RDW_INVALIDATE | RDW_INTERNALPAINT);
+  }
 }
 
 /**************************************************************
diff --git a/widget/windows/nsWindow.h b/widget/windows/nsWindow.h
index 0cbdce82f17f..2ae8dd245247 100644
--- a/widget/windows/nsWindow.h
+++ b/widget/windows/nsWindow.h
@@ -229,7 +229,6 @@ class nsWindow final : public nsBaseWidget {
   [[nodiscard]] nsresult GetRestoredBounds(LayoutDeviceIntRect& aRect) override;
   LayoutDeviceIntRect GetClientBounds() override;
   LayoutDeviceIntPoint GetClientOffset() override;
-  LayoutDeviceIntSize GetSize() const;
   void SetCursor(const Cursor&) override;
   bool PrepareForFullscreenTransition(nsISupports** aData) override;
   void PerformFullscreenTransition(FullscreenTransitionStage aStage,
@@ -551,7 +550,7 @@ class nsWindow final : public nsBaseWidget {
   bool UpdateNonClientMargins(bool aReflowWindow = true);
   void UpdateDarkModeToolbar();
   void ResetLayout();
-  LayoutDeviceIntRegion ComputeNonClientRegion();
+  nsAutoRegion ComputeNonClientHRGN();
   HWND GetOwnerWnd() const { return ::GetWindow(mWnd, GW_OWNER); }
   bool IsOwnerForegroundWindow() const {
     HWND owner = GetOwnerWnd();
@@ -643,10 +642,6 @@ class nsWindow final : public nsBaseWidget {
   LayoutDeviceIntRegion GetOpaqueRegionForTesting() const override {
     return mOpaqueRegion;
   }
-  // Gets the translucent region, relative to the whole window, including the
-  // NC area.
-  LayoutDeviceIntRegion GetTranslucentRegion();
-  void MaybeInvalidateTranslucentRegion();
 
   void SetColorScheme(const mozilla::Maybe<mozilla::ColorScheme>&) override;
   void SetMicaBackdrop(bool) override;
@@ -840,9 +835,6 @@ class nsWindow final : public nsBaseWidget {
 
   // Graphics
   LayoutDeviceIntRect mLastPaintBounds;
-  // The region of the window we know is cleared to transparent already
-  // (relative to the whole window).
-  LayoutDeviceIntRegion mClearedRegion;
 
   ResizeState mResizeState = NOT_RESIZING;
 
diff --git a/widget/windows/nsWindowGfx.cpp b/widget/windows/nsWindowGfx.cpp
index 4c10acf9fa11..f097a4d802a1 100644
--- a/widget/windows/nsWindowGfx.cpp
+++ b/widget/windows/nsWindowGfx.cpp
@@ -175,6 +175,12 @@ bool nsWindow::OnPaint(uint32_t aNestingLevel) {
     listener->WillPaintWindow(this);
   }
 
+  const LayoutDeviceIntRect winRect = [&] {
+    RECT r;
+    ::GetWindowRect(mWnd, &r);
+    ::MapWindowPoints(nullptr, mWnd, (LPPOINT)&r, 2);
+    return WinUtils::ToIntRect(r);
+  }();
   // BeginPaint/EndPaint must be called to make Windows think that invalid
   // area is painted. Otherwise it will continue sending the same message
   // endlessly. Note that we need to call it after WillPaintWindow, which
@@ -184,29 +190,26 @@ bool nsWindow::OnPaint(uint32_t aNestingLevel) {
   // https://learn.microsoft.com/en-us/windows/win32/gdi/the-wm-paint-message
   HDC hDC = ::BeginPaint(mWnd, &ps);
   LayoutDeviceIntRegion region = GetRegionToPaint(ps, hDC);
-  LayoutDeviceIntRegion regionToClear;
-  // Clear the translucent region if needed.
+  LayoutDeviceIntRegion translucentRegion;
   if (mTransparencyMode == TransparencyMode::Transparent) {
-    auto translucentRegion = GetTranslucentRegion();
+    translucentRegion = LayoutDeviceIntRegion{winRect};
+    translucentRegion.SubOut(mOpaqueRegion);
     region.OrWith(translucentRegion);
-    // Clear the parts of the translucent region that aren't clear already.
-    regionToClear = translucentRegion;
-    if (!didResize && !mClearedRegion.IsEmpty()) {
-      // TODO(emilio): Would be nice to reuse some of the cleared region when
-      // resizing, but it's not trivial.
-      regionToClear.SubOut(mClearedRegion);
-    }
-    mClearedRegion = std::move(translucentRegion);
-  }
-  if (mNeedsNCAreaClear) {
-    regionToClear.OrWith(ComputeNonClientRegion());
-    mNeedsNCAreaClear = false;
   }
-  if (!regionToClear.IsEmpty()) {
+
+  if (mNeedsNCAreaClear ||
+      (didResize && mTransparencyMode == TransparencyMode::Transparent)) {
+    // We need to clear the non-client-area region, and the transparent parts
+    // of the window to black (once).
     auto black = reinterpret_cast<HBRUSH>(::GetStockObject(BLACK_BRUSH));
-    nsAutoRegion hRgnToClear(WinUtils::RegionToHRGN(regionToClear));
-    ::FillRgn(hDC, hRgnToClear, black);
+    nsAutoRegion regionToClear(ComputeNonClientHRGN());
+    if (!translucentRegion.IsEmpty()) {
+      nsAutoRegion translucent(WinUtils::RegionToHRGN(translucentRegion));
+      ::CombineRgn(regionToClear, regionToClear, translucent, RGN_OR);
+    }
+    ::FillRgn(hDC, regionToClear, black);
   }
+  mNeedsNCAreaClear = false;
 
   bool didPaint = false;
   auto endPaint = MakeScopeExit([&] {
