diff --git a/browser/installer/windows/docs/InstallerBuild.rst b/browser/installer/windows/docs/InstallerBuild.rst
index e86d17449164..0473bb3f1d20 100644
--- a/browser/installer/windows/docs/InstallerBuild.rst
+++ b/browser/installer/windows/docs/InstallerBuild.rst
@@ -5,7 +5,7 @@ Installer Build
 How to build the installers
 ---------------------------
 
-The easiest way to build an installer in your local tree is to run ``mach package``. The finished installers will be in ``$OBJDIR/dist``. You have to have a build of the application already done before that will work, but `artifact builds <https://developer.mozilla.org/en-US/docs/Mozilla/Developer_guide/Build_Instructions/Artifact_builds>`_ are supported, so that can save you a lot of time if you're only doing installer work.
+The easiest way to build an installer in your local tree is to run ``mach package``. The finished installers will be in ``$OBJDIR/dist/install/sea/``. You have to have a build of the application already done before that will work, but `artifact builds <https://developer.mozilla.org/en-US/docs/Mozilla/Developer_guide/Build_Instructions/Artifact_builds>`_ are supported, so that can save you a lot of time if you're only doing installer work.
 
 You'll also need to be on a Windows machine; the installer build depends on tools that aren't available for other platforms.
 
diff --git a/python/mach/mach/command_util.py b/python/mach/mach/command_util.py
index ecf2f89b17c0..9a36b9998948 100644
--- a/python/mach/mach/command_util.py
+++ b/python/mach/mach/command_util.py
@@ -187,9 +187,6 @@ MACH_COMMANDS = {
     "release-history": MachCommandReference("taskcluster/mach_commands.py"),
     "remote": MachCommandReference("remote/mach_commands.py"),
     "repackage": MachCommandReference("python/mozbuild/mozbuild/mach_commands.py"),
-    "repackage-single-locales": MachCommandReference(
-        "python/mozbuild/mozbuild/mach_commands.py"
-    ),
     "resource-usage": MachCommandReference(
         "python/mozbuild/mozbuild/build_commands.py",
     ),
diff --git a/python/mozbuild/mozbuild/backend/base.py b/python/mozbuild/mozbuild/backend/base.py
index 68561b8ba255..037c17970577 100644
--- a/python/mozbuild/mozbuild/backend/base.py
+++ b/python/mozbuild/mozbuild/backend/base.py
@@ -236,7 +236,6 @@ class BuildBackend(LoggingMixin):
             purgecaches_dirs.append(bundledir)
 
         for dir in purgecaches_dirs:
-            os.makedirs(dir, exist_ok=True)
             with open(mozpath.join(dir, ".purgecaches"), "w") as f:
                 f.write("\n")
 
diff --git a/taskcluster/kinds/l10n/kind.yml b/taskcluster/kinds/l10n/kind.yml
index 8b27aaa054b6..537fe85fda49 100644
--- a/taskcluster/kinds/l10n/kind.yml
+++ b/taskcluster/kinds/l10n/kind.yml
@@ -36,7 +36,10 @@ tasks:
                 # OSX has a special locale for japanese
                 macosx64.*: [ja]
                 default: [ja-JP-mac]
-        run-time: 1800
+        run-time:
+            by-build-platform:
+                default: 900
+                win.*: 1800
         secrets: false
         fetches:
             toolchain:
diff --git a/testing/mozharness/configs/single_locale/macosx64.py b/testing/mozharness/configs/single_locale/macosx64.py
index e78dd2decc64..07c1f0722115 100644
--- a/testing/mozharness/configs/single_locale/macosx64.py
+++ b/testing/mozharness/configs/single_locale/macosx64.py
@@ -4,6 +4,9 @@
 
 config = {
     "mozconfig_platform": "macosx64",
+    "upload_env_extra": {
+        "MOZ_PKG_PLATFORM": "mac",
+    },
     # l10n
     "ignore_locales": ["en-US", "ja"],
 }
diff --git a/testing/mozharness/scripts/desktop_l10n.py b/testing/mozharness/scripts/desktop_l10n.py
index fdc66e7e79d5..70d56e978960 100755
--- a/testing/mozharness/scripts/desktop_l10n.py
+++ b/testing/mozharness/scripts/desktop_l10n.py
@@ -7,16 +7,20 @@
 This script manages Desktop repacks for nightly builds.
 """
 
+import glob
 import os
+import shlex
 import sys
 
 # load modules from parent dir
 sys.path.insert(1, os.path.dirname(sys.path[0]))  # noqa
 
+from mozharness.base.errors import MakefileErrorList
 from mozharness.base.script import BaseScript
 from mozharness.base.vcs.vcsbase import VCSMixin
 from mozharness.mozilla.automation import AutomationMixin
 from mozharness.mozilla.building.buildbase import (
+    MakeUploadOutputParser,
     get_mozconfig_path,
 )
 from mozharness.mozilla.l10n.locales import LocalesMixin
@@ -29,6 +33,14 @@ except ImportError:
     import json
 
 
+# needed by _map
+SUCCESS = 0
+FAILURE = 1
+
+SUCCESS_STR = "Success"
+FAILURE_STR = "Failed"
+
+
 # DesktopSingleLocale {{{1
 class DesktopSingleLocale(LocalesMixin, AutomationMixin, VCSMixin, BaseScript):
     """Manages desktop repacks"""
@@ -76,6 +88,10 @@ class DesktopSingleLocale(LocalesMixin, AutomationMixin, VCSMixin, BaseScript):
 
         self.bootstrap_env = None
         self.upload_env = None
+        self.upload_urls = {}
+        self.pushdate = None
+        # upload_files is a dictionary of files to upload, keyed by locale.
+        self.upload_files = {}
 
     # Helper methods {{{2
     def query_bootstrap_env(self):
@@ -111,7 +127,15 @@ class DesktopSingleLocale(LocalesMixin, AutomationMixin, VCSMixin, BaseScript):
             return self.upload_env
         config = self.config
 
-        self.upload_env = self.query_env(partial_env=config.get("upload_env"))
+        upload_env = self.query_env(partial_env=config.get("upload_env"))
+        # check if there are any extra option from the platform configuration
+        # and append them to the env
+
+        if "upload_env_extra" in config:
+            for extra in config["upload_env_extra"]:
+                upload_env[extra] = config["upload_env_extra"][extra]
+
+        self.upload_env = upload_env
         return self.upload_env
 
     def query_l10n_env(self):
@@ -119,14 +143,63 @@ class DesktopSingleLocale(LocalesMixin, AutomationMixin, VCSMixin, BaseScript):
         l10n_env.update(self.query_bootstrap_env())
         return l10n_env
 
+    def _query_make_variable(self, variable, make_args=None):
+        """returns the value of make echo-variable-<variable>
+        it accepts extra make arguements (make_args)
+        """
+        dirs = self.query_abs_dirs()
+        make_args = make_args or []
+        target = ["echo-variable-%s" % variable] + make_args
+        cwd = dirs["abs_locales_dir"]
+        raw_output = self._get_output_from_make(
+            target, cwd=cwd, env=self.query_bootstrap_env()
+        )
+        # we want to log all the messages from make
+        output = []
+        for line in raw_output.split("\n"):
+            output.append(line.strip())
+        output = " ".join(output).strip()
+        self.info("echo-variable-%s: %s" % (variable, output))
+        return output
+
+    def _map(self, func, items):
+        """runs func for any item in items, calls the add_failure() for each
+        error. It assumes that function returns 0 when successful.
+        returns a two element tuple with (success_count, total_count)"""
+        success_count = 0
+        total_count = len(items)
+        name = func.__name__
+        for item in items:
+            result = func(item)
+            if result == SUCCESS:
+                #  success!
+                success_count += 1
+            else:
+                #  func failed...
+                message = "failure: %s(%s)" % (name, item)
+                self.add_failure(item, message)
+        return (success_count, total_count)
+
     # Actions {{{2
     def clone_locales(self):
         self.pull_locale_source()
 
     def setup(self):
         """setup step"""
+        self._run_tooltool()
         self._copy_mozconfig()
         self._mach_configure()
+        self._make_export()
+
+    def _make_export(self):
+        """this step creates nsinstall, needed on Windows hosts
+        and creates buildid.h, used by NSIS installer for Windows UBR telemetry
+        """
+        dirs = self.query_abs_dirs()
+        config_dir = os.path.join(dirs["abs_obj_dir"], "config")
+        env = self.query_bootstrap_env()
+        self._make(target=["export"], cwd=config_dir, env=env)
+        return self._make(target=["buildid.h"], cwd=dirs["abs_obj_dir"], env=env)
 
     def _copy_mozconfig(self):
         """copies the mozconfig file into abs_src_dir/.mozconfig
@@ -159,19 +232,211 @@ class DesktopSingleLocale(LocalesMixin, AutomationMixin, VCSMixin, BaseScript):
     def _get_mach_executable(self):
         return [sys.executable, "mach"]
 
-    def repack(self):
-        env = self.query_bootstrap_env()
-        return self._mach(
-            target=[
-                "repackage-single-locales",
-                "--verbose",
-                "--dest",
-                self.config["upload_env"]["UPLOAD_PATH"],
-                "--locales",
-            ]
-            + list(sorted(self.query_locales())),
+    def _make(
+        self,
+        target,
+        cwd,
+        env,
+        error_list=MakefileErrorList,
+        halt_on_failure=True,
+        output_parser=None,
+    ):
+        """Runs make. Returns the exit code"""
+        make = ["make"]
+        if target:
+            make = make + target
+        return self.run_command(
+            make,
+            cwd=cwd,
             env=env,
+            error_list=error_list,
+            halt_on_failure=halt_on_failure,
+            output_parser=output_parser,
+        )
+
+    def _get_output_from_make(
+        self, target, cwd, env, halt_on_failure=True, ignore_errors=False
+    ):
+        """runs make and returns the output of the command"""
+        return self.get_output_from_command(
+            ["make"] + target,
+            cwd=cwd,
+            env=env,
+            silent=True,
+            halt_on_failure=halt_on_failure,
+            ignore_errors=ignore_errors,
+        )
+
+    def make_upload(self, locale):
+        """wrapper for make upload command"""
+        env = self.query_l10n_env()
+        dirs = self.query_abs_dirs()
+        target = ["upload", "AB_CD=%s" % (locale)]
+        cwd = dirs["abs_locales_dir"]
+        parser = MakeUploadOutputParser(config=self.config, log_obj=self.log_obj)
+        retval = self._make(
+            target=target, cwd=cwd, env=env, halt_on_failure=False, output_parser=parser
         )
+        if retval == SUCCESS:
+            self.info("Upload successful (%s)" % locale)
+            ret = SUCCESS
+        else:
+            self.error("failed to upload %s" % locale)
+            ret = FAILURE
+
+        if ret == FAILURE:
+            # If we failed above, we shouldn't even attempt a SIMPLE_NAME move
+            # even if we are configured to do so
+            return ret
+
+        # XXX Move the files to a SIMPLE_NAME format until we can enable
+        #     Simple names in the build system
+        if self.config.get("simple_name_move"):
+            # Assume an UPLOAD PATH
+            upload_target = self.config["upload_env"]["UPLOAD_PATH"]
+            target_path = os.path.join(upload_target, locale)
+            self.mkdir_p(target_path)
+            glob_name = "*.%s.*" % locale
+            matches = (
+                glob.glob(os.path.join(upload_target, glob_name))
+                + glob.glob(os.path.join(upload_target, "update", glob_name))
+                + glob.glob(os.path.join(upload_target, "*", "xpi", glob_name))
+                + glob.glob(os.path.join(upload_target, "install", "sea", glob_name))
+                + glob.glob(os.path.join(upload_target, "setup.exe"))
+                + glob.glob(os.path.join(upload_target, "setup-stub.exe"))
+            )
+            targets_exts = [
+                "tar.bz2",
+                "tar.xz",
+                "dmg",
+                "langpack.xpi",
+                "checksums",
+                "zip",
+                "installer.exe",
+                "installer-stub.exe",
+            ]
+            targets = [(".%s" % (ext,), "target.%s" % (ext,)) for ext in targets_exts]
+            targets.extend([(f, f) for f in ("setup.exe", "setup-stub.exe")])
+            for f in matches:
+                possible_targets = [
+                    (tail, target_file)
+                    for (tail, target_file) in targets
+                    if f.endswith(tail)
+                ]
+                if len(possible_targets) == 1:
+                    _, target_file = possible_targets[0]
+                    # Remove from list of available options for this locale
+                    targets.remove(possible_targets[0])
+                else:
+                    # wasn't valid (or already matched)
+                    raise RuntimeError(
+                        "Unexpected matching file name encountered: %s" % f
+                    )
+                self.move(os.path.join(f), os.path.join(target_path, target_file))
+            self.log("Converted uploads for %s to simple names" % locale)
+        return ret
+
+    def set_upload_files(self, locale):
+        # The tree doesn't have a good way of exporting the list of files
+        # created during locale generation, but we can grab them by echoing the
+        # UPLOAD_FILES variable for each locale.
+        env = self.query_l10n_env()
+        target = [
+            "echo-variable-UPLOAD_FILES",
+            "echo-variable-CHECKSUM_FILES",
+            "AB_CD=%s" % locale,
+        ]
+        dirs = self.query_abs_dirs()
+        cwd = dirs["abs_locales_dir"]
+        # Bug 1242771 - echo-variable-UPLOAD_FILES via mozharness fails when stderr is found
+        #    we should ignore stderr as unfortunately it's expected when parsing for values
+        output = self._get_output_from_make(
+            target=target, cwd=cwd, env=env, ignore_errors=True
+        )
+        self.info('UPLOAD_FILES is "%s"' % output)
+        files = shlex.split(output)
+        if not files:
+            self.error("failed to get upload file list for locale %s" % locale)
+            return FAILURE
+
+        self.upload_files[locale] = [
+            os.path.abspath(os.path.join(cwd, f)) for f in files
+        ]
+        return SUCCESS
+
+    def make_installers(self, locale):
+        """wrapper for make installers-(locale)"""
+        env = self.query_l10n_env()
+        env["PYTHONIOENCODING"] = "utf-8"
+        self._copy_mozconfig()
+        dirs = self.query_abs_dirs()
+        cwd = os.path.join(dirs["abs_locales_dir"])
+        target = [
+            "installers-%s" % locale,
+        ]
+        return self._make(target=target, cwd=cwd, env=env, halt_on_failure=False)
+
+    def repack_locale(self, locale):
+        """wraps the logic for make installers and generating
+        complete updates."""
+
+        # run make installers
+        if self.make_installers(locale) != SUCCESS:
+            self.error("make installers-%s failed" % (locale))
+            return FAILURE
+
+        # now try to upload the artifacts
+        if self.make_upload(locale):
+            self.error("make upload for locale %s failed!" % (locale))
+            return FAILURE
+
+        # set_upload_files() should be called after make upload, to make sure
+        # we have all files in place (checksums, etc)
+        if self.set_upload_files(locale):
+            self.error("failed to get list of files to upload for locale %s" % locale)
+            return FAILURE
+
+        return SUCCESS
+
+    def repack(self):
+        """creates the repacks and udpates"""
+        self._map(self.repack_locale, self.query_locales())
+
+    def _run_tooltool(self):
+        env = self.query_bootstrap_env()
+        config = self.config
+        dirs = self.query_abs_dirs()
+        manifest_src = os.environ.get("TOOLTOOL_MANIFEST")
+        if not manifest_src:
+            manifest_src = config.get("tooltool_manifest_src")
+        if not manifest_src:
+            return
+        python = sys.executable
+
+        cmd = [
+            python,
+            "-u",
+            os.path.join(dirs["abs_src_dir"], "mach"),
+            "artifact",
+            "toolchain",
+            "-v",
+            "--retry",
+            "4",
+            "--artifact-manifest",
+            os.path.join(dirs["abs_src_dir"], "toolchains.json"),
+        ]
+        if manifest_src:
+            cmd.extend(
+                [
+                    "--tooltool-manifest",
+                    os.path.join(dirs["abs_src_dir"], manifest_src),
+                ]
+            )
+        cache = config["bootstrap_env"].get("TOOLTOOL_CACHE")
+        if cache:
+            cmd.extend(["--cache-dir", cache])
+        self.info(str(cmd))
+        self.run_command(cmd, cwd=dirs["abs_src_dir"], halt_on_failure=True, env=env)
 
 
 # main {{{
diff --git a/toolkit/locales/l10n.mk b/toolkit/locales/l10n.mk
index 3f41b29a8753..4e27d1815a0e 100644
--- a/toolkit/locales/l10n.mk
+++ b/toolkit/locales/l10n.mk
@@ -24,6 +24,9 @@
 # The including makefile should provide values for the variables
 #   MOZ_APP_VERSION and MOZ_LANGPACK_EID.
 
+
+run_for_effects := $(shell if test ! -d $(DIST); then $(NSINSTALL) -D $(DIST); fi)
+
 # This makefile uses variable overrides from the l10n-% target to
 # build non-default locales to non-default dist/ locations. Be aware!
 
@@ -44,6 +47,8 @@ ACDEFINES += \
 	-DMOZ_APP_VERSION=$(MOZ_APP_VERSION) \
 	-DMOZ_APP_MAXVERSION=$(MOZ_APP_MAXVERSION) \
 	-DLOCALE_SRCDIR=$(abspath $(LOCALE_SRCDIR)) \
+	-DPKG_BASENAME='$(PKG_BASENAME)' \
+	-DPKG_INST_BASENAME='$(PKG_INST_BASENAME)' \
 	$(NULL)
 
 # export some global defines for l10n repacks
diff --git a/toolkit/moz.configure b/toolkit/moz.configure
index 11baf88b5801..2a2228f20b04 100644
--- a/toolkit/moz.configure
+++ b/toolkit/moz.configure
@@ -3655,15 +3655,16 @@ set_define(
     when=depends("--enable-parental-controls")(lambda x: not x),
 )
 
-# Package renaming
+# Automation renaming
 # ==============================================================
 
-option(env="MOZ_SIMPLE_PACKAGE_NAME", nargs=1, help="Package name override")
-set_config(
-    "MOZ_SIMPLE_PACKAGE_NAME", depends_if("MOZ_SIMPLE_PACKAGE_NAME")(lambda x: x[0])
-)
-option(env="MOZ_PKG_SPECIAL", nargs=1, help="Name of special moz flavor")
-set_config("MOZ_PKG_SPECIAL", depends_if("MOZ_PKG_SPECIAL")(lambda x: x[0]))
+with only_when(moz_automation):
+    option(env="MOZ_SIMPLE_PACKAGE_NAME", nargs=1, help="Package name override")
+    set_config(
+        "MOZ_SIMPLE_PACKAGE_NAME", depends_if("MOZ_SIMPLE_PACKAGE_NAME")(lambda x: x[0])
+    )
+    option(env="MOZ_PKG_SPECIAL", nargs=1, help="Name of special moz flavor")
+    set_config("MOZ_PKG_SPECIAL", depends_if("MOZ_PKG_SPECIAL")(lambda x: x[0]))
 
 
 # Smart card support
diff --git a/toolkit/mozapps/installer/package-name.mk b/toolkit/mozapps/installer/package-name.mk
index d864f88cc63a..932329ea35c7 100644
--- a/toolkit/mozapps/installer/package-name.mk
+++ b/toolkit/mozapps/installer/package-name.mk
@@ -60,6 +60,7 @@ PKG_PATH =
 SDK_PATH =
 PKG_INST_BASENAME = $(PKG_BASENAME).installer
 PKG_STUB_BASENAME = $(PKG_BASENAME).installer-stub
+PKG_INST_PATH = install/sea/
 PKG_UPDATE_BASENAME = $(PKG_BASENAME)
 CHECKSUMS_FILE_BASENAME = $(PKG_BASENAME)
 MOZ_INFO_BASENAME = $(PKG_BASENAME)
diff --git a/toolkit/mozapps/installer/packager.mk b/toolkit/mozapps/installer/packager.mk
index 60de7c767791..188ce9875ae8 100644
--- a/toolkit/mozapps/installer/packager.mk
+++ b/toolkit/mozapps/installer/packager.mk
@@ -172,13 +172,16 @@ endif
 	ln -s $(installdir)/$(MOZ_APP_NAME) $(DESTDIR)$(bindir)
 
 upload:
-	$(PYTHON3) -u $(MOZILLA_DIR)/build/upload.py --base-path $(ABS_DIST) $(UPLOAD_FILES)
+	$(PYTHON3) -u $(MOZILLA_DIR)/build/upload.py --base-path $(DIST) $(UPLOAD_FILES)
 	mkdir -p `dirname $(CHECKSUM_FILE)`
-	$(PYTHON3) $(MOZILLA_DIR)/build/checksums.py \
+	@$(PYTHON3) $(MOZILLA_DIR)/build/checksums.py \
 		-o $(CHECKSUM_FILE) \
 		$(CHECKSUM_ALGORITHM_PARAM) \
 		$(UPLOAD_PATH)
-	$(PYTHON3) -u $(MOZILLA_DIR)/build/upload.py --base-path $(ABS_DIST) $(CHECKSUM_FILE)
+	@echo 'CHECKSUM FILE START'
+	@cat $(CHECKSUM_FILE)
+	@echo 'CHECKSUM FILE END'
+	$(PYTHON3) -u $(MOZILLA_DIR)/build/upload.py --base-path $(DIST) $(CHECKSUM_FILES)
 
 # source-package creates a source tarball from the files in MOZ_PKG_SRCDIR,
 # which is either set to a clean checkout or defaults to $topsrcdir
diff --git a/toolkit/mozapps/installer/upload-files.mk b/toolkit/mozapps/installer/upload-files.mk
index f301e254e182..3201bc3cda77 100644
--- a/toolkit/mozapps/installer/upload-files.mk
+++ b/toolkit/mozapps/installer/upload-files.mk
@@ -223,7 +223,7 @@ ifneq (android,$(MOZ_WIDGET_TOOLKIT))
 endif
 
 ifeq ($(OS_TARGET), WINNT)
-  INSTALLER_PACKAGE = $(DIST)/$(PKG_PATH)$(PKG_INST_BASENAME).exe
+  INSTALLER_PACKAGE = $(DIST)/$(PKG_INST_PATH)$(PKG_INST_BASENAME).exe
 endif
 
 # These are necessary because some of our packages/installers contain spaces
@@ -240,7 +240,8 @@ ESCAPE_WILDCARD = $(subst $(space),?,$(1))
 CHECKSUM_ALGORITHM_PARAM = -d sha512 -d md5 -d sha1
 
 # This variable defines where the checksum file will be located
-CHECKSUM_FILE = '$(ABS_DIST)/$(PKG_PATH)/$(CHECKSUMS_FILE_BASENAME).checksums'
+CHECKSUM_FILE = '$(DIST)/$(PKG_PATH)/$(CHECKSUMS_FILE_BASENAME).checksums'
+CHECKSUM_FILES = $(CHECKSUM_FILE)
 
 # Upload MAR tools only if AB_CD is unset or en_US
 ifeq (,$(AB_CD:en-US=))
@@ -299,7 +300,7 @@ endif
 endif
 
 ifdef MOZ_STUB_INSTALLER
-  UPLOAD_FILES += $(call QUOTED_WILDCARD,$(DIST)/$(PKG_PATH)$(PKG_STUB_BASENAME).exe)
+  UPLOAD_FILES += $(call QUOTED_WILDCARD,$(DIST)/$(PKG_INST_PATH)$(PKG_STUB_BASENAME).exe)
 endif
 
 # Upload `.xpt` artifacts for use in artifact builds.
diff --git a/toolkit/mozapps/installer/windows/nsis/makensis.mk b/toolkit/mozapps/installer/windows/nsis/makensis.mk
index facc5b1c058b..b4425034fb19 100755
--- a/toolkit/mozapps/installer/windows/nsis/makensis.mk
+++ b/toolkit/mozapps/installer/windows/nsis/makensis.mk
@@ -63,9 +63,10 @@ endif
 
 ifdef ZIP_IN
 installer:: $(CONFIG_DIR)/setup.exe $(ZIP_IN)
-	$(NSINSTALL) -D '$(ABS_DIST)/$(PKG_PATH)'
+	@echo 'Packaging $(WIN32_INSTALLER_OUT).'
+	$(NSINSTALL) -D '$(ABS_DIST)/$(PKG_INST_PATH)'
 	$(PYTHON3) $(MOZILLA_DIR)/mach repackage installer \
-	  -o '$(ABS_DIST)/$(PKG_PATH)$(PKG_INST_BASENAME).exe' \
+	  -o '$(ABS_DIST)/$(PKG_INST_PATH)$(PKG_INST_BASENAME).exe' \
 	  --package-name '$(MOZ_PKG_DIR)' \
 	  --package '$(ZIP_IN)' \
 	  --tag $(topsrcdir)/$(MOZ_BUILD_APP)/installer/windows/app.tag \
@@ -74,7 +75,7 @@ installer:: $(CONFIG_DIR)/setup.exe $(ZIP_IN)
 	  $(USE_UPX)
 ifdef MOZ_STUB_INSTALLER
 	$(PYTHON3) $(MOZILLA_DIR)/mach repackage installer \
-	  -o '$(ABS_DIST)/$(PKG_PATH)$(PKG_STUB_BASENAME).exe' \
+	  -o '$(ABS_DIST)/$(PKG_INST_PATH)$(PKG_STUB_BASENAME).exe' \
 	  --tag $(topsrcdir)/browser/installer/windows/stub.tag \
 	  --setupexe $(CONFIG_DIR)/setup-stub.exe \
 	  --sfx-stub $(SFX_MODULE) \
