diff --git a/net/cert/cert_verify_proc_unittest.cc b/net/cert/cert_verify_proc_unittest.cc
index 73f8fa3018017..ba69033286c0a 100644
--- a/net/cert/cert_verify_proc_unittest.cc
+++ b/net/cert/cert_verify_proc_unittest.cc
@@ -310,6 +310,23 @@ TEST_F(CertVerifyProcTest, RejectExpiredCert) {
   EXPECT_TRUE(verify_result.cert_status & CERT_STATUS_DATE_INVALID);
 }
 
+// Test that verifying an ECDSA certificate doesn't crash on XP. (See
+// crbug.com/144466).
+TEST_F(CertVerifyProcTest, ECDSA_RSA) {
+  base::FilePath certs_dir = GetTestCertsDirectory();
+
+  scoped_refptr<X509Certificate> cert =
+      ImportCertFromFile(certs_dir,
+                         "prime256v1-ecdsa-ee-by-1024-rsa-intermediate.pem");
+
+  CertVerifyResult verify_result;
+  Verify(cert.get(), "127.0.0.1", 0, NULL, empty_cert_list_, &verify_result);
+
+  // We don't check verify_result because the certificate is signed by an
+  // unknown CA and will be considered invalid on XP because of the ECDSA
+  // public key.
+}
+
 // Currently, only RSA and DSA keys are checked for weakness, and our example
 // weak size is 768. These could change in the future.
 //
diff --git a/net/cert/test_root_certs_win.cc b/net/cert/test_root_certs_win.cc
index 8535725cb4b5c..67bf1b09f75ac 100644
--- a/net/cert/test_root_certs_win.cc
+++ b/net/cert/test_root_certs_win.cc
@@ -9,6 +9,7 @@
 #include "base/logging.h"
 #include "base/numerics/safe_conversions.h"
 #include "base/win/win_util.h"
+#include "base/win/windows_version.h"
 #include "net/cert/x509_certificate.h"
 #include "third_party/boringssl/src/include/openssl/pool.h"
 
@@ -172,11 +173,11 @@ HCERTCHAINENGINE TestRootCerts::GetChainEngine() const {
   if (IsEmpty())
     return nullptr;  // Default chain engine will suffice.
 
-  // Windows versions before 8 don't accept the struct size for later versions.
+  // Windows versions before 7 don't accept the struct size for later versions.
   // We report the size of the old struct since we don't need the new members.
   static const DWORD kSizeofCertChainEngineConfig =
-      SIZEOF_STRUCT_WITH_SPECIFIED_LAST_MEMBER(CERT_CHAIN_ENGINE_CONFIG,
-                                               hExclusiveTrustedPeople);
+      SIZEOF_STRUCT_WITH_SPECIFIED_LAST_MEMBER(
+          CERT_CHAIN_ENGINE_CONFIG, CycleDetectionModulus);
 
   // Each HCERTCHAINENGINE caches both the configured system stores and
   // information about each chain that has been built. In order to ensure
diff --git a/net/ssl/ssl_platform_key_win.cc b/net/ssl/ssl_platform_key_win.cc
index 11c9d111a3d0d..d49d72f25f8ed 100644
--- a/net/ssl/ssl_platform_key_win.cc
+++ b/net/ssl/ssl_platform_key_win.cc
@@ -21,6 +21,7 @@
 #include "base/logging.h"
 #include "base/macros.h"
 #include "base/strings/utf_string_conversions.h"
+#include "base/win/windows_version.h"
 #include "crypto/openssl_util.h"
 #include "crypto/scoped_capi_types.h"
 #include "net/base/net_errors.h"
@@ -336,7 +337,9 @@ scoped_refptr<SSLPrivateKey> FetchClientCertPrivateKey(
   HCRYPTPROV_OR_NCRYPT_KEY_HANDLE prov_or_key = 0;
   DWORD key_spec = 0;
   BOOL must_free = FALSE;
-  DWORD flags = CRYPT_ACQUIRE_PREFER_NCRYPT_KEY_FLAG;
+  DWORD flags = 0;
+  if (base::win::GetVersion() >= base::win::VERSION_VISTA)
+    flags |= CRYPT_ACQUIRE_PREFER_NCRYPT_KEY_FLAG;
 
   if (!CryptAcquireCertificatePrivateKey(cert_context, flags, nullptr,
                                          &prov_or_key, &key_spec, &must_free)) {
