diff --git a/third_party/blink/renderer/platform/wtf/threading_primitives.h b/third_party/blink/renderer/platform/wtf/threading_primitives.h
index 847f6c6e79b72..95a0781f84e39 100644
--- a/third_party/blink/renderer/platform/wtf/threading_primitives.h
+++ b/third_party/blink/renderer/platform/wtf/threading_primitives.h
@@ -61,7 +61,17 @@ struct PlatformMutex {
   CRITICAL_SECTION internal_mutex_;
   size_t recursion_count_;
 };
-typedef CONDITION_VARIABLE PlatformCondition;
+struct PlatformCondition {
+  size_t waiters_gone_;
+  size_t waiters_blocked_;
+  size_t waiters_to_unblock_;
+  HANDLE block_lock_;
+  HANDLE block_queue_;
+  HANDLE unblock_lock_;
+
+  bool TimedWait(PlatformMutex&, DWORD duration_milliseconds);
+  void Signal(bool unblock_all);
+};
 #elif defined(OS_POSIX) || defined(OS_FUCHSIA)
 struct PlatformMutex {
   pthread_mutex_t internal_mutex_;
diff --git a/third_party/blink/renderer/platform/wtf/threading_win.cc b/third_party/blink/renderer/platform/wtf/threading_win.cc
--- a/third_party/blink/renderer/platform/wtf/threading_win.cc
+++ b/third_party/blink/renderer/platform/wtf/threading_win.cc
@@ -236,27 +236,152 @@ bool RecursiveMutex::TryLock() {
   return true;
 }
 
-ThreadCondition::ThreadCondition(Mutex& mutex)
-  : condition_(CONDITION_VARIABLE_INIT), mutex_(mutex.Impl()) {}
+bool PlatformCondition::TimedWait(PlatformMutex& mutex,
+                                  DWORD duration_milliseconds) {
+  // Enter the wait state.
+  DWORD res = WaitForSingleObject(block_lock_, INFINITE);
+  DCHECK_EQ(res, WAIT_OBJECT_0);
+  ++waiters_blocked_;
+  res = ReleaseSemaphore(block_lock_, 1, 0);
+  DCHECK(res);
+
+  --mutex.recursion_count_;
+  LeaveCriticalSection(&mutex.internal_mutex_);
+
+  // Main wait - use timeout.
+  bool timed_out = (WaitForSingleObject(block_queue_, duration_milliseconds) ==
+                    WAIT_TIMEOUT);
+
+  res = WaitForSingleObject(unblock_lock_, INFINITE);
+  DCHECK_EQ(res, WAIT_OBJECT_0);
+
+  int signals_left = waiters_to_unblock_;
+
+  if (waiters_to_unblock_) {
+    --waiters_to_unblock_;
+  } else if (++waiters_gone_ == (INT_MAX / 2)) {
+    // timeout/canceled or spurious semaphore timeout or spurious wakeup
+    // occured, normalize the m_waitersGone count this may occur if many
+    // calls to wait with a timeout are made and no call to notify_* is made
+    res = WaitForSingleObject(block_lock_, INFINITE);
+    DCHECK_EQ(res, WAIT_OBJECT_0);
+    waiters_blocked_ -= waiters_gone_;
+    res = ReleaseSemaphore(block_lock_, 1, 0);
+    DCHECK(res);
+    waiters_gone_ = 0;
+  }
+
+  res = ReleaseMutex(unblock_lock_);
+  DCHECK(res);
+
+  if (signals_left == 1) {
+    res = ReleaseSemaphore(block_lock_, 1, 0);  // Open the gate.
+    DCHECK(res);
+  }
+
+  EnterCriticalSection(&mutex.internal_mutex_);
+  ++mutex.recursion_count_;
+
+  return !timed_out;
+}
+
+void PlatformCondition::Signal(bool unblock_all) {
+  unsigned signals_to_issue = 0;
+
+  DWORD res = WaitForSingleObject(unblock_lock_, INFINITE);
+  DCHECK_EQ(res, WAIT_OBJECT_0);
+
+  if (waiters_to_unblock_) {  // the gate is already closed
+    if (!waiters_blocked_) {  // no-op
+      res = ReleaseMutex(unblock_lock_);
+      DCHECK(res);
+      return;
+    }
+
+    if (unblock_all) {
+      signals_to_issue = waiters_blocked_;
+      waiters_to_unblock_ += waiters_blocked_;
+      waiters_blocked_ = 0;
+    } else {
+      signals_to_issue = 1;
+      ++waiters_to_unblock_;
+      --waiters_blocked_;
+    }
+  } else if (waiters_blocked_ > waiters_gone_) {
+    res = WaitForSingleObject(block_lock_, INFINITE);  // Close the gate.
+    DCHECK_EQ(res, WAIT_OBJECT_0);
+    if (waiters_gone_ != 0) {
+      waiters_blocked_ -= waiters_gone_;
+      waiters_gone_ = 0;
+    }
+    if (unblock_all) {
+      signals_to_issue = waiters_blocked_;
+      waiters_to_unblock_ = waiters_blocked_;
+      waiters_blocked_ = 0;
+    } else {
+      signals_to_issue = 1;
+      waiters_to_unblock_ = 1;
+      --waiters_blocked_;
+    }
+  } else {  // No-op.
+    res = ReleaseMutex(unblock_lock_);
+    DCHECK(res);
+    return;
+  }
+
+  res = ReleaseMutex(unblock_lock_);
+  DCHECK(res);
+
+  if (signals_to_issue) {
+    res = ReleaseSemaphore(block_queue_, signals_to_issue, 0);
+    DCHECK(res);
+  }
+}
+
+static const long kMaxSemaphoreCount = static_cast<long>(~0UL >> 1);
+
+ThreadCondition::ThreadCondition(Mutex& mutex) : mutex_(mutex.Impl()) {
+  condition_.waiters_gone_ = 0;
+  condition_.waiters_blocked_ = 0;
+  condition_.waiters_to_unblock_ = 0;
+  condition_.block_lock_ = CreateSemaphore(0, 1, 1, 0);
+  condition_.block_queue_ = CreateSemaphore(0, 0, kMaxSemaphoreCount, 0);
+  condition_.unblock_lock_ = CreateMutex(0, 0, 0);
+
+  if (!condition_.block_lock_ || !condition_.block_queue_ ||
+      !condition_.unblock_lock_) {
+    if (condition_.block_lock_)
+      CloseHandle(condition_.block_lock_);
+    if (condition_.block_queue_)
+      CloseHandle(condition_.block_queue_);
+    if (condition_.unblock_lock_)
+      CloseHandle(condition_.unblock_lock_);
+
+    condition_.block_lock_ = nullptr;
+    condition_.block_queue_ = nullptr;
+    condition_.unblock_lock_ = nullptr;
+  }
+}
 
-ThreadCondition::~ThreadCondition() {}
+ThreadCondition::~ThreadCondition() {
+  if (condition_.block_lock_)
+    CloseHandle(condition_.block_lock_);
+  if (condition_.block_queue_)
+    CloseHandle(condition_.block_queue_);
+  if (condition_.unblock_lock_)
+    CloseHandle(condition_.unblock_lock_);
+}
 
 void ThreadCondition::Wait() {
-  base::ScopedBlockingCall scoped_blocking_call(FROM_HERE,
-                                                base::BlockingType::MAY_BLOCK);
-  --mutex_.recursion_count_;
-  BOOL result =
-      SleepConditionVariableCS(&condition_, &mutex_.internal_mutex_, INFINITE);
-  DCHECK_NE(result, 0);
-  ++mutex_.recursion_count_;
+  condition_.TimedWait(mutex_, INFINITE);
 }
 
 void ThreadCondition::Signal() {
-  WakeConditionVariable(&condition_);
+  condition_.Signal(false);  // Unblock only 1 thread.
 }
 
 void ThreadCondition::Broadcast() {
-  WakeAllConditionVariable(&condition_);
+  condition_.Signal(true);  // Unblock all threads.
 }
 
 }  // namespace WTF
