diff --git a/base/synchronization/condition_variable.h b/base/synchronization/condition_variable.h
index ebf90d2..7dda21f 100644
--- a/base/synchronization/condition_variable.h
+++ b/base/synchronization/condition_variable.h
@@ -104,7 +104,7 @@ class BASE_EXPORT ConditionVariable {
 
 #if defined(OS_WIN)
   CHROME_CONDITION_VARIABLE cv_;
-  CHROME_SRWLOCK* const srwlock_;
+  char const crit_sec_[100];
 #elif defined(OS_POSIX) || defined(OS_FUCHSIA)
   pthread_cond_t condition_;
   pthread_mutex_t* user_mutex_;
diff --git a/base/synchronization/condition_variable_win.cc b/base/synchronization/condition_variable_win.cc
index 61c6a71..d598e33 100644
--- a/base/synchronization/condition_variable_win.cc
+++ b/base/synchronization/condition_variable_win.cc
@@ -11,7 +11,7 @@
 namespace base {
 
 ConditionVariable::ConditionVariable(Lock* user_lock)
-    : srwlock_(user_lock->lock_.native_handle())
+    : crit_sec_(user_lock->lock_.native_handle())
 #if DCHECK_IS_ON()
     , user_lock_(user_lock)
 #endif
@@ -34,7 +34,7 @@ void ConditionVariable::TimedWait(const TimeDelta& max_time) {
 #endif
 
   if (!SleepConditionVariableSRW(reinterpret_cast<PCONDITION_VARIABLE>(&cv_),
-                                 reinterpret_cast<PSRWLOCK>(srwlock_), timeout,
+                                 reinterpret_cast<PCRITICAL_SECTION>(crit_sec_), timeout,
                                  0)) {
     // On failure, we only expect the CV to timeout. Any other error value means
     // that we've unexpectedly woken up.
diff --git a/base/synchronization/lock.h b/base/synchronization/lock.h
index fbf6cef..39a26b7 100644
--- a/base/synchronization/lock.h
+++ b/base/synchronization/lock.h
@@ -38,9 +38,9 @@ class BASE_EXPORT Lock {
   Lock();
   ~Lock();
 
-  // NOTE: We do not permit recursive locks and will commonly fire a DCHECK() if
-  // a thread attempts to acquire the lock a second time (while already holding
-  // it).
+  // NOTE: Although windows critical sections support recursive locks, we do not
+  // allow this, and we will commonly fire a DCHECK() if a thread attempts to
+  // acquire the lock a second time (while already holding it).
   void Acquire() {
     lock_.Lock();
     CheckUnheldAndMark();
diff --git a/base/synchronization/lock_impl.h b/base/synchronization/lock_impl.h
index cbaabc7..ed85987 100644
--- a/base/synchronization/lock_impl.h
+++ b/base/synchronization/lock_impl.h
@@ -24,9 +24,9 @@ namespace internal {
 class BASE_EXPORT LockImpl {
  public:
 #if defined(OS_WIN)
-  using NativeHandle = CHROME_SRWLOCK;
+  typedef char NativeHandle[100];
 #elif defined(OS_POSIX) || defined(OS_FUCHSIA)
-  using NativeHandle =  pthread_mutex_t;
+  typedef pthread_mutex_t NativeHandle;
 #endif
 
   LockImpl();
@@ -59,7 +61,7 @@ class BASE_EXPORT LockImpl {
 
 #if defined(OS_WIN)
 void LockImpl::Unlock() {
-  ::ReleaseSRWLockExclusive(reinterpret_cast<PSRWLOCK>(&native_handle_));
+  ::LeaveCriticalSection(reinterpret_cast<PCRITICAL_SECTION>(&native_handle_));
 }
 #elif defined(OS_POSIX) || defined(OS_FUCHSIA)
 void LockImpl::Unlock() {
diff --git a/base/synchronization/lock_impl_win.cc b/base/synchronization/lock_impl_win.cc
index d89a131..fbc1bdd 100644
--- a/base/synchronization/lock_impl_win.cc
+++ b/base/synchronization/lock_impl_win.cc
@@ -7,29 +7,37 @@
 namespace base {
 namespace internal {
 
-LockImpl::LockImpl() : native_handle_(SRWLOCK_INIT) {}
+LockImpl::LockImpl() {
+  // The second parameter is the spin count, for short-held locks it avoid the
+  // contending thread from going to sleep which helps performance greatly.
+  ::InitializeCriticalSectionAndSpinCount(reinterpret_cast<PCRITICAL_SECTION>(&native_handle_), 2000);
+}
 
-LockImpl::~LockImpl() = default;
+LockImpl::~LockImpl() {
+  ::DeleteCriticalSection(reinterpret_cast<PCRITICAL_SECTION>(&native_handle_));
+}
 
 bool LockImpl::Try() {
-  return !!::TryAcquireSRWLockExclusive(
-      reinterpret_cast<PSRWLOCK>(&native_handle_));
+  if (::TryEnterCriticalSection(reinterpret_cast<PCRITICAL_SECTION>(&native_handle_)) != FALSE) {
+    return true;
+  }
+  return false;
 }
 
 void LockImpl::Lock() {
   // The ScopedLockAcquireActivity below is relatively expensive and so its
   // actions can become significant due to the very large number of locks
   // that tend to be used throughout the build. To avoid this cost in the
   // vast majority of the calls, simply "try" the lock first and only do the
   // (tracked) blocking call if that fails. Since "try" itself is a system
   // call, and thus also somewhat expensive, don't bother with it unless
   // tracking is actually enabled.
   if (base::debug::GlobalActivityTracker::IsEnabled())
     if (Try())
       return;
 
   base::debug::ScopedLockAcquireActivity lock_activity(this);
-  ::AcquireSRWLockExclusive(reinterpret_cast<PSRWLOCK>(&native_handle_));
+  ::EnterCriticalSection(reinterpret_cast<PCRITICAL_SECTION>(&native_handle_));
 }
 
 }  // namespace internal
diff --git a/base/win/windows_types.h b/base/win/windows_types.h
index 091e47fff25a6..8060f03911171 100644
--- a/base/win/windows_types.h
+++ b/base/win/windows_types.h
@@ -102,6 +103,24 @@ typedef struct _FILETIME FILETIME;
 WINBASEAPI _Releases_exclusive_lock_(*SRWLock) VOID WINAPI
     ReleaseSRWLockExclusive(_Inout_ PSRWLOCK SRWLock);
 
+#ifndef IN
+#define IN
+#endif
+
+#ifndef OUT
+#define OUT
+#endif
+
+typedef struct _RTL_CRITICAL_SECTION *PCRITICAL_SECTION;
+typedef struct _RTL_CRITICAL_SECTION *LPCRITICAL_SECTION;
+
+WINBASEAPI
+VOID
+WINAPI
+LeaveCriticalSection(
+    IN OUT LPCRITICAL_SECTION lpCriticalSection
+    );
+
 // Needed to support protobuf's GetMessage macro magic.
 WINUSERAPI BOOL WINAPI GetMessageW(_Out_ LPMSG lpMsg,
                                    _In_opt_ HWND hWnd,
