diff --git a/build/config/c++/c++.gni b/build/config/c++/c++.gni
index 040c9358f1a1a..61f07bc0804fc 100644
--- a/build/config/c++/c++.gni
+++ b/build/config/c++/c++.gni
@@ -11,7 +11,7 @@ declare_args() {
   # Don't check in changes that set this to false for more platforms; doing so
   # is not supported.
   use_custom_libcxx =
-      is_fuchsia || is_android || is_mac || (is_win && is_clang) ||
+      is_fuchsia || is_android || is_mac ||
       (is_linux &&
        (!is_chromeos || default_toolchain != "//build/toolchain/cros:target"))
 
