diff --git a/third_party/libvpx/source/libvpx/vpx_util/vpx_thread.h b/third_party/libvpx/source/libvpx/vpx_util/vpx_thread.h
index cf55a75e7..b63a01eba 100644
--- a/third_party/libvpx/source/libvpx/vpx_util/vpx_thread.h
+++ b/third_party/libvpx/source/libvpx/vpx_util/vpx_thread.h
@@ -77,11 +77,7 @@ static INLINE int pthread_join(pthread_t thread, void** value_ptr) {
 static INLINE int pthread_mutex_init(pthread_mutex_t *const mutex,
                                      void *mutexattr) {
   (void)mutexattr;
-#if _WIN32_WINNT >= 0x0600  // Windows Vista / Server 2008 or greater
-  InitializeCriticalSectionEx(mutex, 0 /*dwSpinCount*/, 0 /*Flags*/);
-#else
   InitializeCriticalSection(mutex);
-#endif
   return 0;
 }
 
