diff --git a/ui/base/ime/win/tsf_input_scope.cc b/ui/base/ime/win/tsf_input_scope.cc
index d8025efae6e47..e2932190581a3 100644
--- a/ui/base/ime/win/tsf_input_scope.cc
+++ b/ui/base/ime/win/tsf_input_scope.cc
@@ -165,6 +165,10 @@ void InitializeTsfForInputScopes() {
   if (!g_get_proc_done) {
     g_get_proc_done = true;
 
+    // For stability reasons, we do not support Windows XP.
+    if (base::win::GetVersion() < base::win::VERSION_VISTA)
+      return;
+
     HMODULE module = NULL;
     if (GetModuleHandleExW(GET_MODULE_HANDLE_EX_FLAG_PIN, L"msctf.dll",
         &module)) {
diff --git a/ui/base/l10n/l10n_util_win.cc b/ui/base/l10n/l10n_util_win.cc
index 8acfde020eea0..01bbacc66e24f 100644
--- a/ui/base/l10n/l10n_util_win.cc
+++ b/ui/base/l10n/l10n_util_win.cc
@@ -15,6 +15,7 @@
 #include "base/strings/string_util.h"
 #include "base/strings/utf_string_conversions.h"
 #include "base/win/i18n.h"
+#include "base/win/windows_version.h"
 #include "ui/base/l10n/l10n_util.h"
 #include "ui/display/display.h"
 #include "ui/display/win/screen_win.h"
@@ -46,6 +47,23 @@ void AdjustLogFont(const base::string16& font_family,
 
 namespace {
 
+bool IsFontPresent(const wchar_t* font_name) {
+  HFONT hfont = CreateFont(12, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                           font_name);
+  if (hfont == NULL)
+    return false;
+  HDC dc = GetDC(0);
+  HGDIOBJ oldFont = static_cast<HFONT>(SelectObject(dc, hfont));
+  WCHAR actual_font_name[LF_FACESIZE];
+  int size_ret = GetTextFace(dc, LF_FACESIZE, actual_font_name);
+  SelectObject(dc, oldFont);
+  DeleteObject(hfont);
+  ReleaseDC(0, dc);
+  // We don't have to worry about East Asian fonts with locale-dependent
+  // names here.
+  return (size_ret != 0) && wcscmp(font_name, actual_font_name) == 0;
+}
+
 class OverrideLocaleHolder {
  public:
   OverrideLocaleHolder() {}
@@ -89,7 +107,13 @@ void HWNDSetRTLLayout(HWND hwnd) {
 }
 
 bool IsLocaleSupportedByOS(const std::string& locale) {
-  return true;
+  // Block Amharic on Windows XP unless 'Abyssinica SIL' font is present.
+  // On Win XP, no Ethiopic/Amahric font is availabel out of box. We hard-coded
+  // 'Abyssinica SIL' in the resource bundle to use in the UI. Check
+  // for its presence to determine whether or not to support Amharic UI on XP.
+  return (base::win::GetVersion() >= base::win::VERSION_VISTA ||
+      !base::LowerCaseEqualsASCII(locale, "am") ||
+      IsFontPresent(L"Abyssinica SIL"));
 }
 
 bool NeedOverrideDefaultUIFont(base::string16* override_font_family,
@@ -99,9 +123,19 @@ bool NeedOverrideDefaultUIFont(base::string16* override_font_family,
   // the default Windows fonts are too small to be legible.  For those
   // locales, IDS_UI_FONT_FAMILY is set to an actual font family to
   // use while for other locales, it's set to 'default'.
-  base::string16 ui_font_family = GetStringUTF16(IDS_UI_FONT_FAMILY);
+
+  // XP and Vista or later have different font size issues and
+  // we need separate ui font specifications.
+  int ui_font_family_id = IDS_UI_FONT_FAMILY;
+  int ui_font_size_scaler_id = IDS_UI_FONT_SIZE_SCALER;
+  if (base::win::GetVersion() < base::win::VERSION_VISTA) {
+    ui_font_family_id = IDS_UI_FONT_FAMILY_XP;
+    ui_font_size_scaler_id = IDS_UI_FONT_SIZE_SCALER_XP;
+  }
+
+  base::string16 ui_font_family = GetStringUTF16(ui_font_family_id);
   int scaler100;
-  if (!base::StringToInt(l10n_util::GetStringUTF16(IDS_UI_FONT_SIZE_SCALER),
+  if (!base::StringToInt(l10n_util::GetStringUTF16(ui_font_size_scaler_id),
                          &scaler100))
     return false;
 
diff --git a/ui/base/webui/web_ui_util.cc b/ui/base/webui/web_ui_util.cc
index e6a9c5c23df8b..8b04a37c6d2bb 100644
--- a/ui/base/webui/web_ui_util.cc
+++ b/ui/base/webui/web_ui_util.cc
@@ -160,7 +160,12 @@ void AppendWebUiCssTextDefaults(std::string* html) {
 }
 
 std::string GetFontFamily() {
-  std::string font_family = l10n_util::GetStringUTF8(IDS_WEB_FONT_FAMILY);
+  std::string font_family = l10n_util::GetStringUTF8(
+#if defined(OS_WIN)
+      base::win::GetVersion() < base::win::VERSION_VISTA ?
+          IDS_WEB_FONT_FAMILY_XP :
+#endif
+          IDS_WEB_FONT_FAMILY);
 
 // TODO(dnicoara) Remove Ozone check when PlatformFont support is introduced
 // into Ozone: crbug.com/320050
@@ -173,7 +178,12 @@ std::string GetFontFamily() {
 }
 
 std::string GetFontSize() {
-  return l10n_util::GetStringUTF8(IDS_WEB_FONT_SIZE);
+  return l10n_util::GetStringUTF8(
+#if defined(OS_WIN)
+      base::win::GetVersion() < base::win::VERSION_VISTA ?
+          IDS_WEB_FONT_SIZE_XP :
+#endif
+          IDS_WEB_FONT_SIZE);
 }
 
 std::string GetTextDirection() {
diff --git a/ui/gfx/animation/animation_win.cc b/ui/gfx/animation/animation_win.cc
index ead2e07a3a70d..b8a7d2880a7b5 100644
--- a/ui/gfx/animation/animation_win.cc
+++ b/ui/gfx/animation/animation_win.cc
@@ -4,14 +4,18 @@
 
 #include "base/message_loop/message_loop.h"
 
+#include "base/win/windows_version.h"
+
 namespace gfx {
 
 // static
 bool Animation::ShouldRenderRichAnimationImpl() {
-  BOOL result;
-  // Get "Turn off all unnecessary animations" value.
-  if (::SystemParametersInfo(SPI_GETCLIENTAREAANIMATION, 0, &result, 0)) {
-    return !!result;
+  if (base::win::GetVersion() >= base::win::VERSION_VISTA) {
+    BOOL result;
+    // Get "Turn off all unnecessary animations" value.
+    if (::SystemParametersInfo(SPI_GETCLIENTAREAANIMATION, 0, &result, 0)) {
+      return !!result;
+    }
   }
   return !::GetSystemMetrics(SM_REMOTESESSION);
 }
diff --git a/ui/native_theme/native_theme_win.cc b/ui/native_theme/native_theme_win.cc
index 3340f3744cf6d..70c83cd9c5759 100644
--- a/ui/native_theme/native_theme_win.cc
+++ b/ui/native_theme/native_theme_win.cc
@@ -16,6 +16,7 @@
 #include "base/win/scoped_hdc.h"
 #include "base/win/scoped_select_object.h"
 #include "base/win/win_util.h"
+#include "base/win/windows_version.h"
 #include "cc/paint/paint_canvas.h"
 #include "cc/paint/paint_flags.h"
 #include "skia/ext/platform_canvas.h"
@@ -107,6 +108,12 @@ int ComputeAnimationProgress(int frame_width,
   return static_cast<int>(animation_width * ratio) - object_width;
 }
 
+RECT InsetRect(const RECT* rect, int size) {
+  gfx::Rect result(*rect);
+  result.Inset(size, size);
+  return result.ToRECT();
+}
+
 // Custom scoped object for storing DC and a bitmap that was selected into it,
 // and making sure that they are deleted in the right order.
 class ScopedCreateDCWithBitmap {
@@ -1358,7 +1365,10 @@ HRESULT NativeThemeWin::PaintProgressBar(
   const int kDeterminateOverlayPixelsPerSecond = 300;
   const int kDeterminateOverlayWidth = 120;
   const int kIndeterminateOverlayPixelsPerSecond =  175;
-  const int kIndeterminateOverlayWidth = 120;
+  const int kVistaIndeterminateOverlayWidth = 120;
+  const int kXPIndeterminateOverlayWidth = 55;
+  // The thickness of the bar frame inside |value_rect|
+  const int kXPBarPadding = 3;
 
   RECT bar_rect = rect.ToRECT();
   RECT value_rect = gfx::Rect(extra.value_rect_x,
@@ -1376,19 +1386,27 @@ HRESULT NativeThemeWin::PaintProgressBar(
 
   draw_theme_(handle, hdc, PP_BAR, 0, &bar_rect, NULL);
 
+  bool pre_vista = base::win::GetVersion() < base::win::VERSION_VISTA;
   int bar_width = bar_rect.right - bar_rect.left;
   if (!extra.determinate) {
     // The glossy overlay for the indeterminate progress bar has a small pause
     // after each animation. We emulate this by adding an invisible margin the
     // animation has to traverse.
     int width_with_margin = bar_width + kIndeterminateOverlayPixelsPerSecond;
-    int overlay_width = kIndeterminateOverlayWidth;
+    int overlay_width = pre_vista ?
+        kXPIndeterminateOverlayWidth : kVistaIndeterminateOverlayWidth;
     RECT overlay_rect = bar_rect;
     overlay_rect.left += ComputeAnimationProgress(
         width_with_margin, overlay_width, kIndeterminateOverlayPixelsPerSecond,
         extra.animated_seconds);
     overlay_rect.right = overlay_rect.left + overlay_width;
-    draw_theme_(handle, hdc, PP_MOVEOVERLAY, 0, &overlay_rect, &bar_rect);
+    if (pre_vista) {
+      RECT shrunk_rect = InsetRect(&overlay_rect, kXPBarPadding);
+      RECT shrunk_bar_rect = InsetRect(&bar_rect, kXPBarPadding);
+      draw_theme_(handle, hdc, PP_CHUNK, 0, &shrunk_rect, &shrunk_bar_rect);
+    } else {
+      draw_theme_(handle, hdc, PP_MOVEOVERLAY, 0, &overlay_rect, &bar_rect);
+    }
     return S_OK;
   }
 
@@ -1401,18 +1419,26 @@ HRESULT NativeThemeWin::PaintProgressBar(
         DTBG_MIRRORDC : 0u,
     bar_rect
   };
-
-  // On Vista or later, the progress bar part has a single-block value part
-  // and a glossy effect. The value part has exactly same height as the bar
-  // part, so we don't need to shrink the rect.
-  draw_theme_ex_(handle, hdc, PP_FILL, 0, &value_rect, &value_draw_options);
-
-  RECT overlay_rect = value_rect;
-  overlay_rect.left += ComputeAnimationProgress(
-      bar_width, kDeterminateOverlayWidth, kDeterminateOverlayPixelsPerSecond,
-      extra.animated_seconds);
-  overlay_rect.right = overlay_rect.left + kDeterminateOverlayWidth;
-  draw_theme_(handle, hdc, PP_MOVEOVERLAY, 0, &overlay_rect, &value_rect);
+  if (pre_vista) {
+    // On XP, the progress bar is chunk-style and has no glossy effect.  We need
+    // to shrink the destination rect to fit the part inside the bar with an
+    // appropriate margin.
+    RECT shrunk_value_rect = InsetRect(&value_rect, kXPBarPadding);
+    draw_theme_ex_(handle, hdc, PP_CHUNK, 0, &shrunk_value_rect,
+                   &value_draw_options);
+  } else  {
+    // On Vista or later, the progress bar part has a single-block value part
+    // and a glossy effect.  The value part has exactly same height as the bar
+    // part, so we don't need to shrink the rect.
+    draw_theme_ex_(handle, hdc, PP_FILL, 0, &value_rect, &value_draw_options);
+
+    RECT overlay_rect = value_rect;
+    overlay_rect.left += ComputeAnimationProgress(
+        bar_width, kDeterminateOverlayWidth, kDeterminateOverlayPixelsPerSecond,
+        extra.animated_seconds);
+    overlay_rect.right = overlay_rect.left + kDeterminateOverlayWidth;
+    draw_theme_(handle, hdc, PP_MOVEOVERLAY, 0, &overlay_rect, &value_rect);
+  }
   return S_OK;
 }
 
@@ -1738,7 +1764,9 @@ int NativeThemeWin::GetWindowsState(Part part,
         case kDisabled:
           return ABS_DOWNDISABLED;
         case kHovered:
-          return ABS_DOWNHOVER;
+          // Mimic ScrollbarThemeChromiumWin.cpp in WebKit.
+          return base::win::GetVersion() < base::win::VERSION_VISTA ?
+              ABS_DOWNHOT : ABS_DOWNHOVER;
         case kNormal:
           return ABS_DOWNNORMAL;
         case kPressed:
@@ -1752,7 +1780,9 @@ int NativeThemeWin::GetWindowsState(Part part,
         case kDisabled:
           return ABS_LEFTDISABLED;
         case kHovered:
-          return ABS_LEFTHOVER;
+          // Mimic ScrollbarThemeChromiumWin.cpp in WebKit.
+          return base::win::GetVersion() < base::win::VERSION_VISTA ?
+              ABS_LEFTHOT : ABS_LEFTHOVER;
         case kNormal:
           return ABS_LEFTNORMAL;
         case kPressed:
@@ -1766,7 +1796,9 @@ int NativeThemeWin::GetWindowsState(Part part,
         case kDisabled:
           return ABS_RIGHTDISABLED;
         case kHovered:
-          return ABS_RIGHTHOVER;
+          // Mimic ScrollbarThemeChromiumWin.cpp in WebKit.
+          return base::win::GetVersion() < base::win::VERSION_VISTA ?
+              ABS_RIGHTHOT : ABS_RIGHTHOVER;
         case kNormal:
           return ABS_RIGHTNORMAL;
         case kPressed:
@@ -1781,7 +1813,9 @@ int NativeThemeWin::GetWindowsState(Part part,
         case kDisabled:
           return ABS_UPDISABLED;
         case kHovered:
-          return ABS_UPHOVER;
+          // Mimic ScrollbarThemeChromiumWin.cpp in WebKit.
+          return base::win::GetVersion() < base::win::VERSION_VISTA ?
+              ABS_UPHOT : ABS_UPHOVER;
         case kNormal:
           return ABS_UPNORMAL;
         case kPressed:
@@ -1797,7 +1831,9 @@ int NativeThemeWin::GetWindowsState(Part part,
         case kDisabled:
           return SCRBS_DISABLED;
         case kHovered:
-          return SCRBS_HOVER;
+          // Mimic WebKit's behaviour in ScrollbarThemeChromiumWin.cpp.
+          return base::win::GetVersion() < base::win::VERSION_VISTA ?
+              SCRBS_HOT : SCRBS_HOVER;
         case kNormal:
           return SCRBS_NORMAL;
         case kPressed:
diff --git a/ui/strings/app_locale_settings.grd b/ui/strings/app_locale_settings.grd
index 3417591ad8888..a6de88ab52320 100644
--- a/ui/strings/app_locale_settings.grd
+++ b/ui/strings/app_locale_settings.grd
@@ -148,6 +148,10 @@
         <message name="IDS_UI_FONT_FAMILY" use_name_for_id="true">
           default
         </message>
+        <!-- For Windows XP -->
+        <message name="IDS_UI_FONT_FAMILY_XP" use_name_for_id="true">
+          default
+        </message>
 
         <!-- To get the actual UI font size for native UI components
              (e.g. menu), the system UI font size is scaled with this value/100.
@@ -155,15 +159,29 @@
         <message name="IDS_UI_FONT_SIZE_SCALER" use_name_for_id="true">
           100
         </message>
+        <!-- For Windows XP -->
+        <message name="IDS_UI_FONT_SIZE_SCALER_XP" use_name_for_id="true">
+          100
+        </message>
 
         <!-- The font used in Web UI (e.g. History). -->
         <message name="IDS_WEB_FONT_FAMILY" use_name_for_id="true">
           'Segoe UI', Tahoma, sans-serif
         </message>
+        <!-- The font used in Web UI on Windows XP (e.g. History). -->
+        <message name="IDS_WEB_FONT_FAMILY_XP" use_name_for_id="true">
+          Tahoma, sans-serif
+        </message>
+
         <!-- The relative font size in % used in Web UI (e.g. History).  -->
         <message name="IDS_WEB_FONT_SIZE" use_name_for_id="true">
           75%
         </message>
+        <!-- The relative font size in % used in web-style native pages
+        on Windows earlier than Vista (e.g. History).  -->
+        <message name="IDS_WEB_FONT_SIZE_XP" use_name_for_id="true">
+          75%
+        </message>
       </if>
       <if expr="is_macosx">
         <!-- The font used in Web UI (e.g. History). -->
