diff --git a/base/BUILD.gn b/base/BUILD.gn
index b2c4562aa4536..acec47a0be298 100644
--- a/base/BUILD.gn
+++ b/base/BUILD.gn
@@ -1232,6 +1232,10 @@ component("base") {
       "/DELAYLOAD:cfgmgr32.dll",
       "/DELAYLOAD:powrprof.dll",
       "/DELAYLOAD:setupapi.dll",
+      "/DELAYLOAD:dbghelp.dll",
+      "/DELAYLOAD:dwmapi.dll",
+      "/DELAYLOAD:shell32.dll",
+      "/DELAYLOAD:uxtheme.dll",
     ]
   }
 }
diff --git a/chrome/BUILD.gn b/chrome/BUILD.gn
index 5c1f067f30ba5..ef1dd234ff9c5 100644
--- a/chrome/BUILD.gn
+++ b/chrome/BUILD.gn
@@ -236,12 +236,25 @@ if (!is_android && !is_mac) {
         }
 
         ldflags = [
+          "/DELAYLOAD:advapi32.dll",
+          "/DELAYLOAD:comdlg32.dll",
           "/DELAYLOAD:dbghelp.dll",
           "/DELAYLOAD:dwmapi.dll",
-          "/DELAYLOAD:uxtheme.dll",
+          "/DELAYLOAD:imm32.dll",
+          "/DELAYLOAD:iphlpapi.dll",
           "/DELAYLOAD:ole32.dll",
           "/DELAYLOAD:oleaut32.dll",
           "/DELAYLOAD:winhttp.dll",
+          "/DELAYLOAD:propsys.dll",
+          "/DELAYLOAD:psapi.dll",
+          "/DELAYLOAD:shell32.dll",
+          "/DELAYLOAD:shlwapi.dll",
+          "/DELAYLOAD:user32.dll",
+          "/DELAYLOAD:uxtheme.dll",
+          "/DELAYLOAD:wer.dll",
+          "/DELAYLOAD:wevtapi.dll",
+          "/DELAYLOAD:wininet.dll",
+          "/DELAYLOAD:wintrust.dll",
         ]
 
       if (current_cpu == "x64") {
@@ -421,22 +435,52 @@ if (is_win) {
     ]
 
     ldflags = [
+      "/DELAYLOAD:advapi32.dll",
+      "/DELAYLOAD:comctl32.dll",
       "/DELAYLOAD:comdlg32.dll",
+      "/DELAYLOAD:credui.dll",
       "/DELAYLOAD:crypt32.dll",
       "/DELAYLOAD:cryptui.dll",
       "/DELAYLOAD:dbghelp.dll",
+      "/DELAYLOAD:d3d11.dll",
+      "/DELAYLOAD:d3d9.dll",
       "/DELAYLOAD:dhcpcsvc.dll",
       "/DELAYLOAD:dwmapi.dll",
+      "/DELAYLOAD:dwrite.dll",
+      "/DELAYLOAD:dxgi.dll",
+      "/DELAYLOAD:esent.dll",
+      "/DELAYLOAD:gdi32.dll",
+      "/DELAYLOAD:hid.dll",
       "/DELAYLOAD:imagehlp.dll",
       "/DELAYLOAD:imm32.dll",
       "/DELAYLOAD:iphlpapi.dll",
+      "/DELAYLOAD:netapi32.dll",
+      "/DELAYLOAD:ole32.dll",
+      "/DELAYLOAD:oleacc.dll",
+      "/DELAYLOAD:oleaut32.dll",
+      "/DELAYLOAD:ncrypt.dll",
+      "/DELAYLOAD:propsys.dll",
+      "/DELAYLOAD:psapi.dll",
+      "/DELAYLOAD:secur32.dll",
       "/DELAYLOAD:setupapi.dll",
+      "/DELAYLOAD:shell32.dll",
+      "/DELAYLOAD:shlwapi.dll",
+      "/DELAYLOAD:uiautomationcore.dll",
       "/DELAYLOAD:urlmon.dll",
+      "/DELAYLOAD:user32.dll",
+      "/DELAYLOAD:userenv.dll",
+      "/DELAYLOAD:usp10.dll",
+      "/DELAYLOAD:wer.dll",
+      "/DELAYLOAD:wevtapi.dll",
       "/DELAYLOAD:winhttp.dll",
       "/DELAYLOAD:wininet.dll",
       "/DELAYLOAD:winspool.drv",
+      "/DELAYLOAD:wintrust.dll",
+      "/DELAYLOAD:winusb.dll",
       "/DELAYLOAD:ws2_32.dll",
       "/DELAYLOAD:wsock32.dll",
+      "/DELAYLOAD:wtsapi32.dll",
+      "/DELAYLOAD:api-ms-win-core-winrt-string-l1-1-0.dll",
     ]
 
       if (!is_component_build) {
@@ -520,11 +565,25 @@ if (is_win) {
       ]
 
       ldflags = [
+        "/DELAYLOAD:comctl32.dll",
+        "/DELAYLOAD:comdlg32.dll",
         "/DELAYLOAD:d3d11.dll",
         "/DELAYLOAD:d3d9.dll",
         "/DELAYLOAD:dwmapi.dll",
+        "/DELAYLOAD:dxgi.dll",
         "/DELAYLOAD:dxva2.dll",
         "/DELAYLOAD:esent.dll",
+        "/DELAYLOAD:gdi32.dll",
+        "/DELAYLOAD:imm32.dll",
+        "/DELAYLOAD:ole32.dll",
+        "/DELAYLOAD:oleacc.dll",
+        "/DELAYLOAD:propsys.dll",
+        "/DELAYLOAD:rstrtmgr.dll",
+        "/DELAYLOAD:shell32.dll",
+        "/DELAYLOAD:shlwapi.dll",
+        "/DELAYLOAD:urlmon.dll",
+        "/DELAYLOAD:user32.dll",
+        "/DELAYLOAD:usp10.dll",
         "/DELAYLOAD:wininet.dll",
       ]
 
