diff --git a/gfx/wr/webrender/src/composite.rs b/gfx/wr/webrender/src/composite.rs
index f4a613915e92..0075a354e511 100644
--- a/gfx/wr/webrender/src/composite.rs
+++ b/gfx/wr/webrender/src/composite.rs
@@ -303,12 +303,6 @@ pub enum CompositorConfig {
         /// Required if webrender must query the backbuffer's age.
         partial_present: Option<Box<dyn PartialPresentCompositor>>,
     },
-    Layer {
-        /// If supplied, composite the frame using the new experimental compositing
-        /// interface. If this is set, it overrides `compositor_config`. These will
-        /// be unified as the interface stabilises.
-        compositor: Box<dyn LayerCompositor>,
-    },
     /// Use a native OS compositor to draw tiles. This requires clients to implement
     /// the Compositor trait, but can be significantly more power efficient on operating
     /// systems that support it.
@@ -324,7 +318,7 @@ impl CompositorConfig {
             CompositorConfig::Native { ref mut compositor, .. } => {
                 Some(compositor)
             }
-            CompositorConfig::Draw { .. } | CompositorConfig::Layer { .. } => {
+            CompositorConfig::Draw { .. } => {
                 None
             }
         }
@@ -338,25 +332,9 @@ impl CompositorConfig {
             CompositorConfig::Draw { ref mut partial_present, .. } => {
                 partial_present.as_mut()
             }
-            CompositorConfig::Layer { .. } => {
-                None
-            }
         }
     }
 
-    pub fn layer_compositor(&mut self) -> Option<&mut Box<dyn LayerCompositor>> {
-        match self {
-            CompositorConfig::Native { .. } => {
-                None
-            }
-            CompositorConfig::Draw { .. } => {
-                None
-            }
-            CompositorConfig::Layer { ref mut compositor } => {
-                Some(compositor)
-            }
-        }
-    }
 }
 
 impl Default for CompositorConfig {
@@ -383,9 +361,6 @@ pub enum CompositorKind {
         max_partial_present_rects: usize,
         /// Draw previous regions when doing partial present.
         draw_previous_partial_present_regions: bool,
-    },
-    Layer {
-
     },
     /// Native OS compositor.
     Native {
@@ -407,7 +382,7 @@ impl Default for CompositorKind {
 impl CompositorKind {
     pub fn get_virtual_surface_size(&self) -> i32 {
         match self {
-            CompositorKind::Draw { .. } | CompositorKind::Layer {  .. }=> 0,
+            CompositorKind::Draw { .. } => 0,
             CompositorKind::Native { capabilities, .. } => capabilities.virtual_surface_size,
         }
     }
@@ -418,7 +393,6 @@ impl CompositorKind {
                 // When partial present is enabled, we need to force redraw.
                 *max_partial_present_rects > 0
             }
-            CompositorKind::Layer {  } => false,    // TODO(gwc): Is this correct?
             CompositorKind::Native { capabilities, .. } => capabilities.redraw_on_invalidation,
         }
     }
@@ -1349,9 +1323,11 @@ pub struct CompositorInputConfig<'a> {
     pub framebuffer_size: DeviceIntSize,
 }
 
-// Trait for implementors of swapchain based compositing.
+// Skeleton trait for implementors of swapchain based compositing. For now
+// the implementation simply binds a framebuffer-sized surface that can
+// be presented by the native compositor.
 // TODO(gw): Extend to handle external surfaces, layers, swgl, etc.
-pub trait LayerCompositor {
+pub trait Compositor2 {
     // Prepare to composite a frame. Ensure that layers are constructed
     // to match the input config
     fn begin_frame(
diff --git a/gfx/wr/webrender/src/lib.rs b/gfx/wr/webrender/src/lib.rs
index 394814c65368..c970ccf299e3 100644
--- a/gfx/wr/webrender/src/lib.rs
+++ b/gfx/wr/webrender/src/lib.rs
@@ -166,7 +166,7 @@ pub extern crate api;
 extern crate webrender_build;
 
 #[doc(hidden)]
-pub use crate::composite::{LayerCompositor, CompositorInputConfig, CompositorSurfaceUsage};
+pub use crate::composite::{Compositor2, CompositorInputConfig, CompositorSurfaceUsage};
 pub use crate::composite::{CompositorConfig, Compositor, CompositorCapabilities, CompositorSurfaceTransform};
 pub use crate::composite::{NativeSurfaceId, NativeTileId, NativeSurfaceInfo, PartialPresentCompositor};
 pub use crate::composite::{MappableCompositor, MappedTileInfo, SWGLCompositeSurfaceInfo, WindowVisibility};
diff --git a/gfx/wr/webrender/src/picture.rs b/gfx/wr/webrender/src/picture.rs
index 7e06bc95e119..5434b746b59c 100644
--- a/gfx/wr/webrender/src/picture.rs
+++ b/gfx/wr/webrender/src/picture.rs
@@ -1269,10 +1269,6 @@ impl Tile {
             CompositorKind::Native { capabilities, .. } => {
                 (capabilities.max_update_rects > 0, false)
             }
-            CompositorKind::Layer { .. } => {
-                // TODO(gwc): Support partial updates here
-                (false, true)
-            }
         };
 
         // TODO(gw): Consider using smaller tiles and/or tile splits for
@@ -1346,7 +1342,7 @@ impl Tile {
                     // descriptor depending on the compositing mode that will read
                     // the output.
                     let descriptor = match state.composite_state.compositor_kind {
-                        CompositorKind::Draw { .. } | CompositorKind::Layer { .. } => {
+                        CompositorKind::Draw { .. } => {
                             // For a texture cache entry, create an invalid handle that
                             // will be allocated when update_picture_cache is called.
                             SurfaceTextureDescriptor::TextureCache {
@@ -2476,7 +2472,7 @@ impl TileCacheInstance {
 
         // If compositor mode is changed, need to drop all incompatible tiles.
         match frame_context.config.compositor_kind {
-            CompositorKind::Draw { .. } | CompositorKind::Layer { .. } => {
+            CompositorKind::Draw { .. } => {
                 for sub_slice in &mut self.sub_slices {
                     for tile in sub_slice.tiles.values_mut() {
                         if let Some(TileSurface::Texture { descriptor: SurfaceTextureDescriptor::Native { ref mut id, .. }, .. }) = tile.surface {
@@ -2825,7 +2821,7 @@ impl TileCacheInstance {
         // also determine whether this needs to be updated, depending on whether the
         // image generation(s) of the planes have changed since last composite.
         let (native_surface_id, update_params) = match composite_state.compositor_kind {
-            CompositorKind::Draw { .. } | CompositorKind::Layer { .. } => {
+            CompositorKind::Draw { .. } => {
                 (None, None)
             }
             CompositorKind::Native { .. } => {
@@ -5724,7 +5720,7 @@ impl PicturePrimitive {
                 let mut backdrop_in_use_and_visible = false;
                 if let Some(backdrop_rect) = backdrop_rect {
                     let supports_surface_for_backdrop = match frame_state.composite_state.compositor_kind {
-                        CompositorKind::Draw { .. } | CompositorKind::Layer { .. } => {
+                        CompositorKind::Draw { .. } => {
                             false
                         }
                         CompositorKind::Native { capabilities, .. } => {
diff --git a/gfx/wr/webrender/src/renderer/init.rs b/gfx/wr/webrender/src/renderer/init.rs
index 7d477a8dff39..0b250dfdf68f 100644
--- a/gfx/wr/webrender/src/renderer/init.rs
+++ b/gfx/wr/webrender/src/renderer/init.rs
@@ -20,7 +20,7 @@ use crate::glyph_cache::GlyphCache;
 use glyph_rasterizer::{GlyphRasterThread, GlyphRasterizer, SharedFontResources};
 use crate::gpu_types::PrimitiveInstanceData;
 use crate::internal_types::{FastHashMap, FastHashSet, FrameId};
-use crate::picture;
+use crate::{picture, Compositor2};
 use crate::profiler::{self, Profiler, TransactionProfile};
 use crate::device::query::{GpuProfiler, GpuDebugMethod};
 use crate::render_backend::RenderBackend;
@@ -202,6 +202,10 @@ pub struct WebRenderOptions {
     /// make the result look quite close to the high-quality zoom, except for glyphs.
     pub low_quality_pinch_zoom: bool,
     pub max_shared_surface_size: i32,
+    /// If supplied, composite the frame using the new experimental compositing
+    /// interface. If this is set, it overrides `compositor_config`. These will
+    /// be unified as the interface stabilises.
+    pub compositor2: Option<Box<dyn Compositor2>>,
 }
 
 impl WebRenderOptions {
@@ -274,6 +278,7 @@ impl Default for WebRenderOptions {
             reject_software_rasterizer: false,
             low_quality_pinch_zoom: false,
             max_shared_surface_size: 2048,
+            compositor2: None,
         }
     }
 }
@@ -318,13 +323,8 @@ pub fn create_webrender_instance(
     HAS_BEEN_INITIALIZED.store(true, Ordering::SeqCst);
 
     // For now, we assume that native OS compositors are top-left origin. If that doesn't
-    // turn out to be the case, we can add a query method on `LayerCompositor`.
-    match options.compositor_config {
-        CompositorConfig::Draw { .. } | CompositorConfig::Native { .. } => {}
-        CompositorConfig::Layer { .. } => {
-            options.surface_origin_is_top_left = true;
-        }
-    }
+    // turn out to be the case, we can add a query method on `Compositor2`.
+    options.surface_origin_is_top_left |= options.compositor2.is_some();
 
     let (api_tx, api_rx) = unbounded_channel();
     let (result_tx, result_rx) = unbounded_channel();
@@ -536,10 +536,6 @@ pub fn create_webrender_instance(
                 capabilities,
             }
         }
-        CompositorConfig::Layer { .. } => {
-            CompositorKind::Layer {
-            }
-        }
     };
 
     let config = FrameBuilderConfig {
@@ -814,6 +810,7 @@ pub fn create_webrender_instance(
         consecutive_oom_frames: 0,
         target_frame_publish_id: None,
         pending_result_msg: None,
+        compositor2: options.compositor2,
     };
 
     // We initially set the flags to default and then now call set_debug_flags
diff --git a/gfx/wr/webrender/src/renderer/mod.rs b/gfx/wr/webrender/src/renderer/mod.rs
index 8e62ed11d450..11f67d25504e 100644
--- a/gfx/wr/webrender/src/renderer/mod.rs
+++ b/gfx/wr/webrender/src/renderer/mod.rs
@@ -58,7 +58,7 @@ use crate::composite::{CompositeState, CompositeTileSurface, CompositorInputLaye
 use crate::composite::{CompositorKind, Compositor, NativeTileId, CompositeFeatures, CompositeSurfaceFormat, ResolvedExternalSurfaceColorData};
 use crate::composite::{CompositorConfig, NativeSurfaceOperationDetails, NativeSurfaceId, NativeSurfaceOperation};
 use crate::composite::TileKind;
-use crate::{debug_colors, CompositorInputConfig, CompositorSurfaceUsage};
+use crate::{debug_colors, Compositor2, CompositorInputConfig, CompositorSurfaceUsage};
 use crate::device::{DepthFunction, Device, DrawTarget, ExternalTexture, GpuFrameId, UploadPBOPool};
 use crate::device::{ReadTarget, ShaderError, Texture, TextureFilter, TextureFlags, TextureSlot, Texel};
 use crate::device::query::{GpuSampler, GpuTimer};
@@ -873,6 +873,7 @@ pub struct Renderer {
     /// The compositing config, affecting how WR composites into the final scene.
     compositor_config: CompositorConfig,
     current_compositor_kind: CompositorKind,
+    compositor2: Option<Box<dyn Compositor2>>,
 
     /// Maintains a set of allocated native composite surfaces. This allows any
     /// currently allocated surfaces to be cleaned up as soon as deinit() is
@@ -1705,7 +1706,7 @@ impl Renderer {
             //           with the (non-compositor) surface y-flip options.
             let surface_origin_is_top_left = match self.current_compositor_kind {
                 CompositorKind::Native { .. } => true,
-                CompositorKind::Draw { .. } | CompositorKind::Layer { .. } => self.device.surface_origin_is_top_left(),
+                CompositorKind::Draw { .. } => self.device.surface_origin_is_top_left(),
             };
             // If there is a debug overlay, render it. Otherwise, just clear
             // the debug renderer.
@@ -3335,7 +3336,7 @@ impl Renderer {
         }
     }
 
-    // Composite tiles in a swapchain. When using LayerCompositor, we may
+    // Composite tiles in a swapchain. When using Compositor2, we may
     // split the compositing in to multiple swapchains.
     fn composite_pass(
         &mut self,
@@ -3537,11 +3538,8 @@ impl Renderer {
                         (CompositorSurfaceUsage::Content, CompositorSurfaceUsage::External) |
                         (CompositorSurfaceUsage::External, CompositorSurfaceUsage::Content) |
                         (CompositorSurfaceUsage::External, CompositorSurfaceUsage::External) => {
-                            // Only create a new layer if we're using LayerCompositor
-                            match self.compositor_config {
-                                CompositorConfig::Draw { .. } | CompositorConfig::Native { .. } => None,
-                                CompositorConfig::Layer { .. } => Some(usage),
-                            }
+                            // Only create a new layer if we're using Compositor2
+                            self.compositor2.as_ref().map(|_| usage)
                         }
                     }
                 }
@@ -3610,7 +3608,7 @@ impl Renderer {
         }
 
         // Start compositing if using OS compositor
-        if let Some(ref mut compositor) = self.compositor_config.layer_compositor() {
+        if let Some(ref mut compositor) = self.compositor2 {
             let input = CompositorInputConfig {
                 layers: &input_layers,
                 framebuffer_size: fb_draw_target.dimensions(),
@@ -3628,8 +3626,8 @@ impl Renderer {
                 ColorF::TRANSPARENT
             };
 
-            let draw_target = match self.compositor_config {
-                CompositorConfig::Layer { ref mut compositor } => {
+            let draw_target = match self.compositor2 {
+                Some(ref mut compositor) => {
                     compositor.bind_layer(layer_index);
 
                     DrawTarget::NativeSurface {
@@ -3638,8 +3636,7 @@ impl Renderer {
                         dimensions: fb_draw_target.dimensions(),
                     }
                 }
-                // Native can be hit when switching compositors (disable when using Layer)
-                CompositorConfig::Draw { .. } | CompositorConfig::Native { .. } => {
+                None => {
                     fb_draw_target
                 }
             };
@@ -3658,13 +3655,13 @@ impl Renderer {
                 swapchain_layer,
             );
 
-            if let Some(ref mut compositor) = self.compositor_config.layer_compositor() {
+            if let Some(ref mut compositor) = self.compositor2 {
                 compositor.present_layer(layer_index);
             }
         }
 
         // End frame notify for experimental compositor
-        if let Some(ref mut compositor) = self.compositor_config.layer_compositor() {
+        if let Some(ref mut compositor) = self.compositor2 {
             for (layer_index, layer) in input_layers.iter().enumerate() {
                 compositor.add_surface(
                     layer_index,
@@ -4417,9 +4414,6 @@ impl Renderer {
             CompositorKind::Draw { draw_previous_partial_present_regions, max_partial_present_rects } => {
                 (max_partial_present_rects, draw_previous_partial_present_regions)
             }
-            CompositorKind::Layer { .. } => {
-                (0, false)
-            }
         };
 
         if max_partial_present_rects > 0 {
@@ -4574,7 +4568,7 @@ impl Renderer {
                     }
                 }
             }
-            CompositorConfig::Draw { .. } | CompositorConfig::Layer { .. } => {
+            CompositorConfig::Draw { .. } => {
                 // Ensure nothing is added in simple composite mode, since otherwise
                 // memory will leak as this doesn't get drained
                 debug_assert!(self.pending_native_surface_updates.is_empty());
@@ -4781,7 +4775,7 @@ impl Renderer {
                                         picture_target.valid_rect,
                                     )
                                 }
-                                CompositorKind::Draw { .. } | CompositorKind::Layer { .. } => {
+                                CompositorKind::Draw { .. } => {
                                     unreachable!();
                                 }
                             };
@@ -4818,7 +4812,7 @@ impl Renderer {
                                 let compositor = self.compositor_config.compositor().unwrap();
                                 compositor.unbind(&mut self.device);
                             }
-                            CompositorKind::Draw { .. } | CompositorKind::Layer { .. } => {
+                            CompositorKind::Draw { .. } => {
                                 unreachable!();
                             }
                         }
@@ -4938,7 +4932,7 @@ impl Renderer {
                         results,
                     );
                 }
-                CompositorKind::Draw { .. } | CompositorKind::Layer { .. } => {
+                CompositorKind::Draw { .. } => {
                     self.composite_simple(
                         &frame.composite_state,
                         draw_target,
diff --git a/gfx/wr/wrench/src/composite.cpp b/gfx/wr/wrench/src/composite.cpp
index 7746c260c573..1922d132ae11 100644
--- a/gfx/wr/wrench/src/composite.cpp
+++ b/gfx/wr/wrench/src/composite.cpp
@@ -139,7 +139,7 @@ class Compositor {
   ComPtr<IDXGIFactory2> pIDXGIFactory;
 };
 
-// Bindings called by wrench rust impl of the LayerCompositor trait
+// Bindings called by wrench rust impl of the Compositor2 trait
 extern "C" {
 Compositor* wrc_new(void* d3d11_device, void* hwnd) {
   return new Compositor(static_cast<ID3D11Device*>(d3d11_device),
diff --git a/gfx/wr/wrench/src/composite.rs b/gfx/wr/wrench/src/composite.rs
index 2c71ff798feb..13c44c2e5dcf 100644
--- a/gfx/wr/wrench/src/composite.rs
+++ b/gfx/wr/wrench/src/composite.rs
@@ -6,7 +6,7 @@ use std::os::raw::c_void;
 use std::ptr;
 
 use glutin::platform::windows::EGLContext;
-use webrender::{LayerCompositor, CompositorInputConfig};
+use webrender::{Compositor2, CompositorInputConfig};
 use winit::platform::windows::WindowExtWindows;
 
 use crate::WindowWrapper;
@@ -14,7 +14,7 @@ use crate::WindowWrapper;
 use mozangle::egl::ffi::types::{EGLDisplay, EGLSurface, EGLint};
 use mozangle::egl;
 
-// A simplistic implementation of the `LayerCompositor` trait to allow wrench to
+// A simplistic implementation of the `Compositor2` trait to allow wrench to
 // composite via DirectComposition. In this initial version, only a single
 // swap-chain is supported. Follow up patches will add layer and external
 // surface support.
@@ -69,7 +69,7 @@ impl WrLayer {
     }
 }
 
-// A basic `LayerCompositor` implementation for wrench
+// A basic `Compositor2` implementation for wrench
 pub struct WrCompositor {
     // EGL display and content, provided by winit
     context: EGLContext,
@@ -115,7 +115,7 @@ impl Drop for WrCompositor {
     }
 }
 
-impl LayerCompositor for WrCompositor {
+impl Compositor2 for WrCompositor {
     // Begin compositing a frame with the supplied input config
     // The main job of this method is to inspect the input config, create the output
     // config, and create any native OS resources that will be needed as layers get
diff --git a/gfx/wr/wrench/src/main.rs b/gfx/wr/wrench/src/main.rs
index fb008b00ff90..9bea13394931 100644
--- a/gfx/wr/wrench/src/main.rs
+++ b/gfx/wr/wrench/src/main.rs
@@ -48,7 +48,7 @@ use std::rc::Rc;
 #[cfg(feature = "software")]
 use std::slice;
 use std::sync::mpsc::{channel, Sender, Receiver};
-use webrender::{DebugFlags, LayerCompositor};
+use webrender::{DebugFlags, Compositor2};
 use webrender::api::*;
 use webrender::render_api::*;
 use webrender::api::units::*;
@@ -311,12 +311,12 @@ impl WindowWrapper {
     }
 
     #[cfg(target_os = "windows")]
-    pub fn create_compositor(&self) -> Option<Box<dyn LayerCompositor>> {
-        Some(Box::new(WrCompositor::new(self)) as Box<dyn LayerCompositor>)
+    pub fn create_compositor(&self) -> Option<Box<dyn Compositor2>> {
+        Some(Box::new(WrCompositor::new(self)) as Box<dyn Compositor2>)
     }
 
     #[cfg(not(target_os = "windows"))]
-    pub fn create_compositor(&self) -> Option<Box<dyn LayerCompositor>> {
+    pub fn create_compositor(&self) -> Option<Box<dyn Compositor2>> {
         None
     }
 }
@@ -734,7 +734,7 @@ pub fn main() {
         (None, None)
     };
 
-    let layer_compositor = if using_compositor {
+    let compositor2 = if using_compositor {
         window.create_compositor()
     } else {
         None
@@ -754,7 +754,7 @@ pub fn main() {
         args.is_present("precache"),
         dump_shader_source,
         notifier,
-        layer_compositor,
+        compositor2,
     );
 
     if let Some(ui_str) = args.value_of("profiler_ui") {
diff --git a/gfx/wr/wrench/src/wrench.rs b/gfx/wr/wrench/src/wrench.rs
index 80b9bade0ba7..26f348584e07 100644
--- a/gfx/wr/wrench/src/wrench.rs
+++ b/gfx/wr/wrench/src/wrench.rs
@@ -15,10 +15,10 @@ use std::path::{Path, PathBuf};
 use std::sync::{Arc, Mutex};
 use std::sync::mpsc::Receiver;
 use std::time::Instant;
-use webrender::{api::*, CompositorConfig};
+use webrender::api::*;
 use webrender::render_api::*;
 use webrender::api::units::*;
-use webrender::{DebugFlags, RenderResults, ShaderPrecacheFlags, LayerCompositor};
+use webrender::{DebugFlags, RenderResults, ShaderPrecacheFlags, Compositor2};
 use crate::{WindowWrapper, NotifierEvent};
 
 #[derive(Clone)]
@@ -229,7 +229,7 @@ impl Wrench {
         precache_shaders: bool,
         dump_shader_source: Option<String>,
         notifier: Option<Box<dyn RenderNotifier>>,
-        layer_compositor: Option<Box<dyn LayerCompositor>>,
+        compositor2: Option<Box<dyn Compositor2>>,
     ) -> Self {
         println!("Shader override path: {:?}", shader_override_path);
 
@@ -243,11 +243,6 @@ impl Wrench {
             ShaderPrecacheFlags::empty()
         };
 
-        let compositor_config = match layer_compositor {
-            Some(compositor) => CompositorConfig::Layer { compositor },
-            None => CompositorConfig::default(),
-        };
-
         let opts = webrender::WebRenderOptions {
             resource_override_path: shader_override_path,
             use_optimized_shaders,
@@ -264,7 +259,7 @@ impl Wrench {
             // SWGL doesn't support the GL_ALWAYS depth comparison function used by
             // `clear_caches_with_quads`, but scissored clears work well.
             clear_caches_with_quads: !window.is_software(),
-            compositor_config,
+            compositor2,
             ..Default::default()
         };
 
